---
output: github_document
---


# assigner <img src="docs/logo.png" align="right" alt="" />

```{r, echo = FALSE}
description <- readLines("DESCRIPTION")
rvers <- stringr::str_match(grep("R \\(", description, value = TRUE), "[0-9]{1,4}\\.[0-9]{1,4}\\.[0-9]{1,4}")[1,1]
version <- gsub(" ", "", gsub("Version:", "", grep("Version:", description, value = TRUE)))
```
<!-- badges: start -->
[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://tidyverse.org/lifecycle/#maturing)
[![Travis-CI Build Status](https://travis-ci.org/thierrygosselin/assigner.svg?branch=master)](https://travis-ci.org/thierrygosselin/assigner) 
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/assigner)](http://cran.r-project.org/package=assigner) 
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![DOI](https://zenodo.org/badge/14548/thierrygosselin/assigner.svg)](https://zenodo.org/badge/latestdoi/14548/thierrygosselin/assigner)
[![packageversion](https://img.shields.io/badge/Package%20version-`r version`-orange.svg)](commits/master)
[![Last-changedate](https://img.shields.io/badge/last%20change-`r gsub('-', '--', Sys.Date())`-brightgreen.svg)](/commits/master)
<!-- badges: end -->
```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/"
)
```


The name **assigner** |əˈsʌɪn| is rooted in the latin word *assignare*. It's first use in french dates back to XIIIe.

Genomic datasets produced by next-generation sequencing techniques that reduce the size of the genome (e.g. genotype-by-sequencing (GBS) and
restriction-site-associated DNA sequencing (RADseq)) have a huge numbers of markers that hold great potential and promises for assignment analysis. After hitting the bioinformatic wall with the different workflows you'll likely end up with several folders containing whitelist and blacklist of markers and individuals, data sets with various *de novo* and/or filtering parameters and ... missing data. This reality of GBS/RADseq data is quite hard on GUI software traditionally used for population assignment analysis. The end results are usually poor data exploration, constrained by time, and poor reproducibility.

**assigner** was tailored to make it easy to conduct population assignment analysis using GBS/RADseq data within R. Additionally, combining the use of tools like [R Notebook](http://rmarkdown.rstudio.com/r_notebooks.html), [RStudio](https://www.rstudio.com) and [GitHub](https://github.com) will make effortless documenting your workflows and pipelines.

The **keywords** here to remember: 

* 3 differents algorithms implemented: frequentist, likelihood and machine learning
* cross-validation techniques: classic Leave-One-Out (LOO) and Training, Holdout, Leave-one-out (THL)
with marker selection
* resampling/bootstrap/subsampling
* fast Fst WC84 implementation)
* ggplot2-based plotting! 
* [https://thierrygosselin.github.io/assigner/](https://thierrygosselin.github.io/assigner/) 


## Installation
To try out the dev version of **assigner**:

```r
if (!require("devtools")) install.packages("devtools")
devtools::install_github("thierrygosselin/assigner")
library(assigner)
```

If you plan on using `gsi_sim` inside assigner, you need an additional step:

With UNIX
```r
assigner::install_gsi_sim(fromSource = TRUE)
```

With PC
```r
assigner::install_gsi_sim()
```

* web site and additional info: [https://thierrygosselin.github.io/assigner/](https://thierrygosselin.github.io/assigner/) 
* [Computer setup - installation - troubleshooting](http://thierrygosselin.github.io/assigner/articles/rad_genomics_computer_setup.html)
* [assigner's assumptions](http://thierrygosselin.github.io/assigner/reference/assignment_ngs.html#assumptions)
* [assigner's features](http://thierrygosselin.github.io/assigner/FEATURES.html)
* [Function's documentation](http://thierrygosselin.github.io/assigner/reference/index.html)
* [Vignettes](http://thierrygosselin.github.io/assigner/articles/index.html)
* How to cite assigner: inside R type `citation("assigner")`

## [Life cycle](https://thierrygosselin.github.io/assigner/articles/life_cycle.html)

assigner is maturing, but in order to make the package better, changes are 
inevitable. Experimental functions will change, argument names will change.
Your codes and workflows might break from time to time until **assigner is stable**. 
Consequently, depending on your tolerance to change, assigner might not be for you.

* Philosophy, major changes and deprecated functions/arguments are documented in
life cycle section of functions. 
* The latest changes are documented ([here](https://thierrygosselin.github.io/assigner/articles/life_cycle.html)) and
in [changelog, versions, new features and bug history](https://thierrygosselin.github.io/assigner/news/index.html)
* [issues](https://github.com/thierrygosselin/assigner/issues/new/choose) and [contributions](https://github.com/thierrygosselin/assigner/issues/new/choose)


