<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • assigner</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><meta property="og:image" content="https://thierrygosselin.github.io/assigner/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">assigner</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../FEATURES.html">Features</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/get_started.html">Get started</a></li>
    <li><a class="dropdown-item" href="../articles/fst_confidence_intervals.html">Fst analysis</a></li>
    <li><a class="dropdown-item" href="../articles/web_only/fst_comparisons.html">Comparisons of Fst in 8 packages/software</a></li>
    <li><a class="dropdown-item" href="../articles/assignment.whitelists.html">Using whitelists in assignment</a></li>
    <li><a class="dropdown-item" href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a></li>
    <li><a class="dropdown-item" href="../articles/life_cycle.html">Life cycle</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/thierrygosselin/assigner" aria-label="GitHub"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="https://github.com/thierrygosselin/assigner/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.7.0" id="assigner-070-2025-07-01">assigner 0.7.0 2025-07-01<a class="anchor" aria-label="anchor" href="#assigner-070-2025-07-01"></a></h2>
<ul><li>
<em>assigner::fst_WC84</em>: work faster and as better reporting</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.0" id="assigner-060-2024-11-06">assigner 0.6.0 2024-11-06<a class="anchor" aria-label="anchor" href="#assigner-060-2024-11-06"></a></h2>
<ul><li>refresh for new R version</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.9" id="assigner-059-2023-03-15">assigner 0.5.9 2023-03-15<a class="anchor" aria-label="anchor" href="#assigner-059-2023-03-15"></a></h2>
<ul><li>work on <em>assigner::fst_WC84</em>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.8" id="assigner-058-2020-10-20">assigner 0.5.8 2020-10-20<a class="anchor" aria-label="anchor" href="#assigner-058-2020-10-20"></a></h2>
<ul><li>Version bump because it updates numerous packages: <code>tidyr</code>, <code>readr</code>, using <code>future</code>, <code>carrier</code>.</li>
<li>GitHub actions to run the <strong>R-CMD-check</strong> on the 3 OS.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.7" id="assigner-057-2020-01-21">assigner 0.5.7 2020-01-21<a class="anchor" aria-label="anchor" href="#assigner-057-2020-01-21"></a></h2>
<ul><li>Bug fix in heatmap: digits was fixed to 5 and when pop.levels = NULL, the heatmap was all mixed up. Thanks to <a href="https://github.com/siberianhigh" class="external-link">@siberianhigh</a> for highlighting the bug.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.6" id="assigner-056-2019-05-01">assigner 0.5.6 2019-05-01<a class="anchor" aria-label="anchor" href="#assigner-056-2019-05-01"></a></h2>
<ul><li>included 2 simulated datasets</li>
<li>updated documentation of <code>assignment_ngs</code>
</li>
<li>vignette to get started with assigner, finally!</li>
<li>this is really starting to smell like a CRAN release</li>
<li>work on travis CI</li>
<li>work on pkgdown</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.5" id="assigner-055-2019-04-25">assigner 0.5.5 2019-04-25<a class="anchor" aria-label="anchor" href="#assigner-055-2019-04-25"></a></h2>
<ul><li>cosmetic changes to the package: using <code>pkgdown</code>
</li>
<li>updated documentation of <code>assignment_ngs</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.4" id="assigner-054-2019-03-12">assigner 0.5.4 2019-03-12<a class="anchor" aria-label="anchor" href="#assigner-054-2019-03-12"></a></h2>
<ul><li>
<code>fst_WC84</code>: work faster</li>
<li>continue to integrate <code>assigner</code> with <code>SeqArray</code> and GDS object/file</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.3" id="assigner-053-2019-02-25">assigner 0.5.3 2019-02-25<a class="anchor" aria-label="anchor" href="#assigner-053-2019-02-25"></a></h2>
<ul><li>
<code>fst_WC84</code>: work with radiator v.1.0</li>
<li>will continue updating fucntions to work with latest radiator release and work toward releasing the official v.1.0 of assigner.</li>
<li>Imputation module was removed from <code>assigner</code> and now lives exclusively in package <code>grur</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.2" id="assigner-052-2018-07-09">assigner 0.5.2 2018-07-09<a class="anchor" aria-label="anchor" href="#assigner-052-2018-07-09"></a></h2>
<ul><li>working to make assigner work correctly with ggplot2 v.3.0.0</li>
<li>assigner ready for R 3.5.1 “Feather Spray” released on 2018/07/05</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.1" id="assigner-051-2018-06-21">assigner 0.5.1 2018-06-21<a class="anchor" aria-label="anchor" href="#assigner-051-2018-06-21"></a></h2>
<ul><li>bug in <code>assignment_mixture</code> generated by <code>purrr::df</code> replaced recently by <code>purrr:dfr</code>. Changed <code>DESCRIPTION</code> field accordingly.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.0" id="assigner-050-2017-12-12">assigner 0.5.0 2017-12-12<a class="anchor" aria-label="anchor" href="#assigner-050-2017-12-12"></a></h2>
<ul><li>
<code>subsample</code> argument in <code>assignment_ngs</code> and <code>assignment_mixture</code> can now automatically detect the smallest sample size in the data’s grouping. So you can use <code>subsample = "min"</code> to let the function decide (if your not sure).</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.9" id="assigner-049-2017-08-18">assigner 0.4.9 2017-08-18<a class="anchor" aria-label="anchor" href="#assigner-049-2017-08-18"></a></h2>
<ul><li>migration of <code>assigner</code> from using <code>stackr</code> -&gt; <code>radiator</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.8" id="assigner-048-2017-08-15">assigner 0.4.8 2017-08-15<a class="anchor" aria-label="anchor" href="#assigner-048-2017-08-15"></a></h2>
<ul><li>restored progress bar when using parallel computing by installing the new dev version of <code>pbmcapply</code> package.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.7" id="assigner-047-2017-08-15">assigner 0.4.7 2017-08-15<a class="anchor" aria-label="anchor" href="#assigner-047-2017-08-15"></a></h2>
<ul><li>bug fix: removed the progress bar when using parallel computing. This is temporary, while waiting for a fix with <code>pbmcapply</code> package.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.6" id="assigner-046-2017-06-17">assigner 0.4.6 2017-06-17<a class="anchor" aria-label="anchor" href="#assigner-046-2017-06-17"></a></h2>
<ul><li>
<strong>assigner</strong> works with <code>dplyr v.0.7.0</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.5" id="assigner-045-2017-04-25">assigner 0.4.5 2017-04-25<a class="anchor" aria-label="anchor" href="#assigner-045-2017-04-25"></a></h2>
<ul><li>
<code>dlr</code>: simplified arguments, faster function and now creates the Dlr plots</li>
<li>dependencies to package <code>SNPRelate</code> are removed until the bugs with Fst calculation are resolved.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.4" id="assigner-044-2017-04-12">assigner 0.4.4 2017-04-12<a class="anchor" aria-label="anchor" href="#assigner-044-2017-04-12"></a></h2>
<ul><li>bug fix in <code>assignment_ngs</code> introduced in last commit that was suppose to be fix. Problem introduced by <code>stackr::change_pop_names</code>.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.3" id="assigner-043-2017-04-02">assigner 0.4.3 2017-04-02<a class="anchor" aria-label="anchor" href="#assigner-043-2017-04-02"></a></h2>
<ul><li>it’s now official, <code>assigner</code> as a logo</li>
<li>faster <code>fst_NEI87</code>
</li>
<li>unbalanced design impact on estimates can be tested with <code>subsample</code> and <code>iteration.subsample</code> in <code>fst_NEI87</code> and <code>fst_WC84</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.2" id="assigner-042">assigner 0.4.2<a class="anchor" aria-label="anchor" href="#assigner-042"></a></h2>
<ul><li>until <code>SNPRelate</code> <a href="https://github.com/zhengxwen/SNPRelate/issues/21" class="external-link">bias issue</a> is resolved the option is unavailable</li>
<li>better use of <code>pbmcapply</code> for Windows</li>
<li>
<strong>imputations</strong> is being reworked and will be buggy until the next update. The codes are being completely re-written and arguments will change (for the better).</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.1" id="assigner-041">assigner 0.4.1<a class="anchor" aria-label="anchor" href="#assigner-041"></a></h2>
<ul><li>debug code to work in parallel with Windows</li>
<li>code cleaning to prep for CRAN</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.0" id="assigner-040">assigner 0.4.0<a class="anchor" aria-label="anchor" href="#assigner-040"></a></h2>
<ul><li>
<code>assignment_ngs</code> and <code>assignment_mixture</code> code cleaning to prep for CRAN and make them easier to debug.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.9" id="assigner-039">assigner 0.3.9<a class="anchor" aria-label="anchor" href="#assigner-039"></a></h2>
<ul><li>I’m pleased to announce that <code>assigner</code> now works in parallel with <strong>Windows</strong>
</li>
<li>bug fix introduce in last commit in <code>write_gsi_sim</code> where the file was not created properly from an internal module.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.8" id="assigner-038">assigner 0.3.8<a class="anchor" aria-label="anchor" href="#assigner-038"></a></h2>
<ul><li>
<code><a href="../reference/fst_WC84.html">assigner::fst_WC84</a></code> can now use <a href="https://github.com/zhengxwen/SNPRelate" class="external-link">SNPRelate</a> to compute Fst. The confidence intervals are not implemented, yet. The speed increase left me speechless, dataset with 30K snp are computed in less than 15 sec!</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.7" id="assigner-037">assigner 0.3.7<a class="anchor" aria-label="anchor" href="#assigner-037"></a></h2>
<ul><li>
<code><a href="../reference/fst_WC84.html">assigner::fst_WC84</a></code> is 40% faster!</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.6" id="assigner-036">assigner 0.3.6<a class="anchor" aria-label="anchor" href="#assigner-036"></a></h2>
<ul><li>bug fix <code>assignment_ngs</code> during imputations, the imputation module could not recognise that REF/ALT alleles are not necessary or usefull for assignment analysis. *enhancement to <code>assignment_ngs</code> and <code>assignment_mixture</code> so that when <code>marker.number</code> include <code>"all"</code> the <code>iteration.method</code> is set automatically to <code>1</code> when conducting the assignment with all the markers. Iterations at this point is useless and a waist of time.</li>
<li>random seed number is now stored in the appropriate files.</li>
<li>
<code>assignment_mixture</code>: with <code>assignment.analysis = "gsi_sim</code> the unknown/mixture samples are compared with baseline populations using common markers between the pair. Now, the tables include the number of markers used. The summary provides the mean number of markers. This number will change each time randomness is used.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.5" id="assigner-035">assigner 0.3.5<a class="anchor" aria-label="anchor" href="#assigner-035"></a></h2>
<ul><li>bug fix in population not recognise properly</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.4" id="assigner-034">assigner 0.3.4<a class="anchor" aria-label="anchor" href="#assigner-034"></a></h2>
<ul><li>
<code>fst_NEI87</code>: very fast function that can compute: the overall and pairwise Nei’s (1987) fst and f’st (prime). Bootstrap resampling of markers is avalaible to build Confidence Intervals. The estimates are available as a data frame and a matrix with upper diagonal filled with Fst values and lower diagonal filled with the confidence intervals. Jost’s D is also given ;)</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.3" id="assigner-033">assigner 0.3.3<a class="anchor" aria-label="anchor" href="#assigner-033"></a></h2>
<ul><li>
<code>fst_WC84</code>: bug fix, the function was not properly configured for multi-allelic markers (e.g. microsatellite, and haplotype format from STACKS). Thanks to Craig McDougall for catching this.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.2" id="assigner-032">assigner 0.3.2<a class="anchor" aria-label="anchor" href="#assigner-032"></a></h2>
<ul><li>
<code>assignment_mixture</code>: added a check to throw an error when pop.levels != the pop.id in strata</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.1" id="assigner-031">assigner 0.3.1<a class="anchor" aria-label="anchor" href="#assigner-031"></a></h2>
<p><code>assignment_mixture</code>: * updated with latest modules from <code>stackr</code>. * simplified the identification of mixture or unknown samples. See doc.</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.0" id="assigner-030">assigner 0.3.0<a class="anchor" aria-label="anchor" href="#assigner-030"></a></h2>
<ul><li>updated vignettes</li>
<li>major bug fix that involved dplyr new version (0.5.0) and mostly with the use of dplyr::distinct</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.9" id="assigner-029">assigner 0.2.9<a class="anchor" aria-label="anchor" href="#assigner-029"></a></h2>
<ul><li>updated vignettes</li>
<li>bug fix in <code>fst_WC84</code>
</li>
<li>bug fix between assinger -&gt; devtools -&gt; github -&gt; travis, [this page helped] (<a href="http://itsalocke.com/using-travis-make-sure-use-github-pat/" class="external-link uri">http://itsalocke.com/using-travis-make-sure-use-github-pat/</a>)</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.8" id="assigner-028">assigner 0.2.8<a class="anchor" aria-label="anchor" href="#assigner-028"></a></h2>
<ul><li>While changing some lines with <code><a href="https://tidyr.tidyverse.org/reference/spread.html" class="external-link">tidyr::spread</a></code> and <code><a href="https://tidyr.tidyverse.org/reference/gather.html" class="external-link">tidyr::gather</a></code> for <code><a href="https://rdatatable.gitlab.io/data.table/reference/dcast.data.table.html" class="external-link">data.table::dcast.data.table</a></code> and <code><a href="https://rdatatable.gitlab.io/data.table/reference/melt.data.table.html" class="external-link">data.table::melt.data.table</a></code> to make the code faster, I forgot to split genotype into alleles for <code>gsi_sim</code>.</li>
<li>please update both <strong>stackr</strong> and <strong>assigner</strong>
<ul><li>the build error from Travis will be fixed soon. It should not affect the package “experience”” in any way.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.7" id="assigner-027">assigner 0.2.7<a class="anchor" aria-label="anchor" href="#assigner-027"></a></h2>
<ul><li><p>you need to update [stackr] (<a href="https://github.com/thierrygosselin/stackr" class="external-link uri">https://github.com/thierrygosselin/stackr</a>) to v.0.2.7 to appreciate this new version of assigner.</p></li>
<li><p>updated <code>assignment_ngs</code> with the separate stackr modules to simplify the function.</p></li>
<li><p>new data file available for <code>assignment_ngs</code>: <code>genepop</code> and <code>genind</code> object.</p></li>
<li><p><code>assignment_ngs</code> now accept any vcf input file! i.e. it’s no longer limited to stacks vcf.</p></li>
<li><p>new arguments in <code>assignment_ngs</code>. The assignment using dapc can now use the optimized alpha score <code>adegenet.dapc.opt == "optim.a.score"</code> or the cross-validation <code>adegenet.dapc.opt == "xval"</code>. This is useful for fine tuning the trade-off between power of discrimination and over-fitting (for stability of group membership probabilities). Cross validation with <code>adegenet.dapc.opt == "xval"</code> doesn’t work with missing data, so it’s only available with <em>imputed data</em> (i.e. <code>imputation.method == "rf" or "max"</code>). With non imputed data or the default: the optimized alpha-score is used (<code>adegenet.dapc.opt == "optim.a.score"</code>). When using <code>adegenet.dapc.opt == "xval"</code>, 2 new arguments are available:</p></li>
</ul><ol style="list-style-type: decimal"><li>
<code>adegenet.n.rep</code> and (2) <code>adegenet.training</code>. See documentation for details.</li>
</ol><ul><li><p>removed arguments in <code>assignment_ngs</code>. Removed the <code>pop.id.start</code> and <code>pop.id.end</code> arguments that were confusing people. For those used to these arguments, they are now recycled in the new function <code>individuals2strata</code> in [stackr] (<a href="https://github.com/thierrygosselin/stackr" class="external-link uri">https://github.com/thierrygosselin/stackr</a>). The strata file created by this function can be used with the <code>strata</code> argument in <code>assignment_ngs</code>.</p></li>
<li><p>2 modified arguments in <code>assignment_ngs</code>: (1) <code>gsi_sim.filename</code> is now <code>filename</code>; and</p></li>
</ul><ol start="2" style="list-style-type: decimal"><li>if you didn’t use the imputation argument, replace <code>imputation.method = FALSE</code> to <code>imputation.method = NULL</code> or leave the argument missing.</li>
</ol><ul><li><p>simplified sections of codes in <code>assignment_ngs</code> that dealt with <code>strata</code>, <code>pop.levels</code> and <code>pop.labels</code>.</p></li>
<li><p>new function: <code>write_gsi_sim</code>. Write a gsi_sim file from a data frame (wide or long/tidy). Used internally in [assigner] (<a href="https://github.com/thierrygosselin/assigner" class="external-link uri">https://github.com/thierrygosselin/assigner</a>) and might be of interest for users.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.6" id="assigner-026">assigner 0.2.6<a class="anchor" aria-label="anchor" href="#assigner-026"></a></h2>
<ul><li>Added a <code>NEWS.md</code> file to track changes to the package.</li>
<li>
<code>fst_WC84</code> is now a separate and very fast function that can compute: the overall and pairwise Weir and Cockerham 1984 Theta/Fst. Bootstrap resampling of markers is avalaible to build Confidence Intervals (For Louis Bernatchez and his students;). The estimates are available as a data frame and a matrix with upper diagonal filled with Fst values and lower diagonal filled with the confidence intervals.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.5" id="assigner-025">assigner 0.2.5<a class="anchor" aria-label="anchor" href="#assigner-025"></a></h2>
<ul><li>cleaner code for strata section</li>
<li>bug fix restricted to <code>assignment_ngs</code> + <code>assignment.analysis = "adegenet"</code> + <code>sampling.method = "ranked"</code>. A line at the beginning of a gsi_sim code section was deleted makig the assignment with adegenet go through that chunk of code and causing 100% assignment! if (assignment.analysis = “gsi_sim”) {code} prevent this problem…</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.4" id="assigner-024">assigner 0.2.4<a class="anchor" aria-label="anchor" href="#assigner-024"></a></h2>
<ul><li>bug fixed using adegenet that was introduced in v.0.2.3</li>
<li>introducing a new function <code>import_subsamples_fst</code> to import the fst ranking results from all the subsample runs inside an assignment folder.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.3" id="assigner-023">assigner 0.2.3<a class="anchor" aria-label="anchor" href="#assigner-023"></a></h2>
<ul><li>bug fixed in the compilation results section when no pop.id.start and end are used.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.1" id="assigner-021">assigner 0.2.1<a class="anchor" aria-label="anchor" href="#assigner-021"></a></h2>
<ul><li>updated the function <code>assignment_mixture</code> with <code>sampling.method = "ranked"</code> and <code>assignment.analysis = "adegenet"</code>.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.0" id="assigner-020">assigner 0.2.0<a class="anchor" aria-label="anchor" href="#assigner-020"></a></h2>
<ul><li>new function: <code>assignment_mixture</code> for mixture analysis.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.9" id="assigner-019">assigner 0.1.9<a class="anchor" aria-label="anchor" href="#assigner-019"></a></h2>
<ul><li>Simplified gsi_sim install</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.8" id="assigner-018">assigner 0.1.8<a class="anchor" aria-label="anchor" href="#assigner-018"></a></h2>
<ul><li>You can now opt between [gsi_sim] (<a href="https://github.com/eriqande/gsi_sim" class="external-link uri">https://github.com/eriqande/gsi_sim</a>) or [adegenet] (<a href="https://github.com/thibautjombart/adegenet" class="external-link uri">https://github.com/thibautjombart/adegenet</a>), a R package developed by Thibaul Jombart, to conduct the assignment analysis</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.7" id="assigner-017">assigner 0.1.7<a class="anchor" aria-label="anchor" href="#assigner-017"></a></h2>
<ul><li>New input file: Re-introduced the haplotype data frame file from stacks.</li>
<li>Argument name change: <code>imputations</code> is now <code>impute.method</code>.</li>
<li>New argument: <code>impute</code> with 2 options: <code>impute = "genotype"</code> or <code>impute = "allele"</code>.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.6" id="assigner-016">assigner 0.1.6<a class="anchor" aria-label="anchor" href="#assigner-016"></a></h2>
<ul><li>Input file argument is now <code>data</code> and covers the three types of files the function can use: VCF file, PLINK tped/tfam or data frame of genotypes file.</li>
<li>Huge number of markers (&gt; 50 000 markers) can now be imported in PLINK tped/tfam format. The first 2 columns of the <code>tfam</code> file will be used for the <code>strata</code> argument, unless a new one is provided. Columns 1, 3 and 4 of the <code>tped</code> are discarded. The remaining columns correspond to the genotype in the format <code>01/04</code> where <code>A = 01, C = 02, G = 03 and T = 04</code>. For <code>A/T</code> format, use PLINK or bash to convert. Use [VCFTOOLS] (<a href="http://vcftools.sourceforge.net/" class="external-link uri">http://vcftools.sourceforge.net/</a>) with <code>--plink-tped</code> to convert very large VCF file. For <code>.ped</code> file conversion to <code>.tped</code> use [PLINK] (<a href="http://pngu.mgh.harvard.edu/~purcell/plink/" class="external-link uri">http://pngu.mgh.harvard.edu/~purcell/plink/</a>) with <code>--recode transpose</code>.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.5" id="assigner-015">assigner 0.1.5<a class="anchor" aria-label="anchor" href="#assigner-015"></a></h2>
<ul><li>bug fix in <code>method = "random"</code> and <code>imputation</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.4" id="assigner-014">assigner 0.1.4<a class="anchor" aria-label="anchor" href="#assigner-014"></a></h2>
<ul><li>Changed function name, from <code>GBS_assignment</code> to <code>assignment_ngs</code>. Stands for assignment with next-generation sequencing data.</li>
<li>New argument <code>df.file</code> if you don’t have a VCF file. See documentation.</li>
<li>New argument <code>strata</code> if you don’t have population id or other metadata info in the individual name. See documentation.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.3" id="assigner-013">assigner 0.1.3<a class="anchor" aria-label="anchor" href="#assigner-013"></a></h2>
<ul><li>Changed arguments <code>THL</code> to <code>thl</code> and <code>snp.LD</code> to <code>snp.ld</code> to follow convention.</li>
<li>
<code>iterations.subsample</code> changed to <code>iteration.subsample</code>.</li>
<li>
<code>iterations</code> changed to <code>iteration.method</code> to avoid confusion with other iteration arguments.</li>
<li>Removed <code>baseline</code> and <code>mixture</code> arguments from the function <code>GBS_assignment</code>. These options will be re-introduce later in a separate function.</li>
<li>Using <code>marker.number</code> higher than the number of markers in the data set was causing problems. This could arise when using arguments that removed markers from the dataset (e.g. <code>snp.ld</code>, <code>common.markers</code>, and <code>maf</code> filters).</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.2" id="assigner-012">assigner 0.1.2<a class="anchor" aria-label="anchor" href="#assigner-012"></a></h2>
<ul><li>new version to update with gsi_sim new install instruction for Linux and Mac. After re-installing assigner package, follow the instruction to re-install the new [gsi_sim] (<a href="https://github.com/eriqande/gsi_sim" class="external-link uri">https://github.com/eriqande/gsi_sim</a>). And delete the old binary ‘gsisim’ in the /usr/local/bin folder with the following Terminal command: <code>sudo rm /usr/local/bin/gsisim</code>
</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

