<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • assigner</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />
<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Changelog" />

<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">assigner</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/fst_confidence_intervals.html">Fst analysis</a>
    </li>
    <li>
      <a href="../articles/assignment.whitelists.html">Using whitelists in assignment</a>
    </li>
    <li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../news/index.html">Change log</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/assigner">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small></small></h1>
      <small>Source: <a href='https://github.com/thierrygosselin/assigner/blob/master/NEWS.md'><code>NEWS.md</code></a></small>
    </div>

    <div id="assigner-0-5-5-2019-04-25" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-5-5-2019-04-25" class="anchor"></a>assigner 0.5.5 2019-04-25</h1>
<ul>
<li>cosmetic changes to the package: using <code>pkgdown</code>
</li>
<li>updated documentation of <code>assignment_ngs</code>
</li>
</ul>
</div>
    <div id="assigner-0-5-4-2019-03-12" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-5-4-2019-03-12" class="anchor"></a>assigner 0.5.4 2019-03-12</h1>
<ul>
<li>
<code>fst_WC84</code>: work faster</li>
<li>continue to integrate <code>assigner</code> with <code>SeqArray</code> and GDS object/file</li>
</ul>
</div>
    <div id="assigner-0-5-3-2019-02-25" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-5-3-2019-02-25" class="anchor"></a>assigner 0.5.3 2019-02-25</h1>
<ul>
<li>
<code>fst_WC84</code>: work with radiator v.1.0</li>
<li>will continue updating fucntions to work with latest radiator release and work toward releasing the official v.1.0 of assigner.</li>
<li>Imputation module was removed from <code>assigner</code> and now lives exclusively in package <code>grur</code>
</li>
</ul>
</div>
    <div id="assigner-0-5-2-2018-07-09" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-5-2-2018-07-09" class="anchor"></a>assigner 0.5.2 2018-07-09</h1>
<ul>
<li>working to make assigner work correctly with ggplot2 v.3.0.0</li>
<li>assigner ready for R 3.5.1 “Feather Spray” released on 2018/07/05</li>
</ul>
</div>
    <div id="assigner-0-5-1-2018-06-21" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-5-1-2018-06-21" class="anchor"></a>assigner 0.5.1 2018-06-21</h1>
<ul>
<li>bug in <code>assignment_mixture</code> generated by <code>purrr::df</code> replaced recently by <code>purrr:dfr</code>. Changed <code>DESCRIPTION</code> field accordingly.</li>
</ul>
</div>
    <div id="assigner-0-5-0-2017-12-12" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-5-0-2017-12-12" class="anchor"></a>assigner 0.5.0 2017-12-12</h1>
<ul>
<li>
<code>subsample</code> argument in <code>assignment_ngs</code> and <code>assignment_mixture</code> can now automatically detect the smallest sample size in the data’s grouping. So you can use <code>subsample = "min"</code> to let the function decide (if your not sure).</li>
</ul>
</div>
    <div id="assigner-0-4-9-2017-08-18" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-4-9-2017-08-18" class="anchor"></a>assigner 0.4.9 2017-08-18</h1>
<ul>
<li>migration of <code>assigner</code> from using <code>stackr</code> -&gt; <code>radiator</code>
</li>
</ul>
</div>
    <div id="assigner-0-4-8-2017-08-15" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-4-8-2017-08-15" class="anchor"></a>assigner 0.4.8 2017-08-15</h1>
<ul>
<li>restored progress bar when using parallel computing by installing the new dev version of <code>pbmcapply</code> package.</li>
</ul>
</div>
    <div id="assigner-0-4-7-2017-08-15" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-4-7-2017-08-15" class="anchor"></a>assigner 0.4.7 2017-08-15</h1>
<ul>
<li>bug fix: removed the progress bar when using parallel computing. This is temporary, while waiting for a fix with <code>pbmcapply</code> package.</li>
</ul>
</div>
    <div id="assigner-0-4-6-2017-06-17" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-4-6-2017-06-17" class="anchor"></a>assigner 0.4.6 2017-06-17</h1>
<ul>
<li>
<strong>assigner</strong> works with <code>dplyr v.0.7.0</code>
</li>
</ul>
</div>
    <div id="assigner-0-4-5-2017-04-25" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-4-5-2017-04-25" class="anchor"></a>assigner 0.4.5 2017-04-25</h1>
<ul>
<li>
<code>dlr</code>: simplified arguments, faster function and now creates the Dlr plots</li>
<li>dependencies to package <code>SNPRelate</code> are removed until the bugs with Fst calculation are resolved.</li>
</ul>
</div>
    <div id="assigner-0-4-4-2017-04-12" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-4-4-2017-04-12" class="anchor"></a>assigner 0.4.4 2017-04-12</h1>
<ul>
<li>bug fix in <code>assignment_ngs</code> introduced in last commit that was suppose to be fix. Problem introduced by <code><a href="https://www.rdocumentation.org/packages/stackr/topics/change_pop_names">stackr::change_pop_names</a></code>.</li>
</ul>
</div>
    <div id="assigner-0-4-3-2017-04-02" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-4-3-2017-04-02" class="anchor"></a>assigner 0.4.3 2017-04-02</h1>
<ul>
<li>it’s now official, <code>assigner</code> as a logo</li>
<li>faster <code>fst_NEI87</code>
</li>
<li>unbalanced design impact on estimates can be tested with <code>subsample</code> and <code>iteration.subsample</code> in <code>fst_NEI87</code> and <code>fst_WC84</code>
</li>
</ul>
</div>
    <div id="assigner-0-4-2" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-4-2" class="anchor"></a>assigner 0.4.2</h1>
<ul>
<li>until <code>SNPRelate</code> <a href="https://github.com/zhengxwen/SNPRelate/issues/21">bias issue</a> is resolved the option is unavailable</li>
<li>better use of <code>pbmcapply</code> for Windows</li>
<li>
<strong>imputations</strong> is being reworked and will be buggy until the next update. The codes are being completely re-written and arguments will change (for the better).</li>
</ul>
</div>
    <div id="assigner-0-4-1" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-4-1" class="anchor"></a>assigner 0.4.1</h1>
<ul>
<li>debug code to work in parallel with Windows</li>
<li>code cleaning to prep for CRAN</li>
</ul>
</div>
    <div id="assigner-0-4-0" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-4-0" class="anchor"></a>assigner 0.4.0</h1>
<ul>
<li>
<code>assignment_ngs</code> and <code>assignment_mixture</code> code cleaning to prep for CRAN and make them easier to debug.</li>
</ul>
</div>
    <div id="assigner-0-3-9" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-3-9" class="anchor"></a>assigner 0.3.9</h1>
<ul>
<li>I’m pleased to announce that <code>assigner</code> now works in parallel with <strong>Windows</strong>
</li>
<li>bug fix introduce in last commit in <code>write_gsi_sim</code> where the file was not created properly from an internal module.</li>
</ul>
</div>
    <div id="assigner-0-3-8" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-3-8" class="anchor"></a>assigner 0.3.8</h1>
<ul>
<li>
<code><a href="../reference/fst_WC84.html">assigner::fst_WC84</a></code> can now use <a href="https://github.com/zhengxwen/SNPRelate">SNPRelate</a> to compute Fst. The confidence intervals are not implemented, yet. The speed increase left me speechless, dataset with 30K snp are computed in less than 15 sec!</li>
</ul>
</div>
    <div id="assigner-0-3-7" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-3-7" class="anchor"></a>assigner 0.3.7</h1>
<ul>
<li>
<code><a href="../reference/fst_WC84.html">assigner::fst_WC84</a></code> is 40% faster!</li>
</ul>
</div>
    <div id="assigner-0-3-6" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-3-6" class="anchor"></a>assigner 0.3.6</h1>
<ul>
<li>bug fix <code>assignment_ngs</code> during imputations, the imputation module could not recognise that REF/ALT alleles are not necessary or usefull for assignment analysis. *enhancement to <code>assignment_ngs</code> and <code>assignment_mixture</code> so that when <code>marker.number</code> include <code>"all"</code> the <code>iteration.method</code> is set automatically to <code>1</code> when conducting the assignment with all the markers. Iterations at this point is useless and a waist of time.</li>
<li>random seed number is now stored in the appropriate files.</li>
<li>
<code>assignment_mixture</code>: with <code>assignment.analysis = "gsi_sim</code> the unknown/mixture samples are compared with baseline populations using common markers between the pair. Now, the tables include the number of markers used. The summary provides the mean number of markers. This number will change each time randomness is used.</li>
</ul>
</div>
    <div id="assigner-0-3-5" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-3-5" class="anchor"></a>assigner 0.3.5</h1>
<ul>
<li>bug fix in population not recognise properly</li>
</ul>
</div>
    <div id="assigner-0-3-4" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-3-4" class="anchor"></a>assigner 0.3.4</h1>
<ul>
<li>
<code>fst_NEI87</code>: very fast function that can compute: the overall and pairwise Nei’s (1987) fst and f’st (prime). Bootstrap resampling of markers is avalaible to build Confidence Intervals. The estimates are available as a data frame and a matrix with upper diagonal filled with Fst values and lower diagonal filled with the confidence intervals. Jost’s D is also given ;)</li>
</ul>
</div>
    <div id="assigner-0-3-3" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-3-3" class="anchor"></a>assigner 0.3.3</h1>
<ul>
<li>
<code>fst_WC84</code>: bug fix, the function was not properly configured for multi-allelic markers (e.g. microsatellite, and haplotype format from STACKS). Thanks to Craig McDougall for catching this.</li>
</ul>
</div>
    <div id="assigner-0-3-2" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-3-2" class="anchor"></a>assigner 0.3.2</h1>
<ul>
<li>
<code>assignment_mixture</code>: added a check to throw an error when pop.levels != the pop.id in strata</li>
</ul>
</div>
    <div id="assigner-0-3-1" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-3-1" class="anchor"></a>assigner 0.3.1</h1>
<p><code>assignment_mixture</code>:</p>
<ul>
<li>updated with latest modules from <code>stackr</code>.</li>
<li>simplified the identification of mixture or unknown samples. See doc.</li>
</ul>
</div>
    <div id="assigner-0-3-0" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-3-0" class="anchor"></a>assigner 0.3.0</h1>
<ul>
<li>updated vignettes</li>
<li>major bug fix that involved dplyr new version (0.5.0) and mostly with the use of dplyr::distinct</li>
</ul>
</div>
    <div id="assigner-0-2-9" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-2-9" class="anchor"></a>assigner 0.2.9</h1>
<ul>
<li>updated vignettes</li>
<li>bug fix in <code>fst_WC84</code>
</li>
<li>bug fix between assinger -&gt; devtools -&gt; github -&gt; travis, [this page helped] (<a href="http://itsalocke.com/using-travis-make-sure-use-github-pat/" class="uri">http://itsalocke.com/using-travis-make-sure-use-github-pat/</a>)</li>
</ul>
</div>
    <div id="assigner-0-2-8" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-2-8" class="anchor"></a>assigner 0.2.8</h1>
<ul>
<li>While changing some lines with <code><a href="https://tidyr.tidyverse.org/reference/spread.html">tidyr::spread</a></code> and <code><a href="https://tidyr.tidyverse.org/reference/gather.html">tidyr::gather</a></code> for <code><a href="https://www.rdocumentation.org/packages/data.table/topics/dcast.data.table">data.table::dcast.data.table</a></code> and <code><a href="https://www.rdocumentation.org/packages/data.table/topics/melt.data.table">data.table::melt.data.table</a></code> to make the code faster, I forgot to split genotype into alleles for <code>gsi_sim</code>.</li>
<li>please update both <strong>stackr</strong> and <strong>assigner</strong>
</li>
<li>the build error from Travis will be fixed soon. It should not affect the package “experience”" in any way.</li>
</ul>
</div>
    <div id="assigner-0-2-7" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-2-7" class="anchor"></a>assigner 0.2.7</h1>
<ul>
<li><p>you need to update [stackr] (<a href="https://github.com/thierrygosselin/stackr" class="uri">https://github.com/thierrygosselin/stackr</a>) to v.0.2.7 to appreciate this new version of assigner.</p></li>
<li><p>updated <code>assignment_ngs</code> with the separate stackr modules to simplify the function.</p></li>
<li><p>new data file available for <code>assignment_ngs</code>: <code>genepop</code> and <code>genind</code> object.</p></li>
<li><p><code>assignment_ngs</code> now accept any vcf input file! i.e. it’s no longer limited to stacks vcf.</p></li>
<li>new arguments in <code>assignment_ngs</code>. The assignment using dapc can now use the optimized alpha score <code>adegenet.dapc.opt == "optim.a.score"</code> or the cross-validation <code>adegenet.dapc.opt == "xval"</code>. This is useful for fine tuning the trade-off between power of discrimination and over-fitting (for stability of group membership probabilities). Cross validation with <code>adegenet.dapc.opt == "xval"</code> doesn’t work with missing data, so it’s only available with <em>imputed data</em> (i.e. <code>imputation.method == "rf" or "max"</code>). With non imputed data or the default: the optimized alpha-score is used (<code>adegenet.dapc.opt == "optim.a.score"</code>). When using <code>adegenet.dapc.opt == "xval"</code>, 2 new arguments are available:</li>
</ul>
<p>(1) <code>adegenet.n.rep</code> and (2) <code>adegenet.training</code>. See documentation for details.</p>
<ul>
<li><p>removed arguments in <code>assignment_ngs</code>. Removed the <code>pop.id.start</code> and <code>pop.id.end</code> arguments that were confusing people. For those used to these arguments, they are now recycled in the new function <code>individuals2strata</code> in [stackr] (<a href="https://github.com/thierrygosselin/stackr" class="uri">https://github.com/thierrygosselin/stackr</a>). The strata file created by this function can be used with the <code>strata</code> argument in <code>assignment_ngs</code>.</p></li>
<li>2 modified arguments in <code>assignment_ngs</code>: (1) <code>gsi_sim.filename</code> is now <code>filename</code>; and</li>
</ul>
<p>(2) if you didn’t use the imputation argument, replace <code>imputation.method = FALSE</code> to <code>imputation.method = NULL</code> or leave the argument missing.</p>
<ul>
<li><p>simplified sections of codes in <code>assignment_ngs</code> that dealt with <code>strata</code>, <code>pop.levels</code> and <code>pop.labels</code>.</p></li>
<li><p>new function: <code>write_gsi_sim</code>. Write a gsi_sim file from a data frame (wide or long/tidy). Used internally in [assigner] (<a href="https://github.com/thierrygosselin/assigner" class="uri">https://github.com/thierrygosselin/assigner</a>) and might be of interest for users.</p></li>
</ul>
</div>
    <div id="assigner-0-2-6" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-2-6" class="anchor"></a>assigner 0.2.6</h1>
<ul>
<li>Added a <code>NEWS.md</code> file to track changes to the package.</li>
<li>
<code>fst_WC84</code> is now a separate and very fast function that can compute: the overall and pairwise Weir and Cockerham 1984 Theta/Fst. Bootstrap resampling of markers is avalaible to build Confidence Intervals (For Louis Bernatchez and his students;). The estimates are available as a data frame and a matrix with upper diagonal filled with Fst values and lower diagonal filled with the confidence intervals.</li>
</ul>
</div>
    <div id="assigner-0-2-5" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-2-5" class="anchor"></a>assigner 0.2.5</h1>
<ul>
<li>cleaner code for strata section</li>
<li>bug fix restricted to <code>assignment_ngs</code> + <code>assignment.analysis = "adegenet"</code> + <code>sampling.method = "ranked"</code>. A line at the beginning of a gsi_sim code section was deleted makig the assignment with adegenet go through that chunk of code and causing 100% assignment! if (assignment.analysis = “gsi_sim”) {code} prevent this problem…</li>
</ul>
</div>
    <div id="assigner-0-2-4" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-2-4" class="anchor"></a>assigner 0.2.4</h1>
<ul>
<li>bug fixed using adegenet that was introduced in v.0.2.3</li>
<li>introducing a new function <code>import_subsamples_fst</code> to import the fst ranking results from all the subsample runs inside an assignment folder.</li>
</ul>
</div>
    <div id="assigner-0-2-3" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-2-3" class="anchor"></a>assigner 0.2.3</h1>
<ul>
<li>bug fixed in the compilation results section when no pop.id.start and end are used.</li>
</ul>
</div>
    <div id="assigner-0-2-1" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-2-1" class="anchor"></a>assigner 0.2.1</h1>
<ul>
<li>updated the function <code>assignment_mixture</code> with <code>sampling.method = "ranked"</code> and <code>assignment.analysis = "adegenet"</code>.</li>
</ul>
</div>
    <div id="assigner-0-2-0" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-2-0" class="anchor"></a>assigner 0.2.0</h1>
<ul>
<li>new function: <code>assignment_mixture</code> for mixture analysis.</li>
</ul>
</div>
    <div id="assigner-0-1-9" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-1-9" class="anchor"></a>assigner 0.1.9</h1>
<ul>
<li>Simplified gsi_sim install</li>
</ul>
</div>
    <div id="assigner-0-1-8" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-1-8" class="anchor"></a>assigner 0.1.8</h1>
<ul>
<li>You can now opt between [gsi_sim] (<a href="https://github.com/eriqande/gsi_sim" class="uri">https://github.com/eriqande/gsi_sim</a>) or [adegenet] (<a href="https://github.com/thibautjombart/adegenet" class="uri">https://github.com/thibautjombart/adegenet</a>), a R package developed by Thibaul Jombart, to conduct the assignment analysis</li>
</ul>
</div>
    <div id="assigner-0-1-7" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-1-7" class="anchor"></a>assigner 0.1.7</h1>
<ul>
<li>New input file: Re-introduced the haplotype data frame file from stacks.</li>
<li>Argument name change: <code>imputations</code> is now <code>impute.method</code>.</li>
<li>New argument: <code>impute</code> with 2 options: <code>impute = "genotype"</code> or <code>impute = "allele"</code>.</li>
</ul>
</div>
    <div id="assigner-0-1-6" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-1-6" class="anchor"></a>assigner 0.1.6</h1>
<ul>
<li>Input file argument is now <code>data</code> and covers the three types of files the function can use: VCF file, PLINK tped/tfam or data frame of genotypes file.</li>
<li>Huge number of markers (&gt; 50 000 markers) can now be imported in PLINK tped/tfam format. The first 2 columns of the <code>tfam</code> file will be used for the <code>strata</code> argument, unless a new one is provided. Columns 1, 3 and 4 of the <code>tped</code> are discarded. The remaining columns correspond to the genotype in the format <code>01/04</code> where <code>A = 01, C = 02, G = 03 and T = 04</code>. For <code>A/T</code> format, use PLINK or bash to convert. Use [VCFTOOLS] (<a href="http://vcftools.sourceforge.net/" class="uri">http://vcftools.sourceforge.net/</a>) with <code>--plink-tped</code> to convert very large VCF file. For <code>.ped</code> file conversion to <code>.tped</code> use [PLINK] (<a href="http://pngu.mgh.harvard.edu/~purcell/plink/" class="uri">http://pngu.mgh.harvard.edu/~purcell/plink/</a>) with <code>--recode transpose</code>.</li>
</ul>
</div>
    <div id="assigner-0-1-5" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-1-5" class="anchor"></a>assigner 0.1.5</h1>
<ul>
<li>bug fix in <code>method = "random"</code> and <code>imputation</code>
</li>
</ul>
</div>
    <div id="assigner-0-1-4" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-1-4" class="anchor"></a>assigner 0.1.4</h1>
<ul>
<li>Changed function name, from <code>GBS_assignment</code> to <code>assignment_ngs</code>. Stands for assignment with next-generation sequencing data.</li>
<li>New argument <code>df.file</code> if you don’t have a VCF file. See documentation.</li>
<li>New argument <code>strata</code> if you don’t have population id or other metadata info in the individual name. See documentation.</li>
</ul>
</div>
    <div id="assigner-0-1-3" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-1-3" class="anchor"></a>assigner 0.1.3</h1>
<ul>
<li>Changed arguments <code>THL</code> to <code>thl</code> and <code>snp.LD</code> to <code>snp.ld</code> to follow convention.</li>
<li>
<code>iterations.subsample</code> changed to <code>iteration.subsample</code>.</li>
<li>
<code>iterations</code> changed to <code>iteration.method</code> to avoid confusion with other iteration arguments.</li>
<li>Removed <code>baseline</code> and <code>mixture</code> arguments from the function <code>GBS_assignment</code>. These options will be re-introduce later in a separate function.</li>
<li>Using <code>marker.number</code> higher than the number of markers in the data set was causing problems. This could arise when using arguments that removed markers from the dataset (e.g. <code>snp.ld</code>, <code>common.markers</code>, and <code>maf</code> filters).</li>
</ul>
</div>
    <div id="assigner-0-1-2" class="section level1">
<h1 class="page-header">
<a href="#assigner-0-1-2" class="anchor"></a>assigner 0.1.2</h1>
<ul>
<li>new version to update with gsi_sim new install instruction for Linux and Mac. After re-installing assigner package, follow the instruction to re-install the new [gsi_sim] (<a href="https://github.com/eriqande/gsi_sim" class="uri">https://github.com/eriqande/gsi_sim</a>). And delete the old binary ‘gsisim’ in the /usr/local/bin folder with the following Terminal command: <code>sudo rm /usr/local/bin/gsisim</code>
</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#assigner-0-5-5-2019-04-25">0.5.5</a></li>
        <li><a href="#assigner-0-5-4-2019-03-12">0.5.4</a></li>
        <li><a href="#assigner-0-5-3-2019-02-25">0.5.3</a></li>
        <li><a href="#assigner-0-5-2-2018-07-09">0.5.2</a></li>
        <li><a href="#assigner-0-5-1-2018-06-21">0.5.1</a></li>
        <li><a href="#assigner-0-5-0-2017-12-12">0.5.0</a></li>
        <li><a href="#assigner-0-4-9-2017-08-18">0.4.9</a></li>
        <li><a href="#assigner-0-4-8-2017-08-15">0.4.8</a></li>
        <li><a href="#assigner-0-4-7-2017-08-15">0.4.7</a></li>
        <li><a href="#assigner-0-4-6-2017-06-17">0.4.6</a></li>
        <li><a href="#assigner-0-4-5-2017-04-25">0.4.5</a></li>
        <li><a href="#assigner-0-4-4-2017-04-12">0.4.4</a></li>
        <li><a href="#assigner-0-4-3-2017-04-02">0.4.3</a></li>
        <li><a href="#assigner-0-4-2">0.4.2</a></li>
        <li><a href="#assigner-0-4-1">0.4.1</a></li>
        <li><a href="#assigner-0-4-0">0.4.0</a></li>
        <li><a href="#assigner-0-3-9">0.3.9</a></li>
        <li><a href="#assigner-0-3-8">0.3.8</a></li>
        <li><a href="#assigner-0-3-7">0.3.7</a></li>
        <li><a href="#assigner-0-3-6">0.3.6</a></li>
        <li><a href="#assigner-0-3-5">0.3.5</a></li>
        <li><a href="#assigner-0-3-4">0.3.4</a></li>
        <li><a href="#assigner-0-3-3">0.3.3</a></li>
        <li><a href="#assigner-0-3-2">0.3.2</a></li>
        <li><a href="#assigner-0-3-1">0.3.1</a></li>
        <li><a href="#assigner-0-3-0">0.3.0</a></li>
        <li><a href="#assigner-0-2-9">0.2.9</a></li>
        <li><a href="#assigner-0-2-8">0.2.8</a></li>
        <li><a href="#assigner-0-2-7">0.2.7</a></li>
        <li><a href="#assigner-0-2-6">0.2.6</a></li>
        <li><a href="#assigner-0-2-5">0.2.5</a></li>
        <li><a href="#assigner-0-2-4">0.2.4</a></li>
        <li><a href="#assigner-0-2-3">0.2.3</a></li>
        <li><a href="#assigner-0-2-1">0.2.1</a></li>
        <li><a href="#assigner-0-2-0">0.2.0</a></li>
        <li><a href="#assigner-0-1-9">0.1.9</a></li>
        <li><a href="#assigner-0-1-8">0.1.8</a></li>
        <li><a href="#assigner-0-1-7">0.1.7</a></li>
        <li><a href="#assigner-0-1-6">0.1.6</a></li>
        <li><a href="#assigner-0-1-5">0.1.5</a></li>
        <li><a href="#assigner-0-1-4">0.1.4</a></li>
        <li><a href="#assigner-0-1-3">0.1.3</a></li>
        <li><a href="#assigner-0-1-2">0.1.2</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

