<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comparisons of Fst computations â€¢ assigner</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Comparisons of Fst computations">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">assigner</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">Installation</a>
</li>
<li>
  <a href="../../FEATURES.html">Features</a>
</li>
<li>
  <a href="../../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/get_started.html">Get started</a>
    </li>
    <li>
      <a href="../../articles/fst_confidence_intervals.html">Fst analysis</a>
    </li>
    <li>
      <a href="../../articles/web_only/fst_comparisons.html">Comparisons of Fst in 8 packages/software</a>
    </li>
    <li>
      <a href="../../articles/assignment.whitelists.html">Using whitelists in assignment</a>
    </li>
    <li>
      <a href="../../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/thierrygosselin/assigner">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Comparisons of Fst computations</h1>
            <h3 class="subtitle">
<code>assigner</code>, <code>diveRsity</code>, <code>GenoDive</code>, <code>hierfstat</code>, <code>SNPRelate</code>, <code>stAMPP</code>, <code>strataG</code>, <code>VCFtools.</code>
</h3>
                        <h4 class="author">Thierry Gosselin</h4>
            
            <h4 class="date">2019-05-17</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/assigner/blob/master/vignettes/web_only/fst_comparisons.Rmd"><code>vignettes/web_only/fst_comparisons.Rmd</code></a></small>
      <div class="hidden name"><code>fst_comparisons.Rmd</code></div>

    </div>

    
    
<div id="prepare-your-r-session" class="section level1">
<h1 class="hasAnchor">
<a href="#prepare-your-r-session" class="anchor"></a>Prepare your R session</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rm">rm</a></span>(<span class="dt">list =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ls">ls</a></span>()) <span class="co"># clean workspace</span></a>
<a class="sourceLine" id="cb1-2" title="2"><span class="co"># install required packages</span></a>
<a class="sourceLine" id="cb1-3" title="3"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"diveRsity"</span>)) <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"diveRsity"</span>)</a>
<a class="sourceLine" id="cb1-4" title="4"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"hierfstat"</span>)) <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"hierfstat"</span>)</a>
<a class="sourceLine" id="cb1-5" title="5"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"StAMPP"</span>)) <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"StAMPP"</span>)</a>
<a class="sourceLine" id="cb1-6" title="6"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"strataG"</span>)) <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"strataG"</span>)</a>
<a class="sourceLine" id="cb1-7" title="7"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"SNPRelate"</span>)) {</a>
<a class="sourceLine" id="cb1-8" title="8">  <span class="cf">if</span>(<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"BiocManager"</span>)) <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"BiocManager"</span>)</a>
<a class="sourceLine" id="cb1-9" title="9">  BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"SNPRelate"</span>)</a>
<a class="sourceLine" id="cb1-10" title="10">}</a>
<a class="sourceLine" id="cb1-11" title="11"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"ggcorrplot"</span>)) <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"ggcorrplot"</span>)</a></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"remotes"</span>)) <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"remotes"</span>)</a></code></pre></div>
<pre><code>## Loading required package: remotes</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"assigner"</span>)) remotes<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/remotes/topics/install_github">install_github</a></span>(<span class="st">"thierrygosselin/assigner"</span>)</a></code></pre></div>
<pre><code>## Loading required package: assigner</code></pre>
<pre><code>## Registered S3 methods overwritten by 'ggplot2':
##   method         from 
##   [.quosures     rlang
##   c.quosures     rlang
##   print.quosures rlang</code></pre>
<pre><code>## Registered S3 method overwritten by 'spdep':
##   method   from
##   plot.mst ape</code></pre>
<p>For macOS: <a href="http://www.bentleydrummer.nl/software/software/GenoDive.html">GenoDive</a>.</p>
<p>VCFtools: <a href="https://vcftools.github.io">install instructions</a> or <a href="https://gbs-cloud-tutorial.readthedocs.io/en/latest/07_start_from_new_image.html#install-gbs-radseq-software-time-required-30min">here</a></p>
<p><strong>Versions of packages and software:</strong></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co"># Check packages versions</span></a>
<a class="sourceLine" id="cb8-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/packageDescription">packageVersion</a></span>(<span class="st">'assigner'</span>) <span class="co"># 0.5.6.</span></a>
<a class="sourceLine" id="cb8-3" title="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/packageDescription">packageVersion</a></span>(<span class="st">'diveRsity'</span>) <span class="co"># 1.9.90</span></a>
<a class="sourceLine" id="cb8-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/packageDescription">packageVersion</a></span>(<span class="st">'hierfstat'</span>) <span class="co"># 0.4.22</span></a>
<a class="sourceLine" id="cb8-5" title="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/packageDescription">packageVersion</a></span>(<span class="st">'stAMPP'</span>) <span class="co"># 1.5.1</span></a>
<a class="sourceLine" id="cb8-6" title="6"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/packageDescription">packageVersion</a></span>(<span class="st">'strataG'</span>) <span class="co"># 2.0.2</span></a>
<a class="sourceLine" id="cb8-7" title="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/packageDescription">packageVersion</a></span>(<span class="st">'SNPRelate'</span>) <span class="co"># 1.18.0</span></a>
<a class="sourceLine" id="cb8-8" title="8"><span class="co">#GenoDive: 2.27</span></a>
<a class="sourceLine" id="cb8-9" title="9"><span class="co"># VCFtools.: 0.1.17</span></a></code></pre></div>
<p>Put this <strong>Rmd notebook</strong> in the same directory as the data or use <code><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd("/YOUR/NEW/PATH/HERE/")</a></code>, to change the working directory.</p>
</div>
<div id="goal-of-this-exercise" class="section level1">
<h1 class="hasAnchor">
<a href="#goal-of-this-exercise" class="anchor"></a>Goal of this exercise</h1>
<p>Letâ€™s see how <code>assigner's</code> Fst calculations (using Weir and Cockerham, 1984) compares with alternatives. To make differences and bottlenecks really stand out we will use a RADseq dataset with <strong>thousands of markers</strong>.</p>
</div>
<div id="datasets" class="section level1">
<h1 class="hasAnchor">
<a href="#datasets" class="anchor"></a>Datasets</h1>
<div id="empirical-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#empirical-dataset" class="anchor"></a>Empirical dataset</h2>
<p>Weâ€™re going to use the unfiltered vcf sunflower dataset from Owens et al.Â 2016, kindly provided by Gregory Owens. This dataset as missing data.</p>
<p>Owens GL, Baute GJ, Rieseberg LH. Revisiting a classic case of introgression: hybridization and gene flow in Californian sunflowers. Abbott RJ, Barton NH, Good JM, editors. Molecular Ecology. 2016;25: 2630â€“2643. <a href="doi:10.1111/mec.13569" class="uri">doi:10.1111/mec.13569</a></p>
<p><strong>Download the files:</strong></p>
<ul>
<li><a href="https://www.dropbox.com/s/3kfx351ijeaa1wr/sunflower.vcf?dl=0">vcf</a></li>
<li>
<a href="https://www.dropbox.com/s/406jleg3imwfaup/sunflower.strata.tsv?dl=0">strata</a>. Whatâ€™s a strata file ? Read the <a href="https://thierrygosselin.github.io/radiator/reference/read_strata.html">doc</a>.</li>
</ul>
</div>
<div id="simulated-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#simulated-datasets" class="anchor"></a>Simulated datasets</h2>
<p>After running this vignette once, I saw something strange in the results. To explore this further, this second grind of testing add 2 additional datasets: <a href="http://thierrygosselin.github.io/assigner/reference/data_assigner_sim_01.html">data_assigner_sim_01</a> and <a href="http://thierrygosselin.github.io/assigner/reference/data_assigner_sim_02.html">data_assigner_sim_02</a>. The documentation explains how these simulated data were produced. They differ mainly in migration rate which resulted in contrastingly different Fst, admixture/membership probability.</p>
</div>
</div>
<div id="generate-the-output-files" class="section level1">
<h1 class="hasAnchor">
<a href="#generate-the-output-files" class="anchor"></a>Generate the output files</h1>
<p>We use <a href="https://thierrygosselin.github.io/radiator/reference/genomic_converter.html">radiator::genomic_converter</a> to produce the required files or R objects:</p>
<div id="empirical-dataset-1" class="section level2">
<h2 class="hasAnchor">
<a href="#empirical-dataset-1" class="anchor"></a>Empirical dataset</h2>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="co"># radiator is installed automatically with assigner</span></a>
<a class="sourceLine" id="cb9-2" title="2">sunflower.filtered &lt;-<span class="st"> </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/genomic_converter">genomic_converter</a></span>(</a>
<a class="sourceLine" id="cb9-3" title="3">  <span class="dt">data =</span> <span class="st">"sunflower.vcf"</span>, </a>
<a class="sourceLine" id="cb9-4" title="4">  <span class="dt">strata =</span> <span class="st">"sunflower.strata.tsv"</span>, </a>
<a class="sourceLine" id="cb9-5" title="5">  <span class="dt">output =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(</a>
<a class="sourceLine" id="cb9-6" title="6">    <span class="st">"genepop"</span>, <span class="co"># for GenoDive and diveRsity</span></a>
<a class="sourceLine" id="cb9-7" title="7">    <span class="st">"gtypes"</span>, <span class="co"># for strataG</span></a>
<a class="sourceLine" id="cb9-8" title="8">    <span class="st">"genlight"</span>, <span class="co"># for StAMPP</span></a>
<a class="sourceLine" id="cb9-9" title="9">    <span class="st">"hierfstat"</span>, <span class="co"># for Hierfstat</span></a>
<a class="sourceLine" id="cb9-10" title="10">    <span class="st">"snprelate"</span> <span class="co"># for SNPRelate</span></a>
<a class="sourceLine" id="cb9-11" title="11">  )</a>
<a class="sourceLine" id="cb9-12" title="12">) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">glimpse</a></span>()</a></code></pre></div>
<p>VCF stats:</p>
<ul>
<li>the original number of markers is 34258.</li>
<li>by default, <code><a href="https://www.rdocumentation.org/packages/radiator/topics/genomic_converter">radiator::genomic_converter</a></code> will only keep polymorphic markers in common between the strata.</li>
<li>15973 biallelic markers are recovered.</li>
<li>Number of populations: 9</li>
<li>Number of individuals: 88</li>
</ul>
<p>The <code>sunflower.filtered</code> object is a list with several objects inside:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(sunflower.filtered)</a>
<a class="sourceLine" id="cb10-2" title="2"><span class="co">#&gt; [1] "hierfstat" "gtypes"    "genlight"  "snprelate" "tidy.data"</span></a>
<a class="sourceLine" id="cb10-3" title="3"><span class="co"># the tidy dataset is used by assigner</span></a></code></pre></div>
<p>The function also produced a folder with several files, including:</p>
<ul>
<li>the genepop file</li>
<li>the SNPRelate GDS file</li>
<li>the hierfstat file</li>
<li>the gtypes file</li>
<li>the genlight file</li>
<li>the tidy data file</li>
</ul>
<p>To learn more about the output folders and files: <a href="https://thierrygosselin.github.io/radiator/reference/genomic_converter.html">radiator::genomic_converter</a></p>
<p><em>Note on timing</em></p>
<p>I could use fancy timing packages (<a href="https://github.com/joshuaulrich/microbenchmark">microbenchmark</a>, <a href="http://bench.r-lib.org">bench</a>), but for simplicity weâ€™re just going to use <code>system.time</code>.</p>
</div>
<div id="simulated-datasets-1" class="section level2">
<h2 class="hasAnchor">
<a href="#simulated-datasets-1" class="anchor"></a>Simulated datasets</h2>
<div id="sim_01" class="section level3">
<h3 class="hasAnchor">
<a href="#sim_01" class="anchor"></a>sim_01</h3>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">data_assigner_sim_<span class="dv">01</span> <span class="op">%&lt;&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb11-2" title="2"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">POP_ID =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(<span class="dt">x =</span> POP_ID, <span class="dt">levels =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP_1"</span>, <span class="st">"POP_2"</span>, <span class="st">"POP_3"</span>, <span class="st">"POP_4"</span>, <span class="st">"POP_5"</span>))) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb11-3" title="3"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(POP_ID)</a>
<a class="sourceLine" id="cb11-4" title="4"></a>
<a class="sourceLine" id="cb11-5" title="5">sim.data1 &lt;-<span class="st"> </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/genomic_converter">genomic_converter</a></span>(</a>
<a class="sourceLine" id="cb11-6" title="6">  <span class="dt">data =</span> data_assigner_sim_<span class="dv">01</span>, </a>
<a class="sourceLine" id="cb11-7" title="7">  <span class="dt">output =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(</a>
<a class="sourceLine" id="cb11-8" title="8">    <span class="st">"genepop"</span>, <span class="co"># for GenoDive and diveRsity</span></a>
<a class="sourceLine" id="cb11-9" title="9">    <span class="st">"gtypes"</span>, <span class="co"># for strataG</span></a>
<a class="sourceLine" id="cb11-10" title="10">    <span class="st">"genlight"</span>, <span class="co"># for StAMPP</span></a>
<a class="sourceLine" id="cb11-11" title="11">    <span class="st">"hierfstat"</span>, <span class="co"># for Hierfstat</span></a>
<a class="sourceLine" id="cb11-12" title="12">    <span class="st">"snprelate"</span> <span class="co"># for SNPRelate</span></a>
<a class="sourceLine" id="cb11-13" title="13">  ),</a>
<a class="sourceLine" id="cb11-14" title="14">  <span class="dt">filename =</span> <span class="st">"sim01"</span>,</a>
<a class="sourceLine" id="cb11-15" title="15">  <span class="dt">path.folder =</span> <span class="st">"fst_sim01"</span>,</a>
<a class="sourceLine" id="cb11-16" title="16">  <span class="dt">internal =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb11-17" title="17">  <span class="dt">verbose =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb11-18" title="18">)</a>
<a class="sourceLine" id="cb11-19" title="19">sim.data1<span class="op">$</span>tidy.data <span class="op">%&lt;&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(POP_ID)</a></code></pre></div>
</div>
<div id="sim_02" class="section level3">
<h3 class="hasAnchor">
<a href="#sim_02" class="anchor"></a>sim_02</h3>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">data_assigner_sim_<span class="dv">02</span> <span class="op">%&lt;&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb12-2" title="2"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">POP_ID =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(<span class="dt">x =</span> POP_ID, <span class="dt">levels =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP_1"</span>, <span class="st">"POP_2"</span>, <span class="st">"POP_3"</span>, <span class="st">"POP_4"</span>, <span class="st">"POP_5"</span>))) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb12-3" title="3"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(POP_ID)</a>
<a class="sourceLine" id="cb12-4" title="4"></a>
<a class="sourceLine" id="cb12-5" title="5">sim.data2 &lt;-<span class="st"> </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/genomic_converter">genomic_converter</a></span>(</a>
<a class="sourceLine" id="cb12-6" title="6">  <span class="dt">data =</span> data_assigner_sim_<span class="dv">02</span>, </a>
<a class="sourceLine" id="cb12-7" title="7">  <span class="dt">output =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(</a>
<a class="sourceLine" id="cb12-8" title="8">    <span class="st">"genepop"</span>, <span class="co"># for GenoDive and diveRsity</span></a>
<a class="sourceLine" id="cb12-9" title="9">    <span class="st">"gtypes"</span>, <span class="co"># for strataG</span></a>
<a class="sourceLine" id="cb12-10" title="10">    <span class="st">"genlight"</span>, <span class="co"># for StAMPP</span></a>
<a class="sourceLine" id="cb12-11" title="11">    <span class="st">"hierfstat"</span>, <span class="co"># for Hierfstat</span></a>
<a class="sourceLine" id="cb12-12" title="12">    <span class="st">"snprelate"</span> <span class="co"># for SNPRelate</span></a>
<a class="sourceLine" id="cb12-13" title="13">  ),</a>
<a class="sourceLine" id="cb12-14" title="14">  <span class="dt">filename =</span> <span class="st">"sim02"</span>,</a>
<a class="sourceLine" id="cb12-15" title="15">  <span class="dt">path.folder =</span> <span class="st">"fst_sim02"</span>,</a>
<a class="sourceLine" id="cb12-16" title="16">  <span class="dt">internal =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb12-17" title="17">  <span class="dt">verbose =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb12-18" title="18">)</a>
<a class="sourceLine" id="cb12-19" title="19">sim.data2<span class="op">$</span>tidy.data <span class="op">%&lt;&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(POP_ID)</a></code></pre></div>
</div>
</div>
</div>
<div id="hierfstat" class="section level1">
<h1 class="hasAnchor">
<a href="#hierfstat" class="anchor"></a>hierfstat</h1>
<p>Letâ€™s start with the Gold Standard, JÃ©rÃ´me Goudetâ€™s <a href="https://github.com/jgx65/hierfstat">hierfstat</a> package.</p>
<div id="empirical-dataset-2" class="section level2">
<h2 class="hasAnchor">
<a href="#empirical-dataset-2" class="anchor"></a>Empirical dataset</h2>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"><span class="co"># ETA ~ 30 min on my MBP, so yes, you have time for coffee...</span></a>
<a class="sourceLine" id="cb13-2" title="2">pops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(sunflower.filtered<span class="op">$</span>tidy.data<span class="op">$</span>POP_ID))<span class="co"># pop string</span></a>
<a class="sourceLine" id="cb13-3" title="3"></a>
<a class="sourceLine" id="cb13-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.time">system.time</a></span>(</a>
<a class="sourceLine" id="cb13-5" title="5">  hierfstat.fst &lt;-<span class="st"> </span>hierfstat<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/hierfstat/topics/pairwise.WCfst">pairwise.WCfst</a></span>(</a>
<a class="sourceLine" id="cb13-6" title="6">    <span class="dt">dat =</span> sunflower.filtered<span class="op">$</span>hierfstat, </a>
<a class="sourceLine" id="cb13-7" title="7">    <span class="dt">diploid =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb13-8" title="8">  ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-9" title="9"><span class="st">    </span><span class="co"># The result need some work</span></a>
<a class="sourceLine" id="cb13-10" title="10"><span class="st">    </span><span class="co"># We transform the matrix into a useful tibble:</span></a>
<a class="sourceLine" id="cb13-11" title="11"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-12" title="12"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_rownames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-13" title="13"><span class="st">    </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb13-14" title="14">      <span class="dt">x =</span> ., <span class="dt">remove.diag =</span> <span class="ot">TRUE</span>, <span class="dt">remove.lower =</span> <span class="ot">TRUE</span>, <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, <span class="dt">relative =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb13-15" title="15">    ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-16" title="16"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>, <span class="st">"HIERFSTAT"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-17" title="17"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(POP1, POP2)</a>
<a class="sourceLine" id="cb13-18" title="18">)</a>
<a class="sourceLine" id="cb13-19" title="19"><span class="co">#&gt; 1233.880sec ~20 min</span></a>
<a class="sourceLine" id="cb13-20" title="20">hierfstat.fst</a></code></pre></div>
</div>
<div id="simulated-datasets-2" class="section level2">
<h2 class="hasAnchor">
<a href="#simulated-datasets-2" class="anchor"></a>Simulated datasets</h2>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1">pops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(sim.data1<span class="op">$</span>tidy.data<span class="op">$</span>POP_ID))<span class="co"># pop string</span></a>
<a class="sourceLine" id="cb14-2" title="2">hierfstat.sim01.fst &lt;-<span class="st"> </span>hierfstat<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/hierfstat/topics/pairwise.WCfst">pairwise.WCfst</a></span>(</a>
<a class="sourceLine" id="cb14-3" title="3">    <span class="dt">dat =</span> sim.data1<span class="op">$</span>hierfstat, </a>
<a class="sourceLine" id="cb14-4" title="4">    <span class="dt">diploid =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb14-5" title="5">  ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-6" title="6"><span class="st">    </span><span class="co"># The result need some work</span></a>
<a class="sourceLine" id="cb14-7" title="7"><span class="st">    </span><span class="co"># We transform the matrix into a useful tibble:</span></a>
<a class="sourceLine" id="cb14-8" title="8"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-9" title="9"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_rownames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-10" title="10"><span class="st">    </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb14-11" title="11">      <span class="dt">x =</span> ., <span class="dt">remove.diag =</span> <span class="ot">TRUE</span>, <span class="dt">remove.lower =</span> <span class="ot">TRUE</span>, <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, <span class="dt">relative =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb14-12" title="12">    ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-13" title="13"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>, <span class="st">"HIERFSTAT_SIM01"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-14" title="14"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(POP1, POP2)</a>
<a class="sourceLine" id="cb14-15" title="15">pops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(sim.data2<span class="op">$</span>tidy.data<span class="op">$</span>POP_ID))<span class="co"># pop string</span></a>
<a class="sourceLine" id="cb14-16" title="16">  hierfstat.sim02.fst &lt;-<span class="st"> </span>hierfstat<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/hierfstat/topics/pairwise.WCfst">pairwise.WCfst</a></span>(</a>
<a class="sourceLine" id="cb14-17" title="17">    <span class="dt">dat =</span> sim.data2<span class="op">$</span>hierfstat, </a>
<a class="sourceLine" id="cb14-18" title="18">    <span class="dt">diploid =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb14-19" title="19">  ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-20" title="20"><span class="st">    </span><span class="co"># The result need some work</span></a>
<a class="sourceLine" id="cb14-21" title="21"><span class="st">    </span><span class="co"># We transform the matrix into a useful tibble:</span></a>
<a class="sourceLine" id="cb14-22" title="22"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-23" title="23"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_rownames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-24" title="24"><span class="st">    </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb14-25" title="25">      <span class="dt">x =</span> ., <span class="dt">remove.diag =</span> <span class="ot">TRUE</span>, <span class="dt">remove.lower =</span> <span class="ot">TRUE</span>, <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, <span class="dt">relative =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb14-26" title="26">    ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-27" title="27"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>, <span class="st">"HIERFSTAT_SIM02"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-28" title="28"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(POP1, POP2)</a></code></pre></div>
</div>
</div>
<div id="stratag" class="section level1">
<h1 class="hasAnchor">
<a href="#stratag" class="anchor"></a>strataG</h1>
<p>This is Eric Archerâ€™s <a href="https://github.com/EricArcher/strataG">strataG</a> package.</p>
<div id="empirical-dataset-3" class="section level2">
<h2 class="hasAnchor">
<a href="#empirical-dataset-3" class="anchor"></a>Empirical dataset</h2>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.time">system.time</a></span>(</a>
<a class="sourceLine" id="cb15-2" title="2">  stratag.fst &lt;-<span class="st"> </span>strataG<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/strataG/topics/popStructTest">popStructTest</a></span>(</a>
<a class="sourceLine" id="cb15-3" title="3">    <span class="dt">g =</span> sunflower.filtered<span class="op">$</span>gtypes, </a>
<a class="sourceLine" id="cb15-4" title="4">    <span class="dt">stats =</span> <span class="st">"fst"</span>, </a>
<a class="sourceLine" id="cb15-5" title="5">    <span class="dt">type =</span> <span class="st">"both"</span>, </a>
<a class="sourceLine" id="cb15-6" title="6">    <span class="dt">quietly =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb15-7" title="7">    <span class="dt">max.cores =</span> parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/detectCores">detectCores</a></span>() <span class="op">-</span><span class="st"> </span><span class="dv">1</span>, </a>
<a class="sourceLine" id="cb15-8" title="8">    <span class="dt">nrep =</span> <span class="dv">0</span>, </a>
<a class="sourceLine" id="cb15-9" title="9">    <span class="dt">keep.null =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb15-10" title="10">    <span class="dt">write.output =</span> <span class="ot">FALSE</span>) <span class="op">%$%</span></a>
<a class="sourceLine" id="cb15-11" title="11"><span class="st">    </span>pairwise<span class="op">$</span>result <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb15-12" title="12"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="dt">POP1 =</span> strata<span class="fl">.1</span>, <span class="dt">POP2 =</span> strata<span class="fl">.2</span>, <span class="dt">STRATAG =</span> Fst)</a>
<a class="sourceLine" id="cb15-13" title="13">)</a>
<a class="sourceLine" id="cb15-14" title="14">stratag.fst</a>
<a class="sourceLine" id="cb15-15" title="15"><span class="co">#&gt;657.377sec ~ 11 min</span></a></code></pre></div>
</div>
<div id="simulated-datasets-3" class="section level2">
<h2 class="hasAnchor">
<a href="#simulated-datasets-3" class="anchor"></a>Simulated datasets</h2>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1">stratag.sim01.fst &lt;-<span class="st"> </span>strataG<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/strataG/topics/popStructTest">popStructTest</a></span>(</a>
<a class="sourceLine" id="cb16-2" title="2">    <span class="dt">g =</span> sim.data1<span class="op">$</span>gtypes, </a>
<a class="sourceLine" id="cb16-3" title="3">    <span class="dt">stats =</span> <span class="st">"fst"</span>, </a>
<a class="sourceLine" id="cb16-4" title="4">    <span class="dt">type =</span> <span class="st">"both"</span>, </a>
<a class="sourceLine" id="cb16-5" title="5">    <span class="dt">quietly =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb16-6" title="6">    <span class="dt">max.cores =</span> parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/detectCores">detectCores</a></span>() <span class="op">-</span><span class="st"> </span><span class="dv">1</span>, </a>
<a class="sourceLine" id="cb16-7" title="7">    <span class="dt">nrep =</span> <span class="dv">0</span>, </a>
<a class="sourceLine" id="cb16-8" title="8">    <span class="dt">keep.null =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb16-9" title="9">    <span class="dt">write.output =</span> <span class="ot">FALSE</span>) <span class="op">%$%</span></a>
<a class="sourceLine" id="cb16-10" title="10"><span class="st">    </span>pairwise<span class="op">$</span>result <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb16-11" title="11"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="dt">POP1 =</span> strata<span class="fl">.1</span>, <span class="dt">POP2 =</span> strata<span class="fl">.2</span>, <span class="dt">STRATAG_SIM01 =</span> Fst)</a>
<a class="sourceLine" id="cb16-12" title="12">stratag.sim02.fst &lt;-<span class="st"> </span>strataG<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/strataG/topics/popStructTest">popStructTest</a></span>(</a>
<a class="sourceLine" id="cb16-13" title="13">    <span class="dt">g =</span> sim.data2<span class="op">$</span>gtypes, </a>
<a class="sourceLine" id="cb16-14" title="14">    <span class="dt">stats =</span> <span class="st">"fst"</span>, </a>
<a class="sourceLine" id="cb16-15" title="15">    <span class="dt">type =</span> <span class="st">"both"</span>, </a>
<a class="sourceLine" id="cb16-16" title="16">    <span class="dt">quietly =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb16-17" title="17">    <span class="dt">max.cores =</span> parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/detectCores">detectCores</a></span>() <span class="op">-</span><span class="st"> </span><span class="dv">1</span>, </a>
<a class="sourceLine" id="cb16-18" title="18">    <span class="dt">nrep =</span> <span class="dv">0</span>, </a>
<a class="sourceLine" id="cb16-19" title="19">    <span class="dt">keep.null =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb16-20" title="20">    <span class="dt">write.output =</span> <span class="ot">FALSE</span>) <span class="op">%$%</span></a>
<a class="sourceLine" id="cb16-21" title="21"><span class="st">    </span>pairwise<span class="op">$</span>result <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb16-22" title="22"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="dt">POP1 =</span> strata<span class="fl">.1</span>, <span class="dt">POP2 =</span> strata<span class="fl">.2</span>, <span class="dt">STRATAG_SIM02 =</span> Fst)</a></code></pre></div>
</div>
</div>
<div id="diversity" class="section level1">
<h1 class="hasAnchor">
<a href="#diversity" class="anchor"></a>diveRsity</h1>
<ul>
<li>Kevin Keenanâ€™s <a href="http://diversityinlife.weebly.com/software.html">diveRsity</a> package requires the <code>genepop</code> file.</li>
<li>the function is optimized with C++, so no need for a coffee (&lt;Â 1 min on my MBP)</li>
</ul>
<div id="empirical-dataset-4" class="section level2">
<h2 class="hasAnchor">
<a href="#empirical-dataset-4" class="anchor"></a>Empirical dataset</h2>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1">pops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(sunflower.filtered<span class="op">$</span>tidy.data<span class="op">$</span>POP_ID)) <span class="co"># pop string</span></a>
<a class="sourceLine" id="cb17-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.time">system.time</a></span>(</a>
<a class="sourceLine" id="cb17-3" title="3">  diversity.fst &lt;-<span class="st"> </span>diveRsity<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/diveRsity/topics/diffCalc">diffCalc</a></span>(</a>
<a class="sourceLine" id="cb17-4" title="4">    <span class="dt">infile =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(</a>
<a class="sourceLine" id="cb17-5" title="5">      <span class="dt">path =</span> <span class="st">"."</span>, </a>
<a class="sourceLine" id="cb17-6" title="6">      <span class="dt">pattern =</span> <span class="st">"_genepop.gen"</span>, </a>
<a class="sourceLine" id="cb17-7" title="7">      <span class="dt">full.names =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb17-8" title="8">      <span class="dt">recursive =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb17-9" title="9">    ), </a>
<a class="sourceLine" id="cb17-10" title="10">    <span class="dt">outfile =</span> <span class="st">"sunflower_diversity_output"</span>, </a>
<a class="sourceLine" id="cb17-11" title="11">    <span class="dt">fst =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb17-12" title="12">    <span class="dt">pairwise =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb17-13" title="13">    <span class="dt">para =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb17-14" title="14">  ) <span class="op">%$%</span></a>
<a class="sourceLine" id="cb17-15" title="15"><span class="st">    </span>pairwise <span class="op">%$%</span></a>
<a class="sourceLine" id="cb17-16" title="16"><span class="st">    </span>Fst <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb17-17" title="17"><span class="st">    </span><span class="co"># The result need some work</span></a>
<a class="sourceLine" id="cb17-18" title="18"><span class="st">    </span><span class="co"># We transform the matrix into a useful tibble:</span></a>
<a class="sourceLine" id="cb17-19" title="19"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb17-20" title="20"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_rownames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb17-21" title="21"><span class="st">    </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb17-22" title="22">      <span class="dt">x =</span> ., </a>
<a class="sourceLine" id="cb17-23" title="23">      <span class="dt">remove.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb17-24" title="24">      <span class="dt">remove.lower =</span> <span class="ot">FALSE</span>, <span class="co"># FALSE because their's only the lower diag</span></a>
<a class="sourceLine" id="cb17-25" title="25">      <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb17-26" title="26">      <span class="dt">relative =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb17-27" title="27">    ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb17-28" title="28"><span class="st">    </span><span class="co"># we switch the POP id column here to match the others</span></a>
<a class="sourceLine" id="cb17-29" title="29"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP2"</span>, <span class="st">"POP1"</span>, <span class="st">"DIVERSITY"</span>))</a>
<a class="sourceLine" id="cb17-30" title="30">)</a>
<a class="sourceLine" id="cb17-31" title="31"><span class="co">#&gt; 42.711 sec Wow! this is very fast, plus other distances are automatically generated!</span></a>
<a class="sourceLine" id="cb17-32" title="32">diversity.fst</a></code></pre></div>
</div>
<div id="simulated-datasets-4" class="section level2">
<h2 class="hasAnchor">
<a href="#simulated-datasets-4" class="anchor"></a>Simulated datasets</h2>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1">pops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(sim.data1<span class="op">$</span>tidy.data<span class="op">$</span>POP_ID)) <span class="co"># pop string</span></a>
<a class="sourceLine" id="cb18-2" title="2">diversity.sim01.fst &lt;-<span class="st"> </span>diveRsity<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/diveRsity/topics/diffCalc">diffCalc</a></span>(</a>
<a class="sourceLine" id="cb18-3" title="3">  <span class="dt">infile =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(</a>
<a class="sourceLine" id="cb18-4" title="4">    <span class="dt">path =</span> <span class="st">"."</span>, </a>
<a class="sourceLine" id="cb18-5" title="5">    <span class="dt">pattern =</span> <span class="st">"sim01_genepop.gen"</span>, </a>
<a class="sourceLine" id="cb18-6" title="6">    <span class="dt">full.names =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb18-7" title="7">    <span class="dt">recursive =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb18-8" title="8">  ), </a>
<a class="sourceLine" id="cb18-9" title="9">  <span class="dt">outfile =</span> <span class="st">"sim_data1_diversity_output"</span>, </a>
<a class="sourceLine" id="cb18-10" title="10">  <span class="dt">fst =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb18-11" title="11">  <span class="dt">pairwise =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb18-12" title="12">  <span class="dt">para =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb18-13" title="13">) <span class="op">%$%</span></a>
<a class="sourceLine" id="cb18-14" title="14"><span class="st">  </span>pairwise <span class="op">%$%</span></a>
<a class="sourceLine" id="cb18-15" title="15"><span class="st">  </span>Fst <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-16" title="16"><span class="st">  </span><span class="co"># The result need some work</span></a>
<a class="sourceLine" id="cb18-17" title="17"><span class="st">  </span><span class="co"># We transform the matrix into a useful tibble:</span></a>
<a class="sourceLine" id="cb18-18" title="18"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-19" title="19"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_rownames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-20" title="20"><span class="st">  </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb18-21" title="21">    <span class="dt">x =</span> ., </a>
<a class="sourceLine" id="cb18-22" title="22">    <span class="dt">remove.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb18-23" title="23">    <span class="dt">remove.lower =</span> <span class="ot">FALSE</span>, <span class="co"># FALSE because their's only the lower diag</span></a>
<a class="sourceLine" id="cb18-24" title="24">    <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb18-25" title="25">    <span class="dt">relative =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb18-26" title="26">  ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-27" title="27"><span class="st">  </span><span class="co"># we switch the POP id column here to match the others</span></a>
<a class="sourceLine" id="cb18-28" title="28"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP2"</span>, <span class="st">"POP1"</span>, <span class="st">"DIVERSITY_SIM01"</span>))</a>
<a class="sourceLine" id="cb18-29" title="29">pops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(sim.data2<span class="op">$</span>tidy.data<span class="op">$</span>POP_ID)) <span class="co"># pop string</span></a>
<a class="sourceLine" id="cb18-30" title="30">diversity.sim02.fst &lt;-<span class="st"> </span>diveRsity<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/diveRsity/topics/diffCalc">diffCalc</a></span>(</a>
<a class="sourceLine" id="cb18-31" title="31">  <span class="dt">infile =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(</a>
<a class="sourceLine" id="cb18-32" title="32">    <span class="dt">path =</span> <span class="st">"."</span>, </a>
<a class="sourceLine" id="cb18-33" title="33">    <span class="dt">pattern =</span> <span class="st">"sim02_genepop.gen"</span>, </a>
<a class="sourceLine" id="cb18-34" title="34">    <span class="dt">full.names =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb18-35" title="35">    <span class="dt">recursive =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb18-36" title="36">  ),  </a>
<a class="sourceLine" id="cb18-37" title="37">  <span class="dt">outfile =</span> <span class="st">"sim_data2_diversity_output"</span>, </a>
<a class="sourceLine" id="cb18-38" title="38">  <span class="dt">fst =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb18-39" title="39">  <span class="dt">pairwise =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb18-40" title="40">  <span class="dt">para =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb18-41" title="41">) <span class="op">%$%</span></a>
<a class="sourceLine" id="cb18-42" title="42"><span class="st">  </span>pairwise <span class="op">%$%</span></a>
<a class="sourceLine" id="cb18-43" title="43"><span class="st">  </span>Fst <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-44" title="44"><span class="st">  </span><span class="co"># The result need some work</span></a>
<a class="sourceLine" id="cb18-45" title="45"><span class="st">  </span><span class="co"># We transform the matrix into a useful tibble:</span></a>
<a class="sourceLine" id="cb18-46" title="46"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-47" title="47"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_rownames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-48" title="48"><span class="st">  </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb18-49" title="49">    <span class="dt">x =</span> ., </a>
<a class="sourceLine" id="cb18-50" title="50">    <span class="dt">remove.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb18-51" title="51">    <span class="dt">remove.lower =</span> <span class="ot">FALSE</span>, <span class="co"># FALSE because their's only the lower diag</span></a>
<a class="sourceLine" id="cb18-52" title="52">    <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb18-53" title="53">    <span class="dt">relative =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb18-54" title="54">  ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-55" title="55"><span class="st">  </span><span class="co"># we switch the POP id column here to match the others</span></a>
<a class="sourceLine" id="cb18-56" title="56"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP2"</span>, <span class="st">"POP1"</span>, <span class="st">"DIVERSITY_SIM02"</span>))</a></code></pre></div>
</div>
</div>
<div id="snprelate" class="section level1">
<h1 class="hasAnchor">
<a href="#snprelate" class="anchor"></a>SNPRelate</h1>
<ul>
<li>Xiuwen Zhengâ€™s <a href="https://github.com/zhengxwen/SNPRelate">SNPRelate</a> package is very fast.</li>
<li>No out-of-the-box pairwise Fst analysis. So I implemented my own below.</li>
</ul>
<div id="empirical-dataset-5" class="section level2">
<h2 class="hasAnchor">
<a href="#empirical-dataset-5" class="anchor"></a>Empirical dataset</h2>
<p>We need the strata</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1">strata &lt;-<span class="st"> </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/read_strata">read_strata</a></span>(<span class="dt">strata =</span> <span class="st">"sunflower.strata.tsv"</span>) <span class="op">%$%</span><span class="st"> </span>strata</a></code></pre></div>
<p>Showing off the global Fst only:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.time">system.time</a></span>(</a>
<a class="sourceLine" id="cb20-2" title="2">  snprelate.fst &lt;-<span class="st"> </span>SNPRelate<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/SNPRelate/topics/snpgdsFst">snpgdsFst</a></span>(</a>
<a class="sourceLine" id="cb20-3" title="3">    <span class="dt">gdsobj =</span> sunflower.filtered<span class="op">$</span>snprelate,</a>
<a class="sourceLine" id="cb20-4" title="4">    <span class="dt">population =</span> strata<span class="op">$</span>STRATA, <span class="co"># factors required</span></a>
<a class="sourceLine" id="cb20-5" title="5">    <span class="dt">sample.id =</span> strata<span class="op">$</span>INDIVIDUALS,</a>
<a class="sourceLine" id="cb20-6" title="6">    <span class="dt">snp.id =</span> <span class="ot">NULL</span>,</a>
<a class="sourceLine" id="cb20-7" title="7">    <span class="dt">method =</span> <span class="st">"W&amp;C84"</span>,</a>
<a class="sourceLine" id="cb20-8" title="8">    <span class="dt">remove.monosnp =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb20-9" title="9">    <span class="dt">maf =</span> <span class="ot">NaN</span>,</a>
<a class="sourceLine" id="cb20-10" title="10">    <span class="dt">missing.rate =</span> <span class="ot">NaN</span>,</a>
<a class="sourceLine" id="cb20-11" title="11">    <span class="dt">autosome.only =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb20-12" title="12">    <span class="dt">with.id =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb20-13" title="13">    <span class="dt">verbose =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb20-14" title="14">  )</a>
<a class="sourceLine" id="cb20-15" title="15">)</a>
<a class="sourceLine" id="cb20-16" title="16"><span class="co">#&gt; 0.029sec wo!</span></a>
<a class="sourceLine" id="cb20-17" title="17"></a>
<a class="sourceLine" id="cb20-18" title="18"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(snprelate.fst)</a>
<a class="sourceLine" id="cb20-19" title="19"><span class="co">#&gt; snprelate.fst$Fst: weighted Fst estimate</span></a>
<a class="sourceLine" id="cb20-20" title="20"><span class="co">#&gt; snprelate.fst$MeanFst: the average of Fst estimates across SNPs</span></a>
<a class="sourceLine" id="cb20-21" title="21"><span class="co">#&gt; snprelate.fst$FstSNP: a vector of Fst for each SNP</span></a></code></pre></div>
<p>Build a function to conduct pairwise Fst with SNPRelate:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1">pairwise_fst_snprelate &lt;-<span class="st"> </span><span class="cf">function</span>(pop.pairwise, data, strata) {</a>
<a class="sourceLine" id="cb21-2" title="2">  strata.temp &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="dt">.data =</span> strata, STRATA <span class="op">%in%</span><span class="st"> </span>pop.pairwise) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb21-3" title="3"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">STRATA =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/droplevels">droplevels</a></span>(STRATA))</a>
<a class="sourceLine" id="cb21-4" title="4">  </a>
<a class="sourceLine" id="cb21-5" title="5">  fst &lt;-<span class="st"> </span>SNPRelate<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/SNPRelate/topics/snpgdsFst">snpgdsFst</a></span>(</a>
<a class="sourceLine" id="cb21-6" title="6">    <span class="dt">gdsobj =</span> data,</a>
<a class="sourceLine" id="cb21-7" title="7">    <span class="dt">population =</span> strata.temp<span class="op">$</span>STRATA, <span class="co"># factors required</span></a>
<a class="sourceLine" id="cb21-8" title="8">    <span class="dt">sample.id =</span> strata.temp<span class="op">$</span>INDIVIDUALS,</a>
<a class="sourceLine" id="cb21-9" title="9">    <span class="dt">snp.id =</span> <span class="ot">NULL</span>,</a>
<a class="sourceLine" id="cb21-10" title="10">    <span class="dt">method =</span> <span class="st">"W&amp;C84"</span>,</a>
<a class="sourceLine" id="cb21-11" title="11">    <span class="dt">remove.monosnp =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb21-12" title="12">    <span class="dt">maf =</span> <span class="ot">NaN</span>,</a>
<a class="sourceLine" id="cb21-13" title="13">    <span class="dt">missing.rate =</span> <span class="ot">NaN</span>,</a>
<a class="sourceLine" id="cb21-14" title="14">    <span class="dt">autosome.only =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb21-15" title="15">    <span class="dt">with.id =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb21-16" title="16">    <span class="dt">verbose =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb21-17" title="17">  )</a>
<a class="sourceLine" id="cb21-18" title="18">  <span class="co">#prepare the results into a tibble:</span></a>
<a class="sourceLine" id="cb21-19" title="19">  fst &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span>(</a>
<a class="sourceLine" id="cb21-20" title="20">    <span class="dt">POP1 =</span> pop.pairwise[<span class="dv">1</span>], </a>
<a class="sourceLine" id="cb21-21" title="21">    <span class="dt">POP2 =</span> pop.pairwise[<span class="dv">2</span>], </a>
<a class="sourceLine" id="cb21-22" title="22">    <span class="dt">SNPRELATE_MEAN =</span> fst<span class="op">$</span>MeanFst,</a>
<a class="sourceLine" id="cb21-23" title="23">    <span class="dt">SNPRELATE_WEIGHTED =</span> fst<span class="op">$</span>Fst</a>
<a class="sourceLine" id="cb21-24" title="24">  )</a>
<a class="sourceLine" id="cb21-25" title="25">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(fst)</a>
<a class="sourceLine" id="cb21-26" title="26">}</a></code></pre></div>
<p>To run we need all combination of populations:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" title="1">pop.pairwise &lt;-<span class="st"> </span>utils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/combn">combn</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(strata<span class="op">$</span>STRATA)), <span class="dv">2</span>, <span class="dt">simplify =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p>Run the pairwise implementation:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.time">system.time</a></span>(</a>
<a class="sourceLine" id="cb23-2" title="2">  snprelate.fst &lt;-<span class="st"> </span>purrr<span class="op">::</span><span class="kw"><a href="https://purrr.tidyverse.org/reference/map.html">map_dfr</a></span>(</a>
<a class="sourceLine" id="cb23-3" title="3">    <span class="dt">.x =</span> pop.pairwise,</a>
<a class="sourceLine" id="cb23-4" title="4">    <span class="dt">.f =</span> pairwise_fst_snprelate,</a>
<a class="sourceLine" id="cb23-5" title="5">    <span class="dt">data =</span> sunflower.filtered<span class="op">$</span>snprelate,</a>
<a class="sourceLine" id="cb23-6" title="6">    <span class="dt">strata =</span> strata</a>
<a class="sourceLine" id="cb23-7" title="7">  )</a>
<a class="sourceLine" id="cb23-8" title="8">)</a>
<a class="sourceLine" id="cb23-9" title="9"><span class="co">#&gt; 0.245 sec that's the fastest so far!</span></a></code></pre></div>
</div>
<div id="simulated-datasets-5" class="section level2">
<h2 class="hasAnchor">
<a href="#simulated-datasets-5" class="anchor"></a>Simulated datasets</h2>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" title="1">strata &lt;-<span class="st"> </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/generate_strata">generate_strata</a></span>(<span class="dt">data =</span> sim.data1<span class="op">$</span>tidy.data) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">rename</a></span>(<span class="dt">STRATA =</span> POP_ID)</a>
<a class="sourceLine" id="cb24-2" title="2">pop.pairwise &lt;-<span class="st"> </span>utils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/combn">combn</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(strata<span class="op">$</span>STRATA)), <span class="dv">2</span>, <span class="dt">simplify =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb24-3" title="3">snprelate.sim01.fst &lt;-<span class="st"> </span>purrr<span class="op">::</span><span class="kw"><a href="https://purrr.tidyverse.org/reference/map.html">map_dfr</a></span>(</a>
<a class="sourceLine" id="cb24-4" title="4">  <span class="dt">.x =</span> pop.pairwise,</a>
<a class="sourceLine" id="cb24-5" title="5">  <span class="dt">.f =</span> pairwise_fst_snprelate,</a>
<a class="sourceLine" id="cb24-6" title="6">  <span class="dt">data =</span> sim.data1<span class="op">$</span>snprelate,</a>
<a class="sourceLine" id="cb24-7" title="7">  <span class="dt">strata =</span> strata</a>
<a class="sourceLine" id="cb24-8" title="8">) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb24-9" title="9"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">rename</a></span>(<span class="dt">SNPRELATE_MEAN_SIM01 =</span> SNPRELATE_MEAN, <span class="dt">SNPRELATE_WEIGHTED_SIM01 =</span> SNPRELATE_WEIGHTED)</a>
<a class="sourceLine" id="cb24-10" title="10"></a>
<a class="sourceLine" id="cb24-11" title="11">strata &lt;-<span class="st"> </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/generate_strata">generate_strata</a></span>(<span class="dt">data =</span> sim.data2<span class="op">$</span>tidy.data) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">rename</a></span>(<span class="dt">STRATA =</span> POP_ID)</a>
<a class="sourceLine" id="cb24-12" title="12">pop.pairwise &lt;-<span class="st"> </span>utils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/combn">combn</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(strata<span class="op">$</span>STRATA)), <span class="dv">2</span>, <span class="dt">simplify =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb24-13" title="13">snprelate.sim02.fst &lt;-<span class="st"> </span>purrr<span class="op">::</span><span class="kw"><a href="https://purrr.tidyverse.org/reference/map.html">map_dfr</a></span>(</a>
<a class="sourceLine" id="cb24-14" title="14">  <span class="dt">.x =</span> pop.pairwise,</a>
<a class="sourceLine" id="cb24-15" title="15">  <span class="dt">.f =</span> pairwise_fst_snprelate,</a>
<a class="sourceLine" id="cb24-16" title="16">  <span class="dt">data =</span> sim.data2<span class="op">$</span>snprelate,</a>
<a class="sourceLine" id="cb24-17" title="17">  <span class="dt">strata =</span> strata</a>
<a class="sourceLine" id="cb24-18" title="18">) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb24-19" title="19"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">rename</a></span>(<span class="dt">SNPRELATE_MEAN_SIM02 =</span> SNPRELATE_MEAN, <span class="dt">SNPRELATE_WEIGHTED_SIM02 =</span> SNPRELATE_WEIGHTED)</a></code></pre></div>
</div>
</div>
<div id="genodive" class="section level1">
<h1 class="hasAnchor">
<a href="#genodive" class="anchor"></a>GenoDive</h1>
<ul>
<li>Inside Patrick Meirmansâ€™s <a href="http://www.bentleydrummer.nl/software/software/GenoDive.html">GenoDive</a> software, <em>import</em> or <em>drag</em> the <code>genepop</code> file</li>
<li>Run: <code>Analysis -&gt; Pairwise Differentiation -&gt; F_st</code>, choose <code>1 permutations</code>.</li>
<li>Save the results or copy/paste in MS Excel or text editor: <strong>genodive.fst.tsv</strong>.</li>
<li>no need for a coffee, <code>GenoDive</code> is coded in Objective-C (~ &lt; 3 sec on my MBP)</li>
</ul>
<div id="empirical-dataset-6" class="section level2">
<h2 class="hasAnchor">
<a href="#empirical-dataset-6" class="anchor"></a>Empirical dataset</h2>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" title="1"><span class="co"># Import in R</span></a>
<a class="sourceLine" id="cb25-2" title="2"><span class="co"># Add the proper columns and rows names</span></a>
<a class="sourceLine" id="cb25-3" title="3">pops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(sunflower.filtered<span class="op">$</span>tidy.data<span class="op">$</span>POP_ID)) <span class="co"># pop string</span></a>
<a class="sourceLine" id="cb25-4" title="4">genodive.fst &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(</a>
<a class="sourceLine" id="cb25-5" title="5">  <span class="dt">file =</span> <span class="st">"genodive.fst.tsv"</span>,</a>
<a class="sourceLine" id="cb25-6" title="6">  <span class="dt">col_names =</span> pops</a>
<a class="sourceLine" id="cb25-7" title="7">) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb25-8" title="8"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/strataG/topics/as.matrix-gtypes-method">as.matrix</a></span>(.) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb25-9" title="9"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_rownames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb25-10" title="10"><span class="st">  </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb25-11" title="11">    <span class="dt">x =</span> ., </a>
<a class="sourceLine" id="cb25-12" title="12">    <span class="dt">remove.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb25-13" title="13">    <span class="dt">remove.lower =</span> <span class="ot">TRUE</span>, <span class="co"># FALSE because their's only the lower diag</span></a>
<a class="sourceLine" id="cb25-14" title="14">    <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb25-15" title="15">    <span class="dt">relative =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb25-16" title="16">  ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb25-17" title="17"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>, <span class="st">"GENODIVE"</span>))</a>
<a class="sourceLine" id="cb25-18" title="18">genodive.fst</a></code></pre></div>
<p>Obviously, here the timing gets doesnâ€™t reflect the whole process:</p>
<ul>
<li>the data as to be imported in GenoDive</li>
<li>several manipulations of the results are required to get from GenoDive to R</li>
</ul>
</div>
<div id="simulated-datasets-6" class="section level2">
<h2 class="hasAnchor">
<a href="#simulated-datasets-6" class="anchor"></a>Simulated datasets</h2>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" title="1">pops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(sim.data1<span class="op">$</span>tidy.data<span class="op">$</span>POP_ID)) <span class="co"># pop string</span></a>
<a class="sourceLine" id="cb26-2" title="2">genodive.sim01.fst &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(</a>
<a class="sourceLine" id="cb26-3" title="3">  <span class="dt">file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(</a>
<a class="sourceLine" id="cb26-4" title="4">      <span class="dt">path =</span> <span class="st">"."</span>, </a>
<a class="sourceLine" id="cb26-5" title="5">      <span class="dt">pattern =</span> <span class="st">"genodive.sim01.fst.tsv"</span>, </a>
<a class="sourceLine" id="cb26-6" title="6">      <span class="dt">full.names =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb26-7" title="7">      <span class="dt">recursive =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb26-8" title="8">    ),</a>
<a class="sourceLine" id="cb26-9" title="9">  <span class="dt">col_names =</span> pops</a>
<a class="sourceLine" id="cb26-10" title="10">) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb26-11" title="11"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/strataG/topics/as.matrix-gtypes-method">as.matrix</a></span>(.) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb26-12" title="12"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_rownames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb26-13" title="13"><span class="st">  </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb26-14" title="14">    <span class="dt">x =</span> ., </a>
<a class="sourceLine" id="cb26-15" title="15">    <span class="dt">remove.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb26-16" title="16">    <span class="dt">remove.lower =</span> <span class="ot">TRUE</span>, <span class="co"># FALSE because their's only the lower diag</span></a>
<a class="sourceLine" id="cb26-17" title="17">    <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb26-18" title="18">    <span class="dt">relative =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb26-19" title="19">  ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb26-20" title="20"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>, <span class="st">"GENODIVE_SIM01"</span>))</a>
<a class="sourceLine" id="cb26-21" title="21">pops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(sim.data2<span class="op">$</span>tidy.data<span class="op">$</span>POP_ID)) <span class="co"># pop string</span></a>
<a class="sourceLine" id="cb26-22" title="22">genodive.sim02.fst &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(</a>
<a class="sourceLine" id="cb26-23" title="23">  <span class="dt">file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(</a>
<a class="sourceLine" id="cb26-24" title="24">      <span class="dt">path =</span> <span class="st">"."</span>, </a>
<a class="sourceLine" id="cb26-25" title="25">      <span class="dt">pattern =</span> <span class="st">"genodive.sim02.fst.tsv"</span>, </a>
<a class="sourceLine" id="cb26-26" title="26">      <span class="dt">full.names =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb26-27" title="27">      <span class="dt">recursive =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb26-28" title="28">    ),</a>
<a class="sourceLine" id="cb26-29" title="29">  <span class="dt">col_names =</span> pops</a>
<a class="sourceLine" id="cb26-30" title="30">) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb26-31" title="31"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/strataG/topics/as.matrix-gtypes-method">as.matrix</a></span>(.) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb26-32" title="32"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_rownames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> pops) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb26-33" title="33"><span class="st">  </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb26-34" title="34">    <span class="dt">x =</span> ., </a>
<a class="sourceLine" id="cb26-35" title="35">    <span class="dt">remove.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb26-36" title="36">    <span class="dt">remove.lower =</span> <span class="ot">TRUE</span>, <span class="co"># FALSE because their's only the lower diag</span></a>
<a class="sourceLine" id="cb26-37" title="37">    <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb26-38" title="38">    <span class="dt">relative =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb26-39" title="39">  ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb26-40" title="40"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>, <span class="st">"GENODIVE_SIM02"</span>))</a></code></pre></div>
</div>
</div>
<div id="stampp" class="section level1">
<h1 class="hasAnchor">
<a href="#stampp" class="anchor"></a>StAMPP</h1>
<p>Luke Pembletonâ€™s <code>StAMPP</code> package use <code>genlight</code> object as input:</p>
<div id="empirical-dataset-7" class="section level2">
<h2 class="hasAnchor">
<a href="#empirical-dataset-7" class="anchor"></a>Empirical dataset</h2>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.time">system.time</a></span>(</a>
<a class="sourceLine" id="cb27-2" title="2">  stampp.fst &lt;-<span class="st"> </span>StAMPP<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/StAMPP/topics/stamppFst">stamppFst</a></span>(</a>
<a class="sourceLine" id="cb27-3" title="3">    <span class="dt">geno =</span> sunflower.filtered<span class="op">$</span>genlight, </a>
<a class="sourceLine" id="cb27-4" title="4">    <span class="dt">nboots =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb27-5" title="5">    <span class="dt">percent =</span> <span class="dv">95</span>, </a>
<a class="sourceLine" id="cb27-6" title="6">    <span class="dt">nclusters =</span> parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/detectCores">detectCores</a></span>() <span class="op">-</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb27-7" title="7">  ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb27-8" title="8"><span class="st">    </span><span class="co"># The distance matrix need some love</span></a>
<a class="sourceLine" id="cb27-9" title="9"><span class="st">    </span><span class="co"># if we want to compare it with the others:</span></a>
<a class="sourceLine" id="cb27-10" title="10"><span class="st">    </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb27-11" title="11">      <span class="dt">x =</span> ., </a>
<a class="sourceLine" id="cb27-12" title="12">      <span class="dt">remove.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb27-13" title="13">      <span class="dt">remove.lower =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb27-14" title="14">      <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb27-15" title="15">      <span class="dt">relative =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb27-16" title="16">    ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb27-17" title="17"><span class="st">    </span><span class="co"># we switch the POP id column here to match the others</span></a>
<a class="sourceLine" id="cb27-18" title="18"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP2"</span>, <span class="st">"POP1"</span>, <span class="st">"STAMPP"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb27-19" title="19"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(POP1, POP2)</a>
<a class="sourceLine" id="cb27-20" title="20">)</a>
<a class="sourceLine" id="cb27-21" title="21"><span class="co">#&gt; 39.281 sec!</span></a></code></pre></div>
</div>
<div id="simulated-datasets-7" class="section level2">
<h2 class="hasAnchor">
<a href="#simulated-datasets-7" class="anchor"></a>Simulated datasets</h2>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1">stampp.sim01.fst &lt;-<span class="st"> </span>StAMPP<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/StAMPP/topics/stamppFst">stamppFst</a></span>(</a>
<a class="sourceLine" id="cb28-2" title="2">    <span class="dt">geno =</span> sim.data1<span class="op">$</span>genlight, </a>
<a class="sourceLine" id="cb28-3" title="3">    <span class="dt">nboots =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb28-4" title="4">    <span class="dt">percent =</span> <span class="dv">95</span>, </a>
<a class="sourceLine" id="cb28-5" title="5">    <span class="dt">nclusters =</span> parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/detectCores">detectCores</a></span>() <span class="op">-</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb28-6" title="6">  ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb28-7" title="7"><span class="st">    </span><span class="co"># The distance matrix need some love</span></a>
<a class="sourceLine" id="cb28-8" title="8"><span class="st">    </span><span class="co"># if we want to compare it with the others:</span></a>
<a class="sourceLine" id="cb28-9" title="9"><span class="st">    </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb28-10" title="10">      <span class="dt">x =</span> ., </a>
<a class="sourceLine" id="cb28-11" title="11">      <span class="dt">remove.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb28-12" title="12">      <span class="dt">remove.lower =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb28-13" title="13">      <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb28-14" title="14">      <span class="dt">relative =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb28-15" title="15">    ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb28-16" title="16"><span class="st">    </span><span class="co"># we switch the POP id column here to match the others</span></a>
<a class="sourceLine" id="cb28-17" title="17"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP2"</span>, <span class="st">"POP1"</span>, <span class="st">"STAMPP_SIM01"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb28-18" title="18"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(POP1, POP2)</a>
<a class="sourceLine" id="cb28-19" title="19"></a>
<a class="sourceLine" id="cb28-20" title="20">stampp.sim02.fst &lt;-<span class="st"> </span>StAMPP<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/StAMPP/topics/stamppFst">stamppFst</a></span>(</a>
<a class="sourceLine" id="cb28-21" title="21">    <span class="dt">geno =</span> sim.data2<span class="op">$</span>genlight, </a>
<a class="sourceLine" id="cb28-22" title="22">    <span class="dt">nboots =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb28-23" title="23">    <span class="dt">percent =</span> <span class="dv">95</span>, </a>
<a class="sourceLine" id="cb28-24" title="24">    <span class="dt">nclusters =</span> parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/detectCores">detectCores</a></span>() <span class="op">-</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb28-25" title="25">  ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb28-26" title="26"><span class="st">    </span><span class="co"># The distance matrix need some love</span></a>
<a class="sourceLine" id="cb28-27" title="27"><span class="st">    </span><span class="co"># if we want to compare it with the others:</span></a>
<a class="sourceLine" id="cb28-28" title="28"><span class="st">    </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb28-29" title="29">      <span class="dt">x =</span> ., </a>
<a class="sourceLine" id="cb28-30" title="30">      <span class="dt">remove.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb28-31" title="31">      <span class="dt">remove.lower =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb28-32" title="32">      <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb28-33" title="33">      <span class="dt">relative =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb28-34" title="34">    ) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb28-35" title="35"><span class="st">    </span><span class="co"># we switch the POP id column here to match the others</span></a>
<a class="sourceLine" id="cb28-36" title="36"><span class="st">    </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP2"</span>, <span class="st">"POP1"</span>, <span class="st">"STAMPP_SIM02"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb28-37" title="37"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(POP1, POP2)</a></code></pre></div>
</div>
</div>
<div id="vcftools" class="section level1">
<h1 class="hasAnchor">
<a href="#vcftools" class="anchor"></a>VCFtools</h1>
<p>Iâ€™m not particularly fond of VCFtools for Fst. Too much manipulation of files. For the sake of the exercise Iâ€™m going to compare it quickly with 2 populations for the empirical dataset only.</p>
<div id="empirical-dataset-8" class="section level2">
<h2 class="hasAnchor">
<a href="#empirical-dataset-8" class="anchor"></a>Empirical dataset</h2>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" title="1">pop1 &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(strata, STRATA <span class="op">==</span><span class="st"> "G100"</span>) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(INDIVIDUALS) <span class="op">%&gt;%</span><span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_tsv</a></span>(<span class="dt">x =</span> ., <span class="dt">path =</span> <span class="st">"pop1.txt"</span>, <span class="dt">col_names =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb29-2" title="2">pop2 &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(strata, STRATA <span class="op">==</span><span class="st"> "G102"</span>) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(INDIVIDUALS) <span class="op">%&gt;%</span><span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_tsv</a></span>(<span class="dt">x =</span> ., <span class="dt">path =</span> <span class="st">"pop2.txt"</span>, <span class="dt">col_names =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p>the command used:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb30-1" title="1"><span class="ex">vcftools</span> --vcf sunflower.vcf --weir-fst-pop pop1.txt --weir-fst-pop pop2.txt --out pop1_vs_pop2 </a>
<a class="sourceLine" id="cb30-2" title="2"><span class="co">#&gt; Fst between G100 and G102:</span></a>
<a class="sourceLine" id="cb30-3" title="3"><span class="co">#&gt; Weir and Cockerham mean Fst estimate: 0.036903</span></a>
<a class="sourceLine" id="cb30-4" title="4"><span class="co">#&gt; Weir and Cockerham weighted Fst estimate: 0.16222</span></a></code></pre></div>
<p>PLINK users, note that the <code>--fst</code> option in <a href="https://www.cog-genomics.org/plink/1.9/basic_stats#fst">PLINK v.1.9</a> is actually a port of VCFtools..</p>
<p>For me, comparison with VCFtools stops here, too painfull to conduct more pairwise, getting the ouput ready for R, etc. This is prone to human error and too much work to automate for no benefit (below youâ€™ll understand how those VCFtools results compares with the other packages).</p>
</div>
<div id="simulated-datasets-8" class="section level2">
<h2 class="hasAnchor">
<a href="#simulated-datasets-8" class="anchor"></a>Simulated datasets</h2>
<p><em>not conducted</em></p>
</div>
</div>
<div id="assigner" class="section level1">
<h1 class="hasAnchor">
<a href="#assigner" class="anchor"></a>assigner</h1>
<div id="empirical-dataset-9" class="section level2">
<h2 class="hasAnchor">
<a href="#empirical-dataset-9" class="anchor"></a>Empirical dataset</h2>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" title="1">assigner.fst &lt;-<span class="st"> </span>assigner<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/assigner/topics/fst_WC84">fst_WC84</a></span>(</a>
<a class="sourceLine" id="cb31-2" title="2">  <span class="dt">data =</span> sunflower.filtered<span class="op">$</span>tidy.data, </a>
<a class="sourceLine" id="cb31-3" title="3">  <span class="dt">pairwise =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb31-4" title="4">  <span class="dt">parallel.core =</span> parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/detectCores">detectCores</a></span>() <span class="op">-</span><span class="st"> </span><span class="dv">1</span>,</a>
<a class="sourceLine" id="cb31-5" title="5">  <span class="dt">verbose =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb31-6" title="6">) <span class="op">%$%</span></a>
<a class="sourceLine" id="cb31-7" title="7"><span class="st">  </span>pairwise.fst <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb31-8" title="8"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">rename</a></span>(<span class="dt">ASSIGNER =</span> FST) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb31-9" title="9"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="op">-</span>N_MARKERS)</a>
<a class="sourceLine" id="cb31-10" title="10"><span class="co">#&gt; 49 sec</span></a>
<a class="sourceLine" id="cb31-11" title="11">assigner.fst</a></code></pre></div>
<p>Learn more about <code><a href="../../reference/fst_WC84.html">assigner::fst_WC84</a></code> with the dedicated <a href="http://thierrygosselin.github.io/assigner/articles/fst_confidence_intervals.html">vignette</a>.</p>
</div>
<div id="simulated-datasets-9" class="section level2">
<h2 class="hasAnchor">
<a href="#simulated-datasets-9" class="anchor"></a>Simulated datasets</h2>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" title="1">assigner.sim01.fst &lt;-<span class="st"> </span>assigner<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/assigner/topics/fst_WC84">fst_WC84</a></span>(</a>
<a class="sourceLine" id="cb32-2" title="2">  <span class="dt">data =</span> sim.data1<span class="op">$</span>tidy.data, </a>
<a class="sourceLine" id="cb32-3" title="3">  <span class="dt">pairwise =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb32-4" title="4">  <span class="dt">parallel.core =</span> parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/detectCores">detectCores</a></span>() <span class="op">-</span><span class="st"> </span><span class="dv">1</span>,</a>
<a class="sourceLine" id="cb32-5" title="5">  <span class="dt">verbose =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb32-6" title="6">) <span class="op">%$%</span></a>
<a class="sourceLine" id="cb32-7" title="7"><span class="st">  </span>pairwise.fst <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb32-8" title="8"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">rename</a></span>(<span class="dt">ASSIGNER_SIM01 =</span> FST) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb32-9" title="9"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="op">-</span>N_MARKERS)</a>
<a class="sourceLine" id="cb32-10" title="10"></a>
<a class="sourceLine" id="cb32-11" title="11">assigner.sim02.fst &lt;-<span class="st"> </span>assigner<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/assigner/topics/fst_WC84">fst_WC84</a></span>(</a>
<a class="sourceLine" id="cb32-12" title="12">  <span class="dt">data =</span> sim.data2<span class="op">$</span>tidy.data, </a>
<a class="sourceLine" id="cb32-13" title="13">  <span class="dt">pairwise =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb32-14" title="14">  <span class="dt">parallel.core =</span> parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/detectCores">detectCores</a></span>() <span class="op">-</span><span class="st"> </span><span class="dv">1</span>,</a>
<a class="sourceLine" id="cb32-15" title="15">  <span class="dt">verbose =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb32-16" title="16">) <span class="op">%$%</span></a>
<a class="sourceLine" id="cb32-17" title="17"><span class="st">  </span>pairwise.fst <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb32-18" title="18"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">rename</a></span>(<span class="dt">ASSIGNER_SIM02 =</span> FST) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb32-19" title="19"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="op">-</span>N_MARKERS)</a></code></pre></div>
</div>
</div>
<div id="comparisons" class="section level1">
<h1 class="hasAnchor">
<a href="#comparisons" class="anchor"></a>Comparisons</h1>
<div id="empirical-dataset-10" class="section level2">
<h2 class="hasAnchor">
<a href="#empirical-dataset-10" class="anchor"></a>Empirical dataset</h2>
<div id="generate-a-summary-tibble" class="section level3">
<h3 class="hasAnchor">
<a href="#generate-a-summary-tibble" class="anchor"></a>Generate a summary tibble</h3>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" title="1">fst.summary &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/warning">suppressWarnings</a></span>(</a>
<a class="sourceLine" id="cb33-2" title="2">  assigner.fst <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb33-3" title="3"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(diversity.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb33-4" title="4"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(genodive.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb33-5" title="5"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(hierfstat.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb33-6" title="6"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(stampp.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb33-7" title="7"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(stratag.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb33-8" title="8"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(snprelate.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>))</a>
<a class="sourceLine" id="cb33-9" title="9">)</a></code></pre></div>
<p>If you havenâ€™t run the codes above:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" title="1">fst.summary &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(<span class="dt">file =</span> <span class="st">"fst.summary.tsv"</span>, <span class="dt">col_types =</span> <span class="st">"ccnnnnnnnn"</span>)</a>
<a class="sourceLine" id="cb34-2" title="2">fst.summary</a></code></pre></div>
<pre><code>## # A tibble: 36 x 10
##    POP1  POP2  ASSIGNER DIVERSITY GENODIVE HIERFSTAT STAMPP STRATAG
##    &lt;chr&gt; &lt;chr&gt;    &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;
##  1 G100  G102     0.123     0.123    0.115     0.123  0.123   0.123
##  2 G100  G103     0.210     0.210    0.215     0.210  0.211   0.211
##  3 G100  G108     0.394     0.395    0.39      0.394  0.394   0.394
##  4 G100  G109     0.299     0.299    0.287     0.299  0.299   0.299
##  5 G100  G111     0.369     0.370    0.356     0.369  0.369   0.369
##  6 G100  G118     0.169     0.169    0.156     0.169  0.169   0.169
##  7 G100  G122     0.269     0.270    0.255     0.269  0.269   0.269
##  8 G100  G124     0.366     0.366    0.362     0.366  0.366   0.366
##  9 G102  G103     0.118     0.118    0.115     0.118  0.119   0.119
## 10 G102  G108     0.322     0.323    0.319     0.322  0.322   0.322
## # â€¦ with 26 more rows, and 2 more variables: SNPRELATE_MEAN &lt;dbl&gt;,
## #   SNPRELATE_WEIGHTED &lt;dbl&gt;</code></pre>
<p>If we highlight the first line, pops G100 and G102:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" title="1">dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(fst.summary, dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/ranking.html">row_number</a></span>() <span class="op">==</span><span class="st"> </span>1L)</a></code></pre></div>
<pre><code>## # A tibble: 1 x 10
##   POP1  POP2  ASSIGNER DIVERSITY GENODIVE HIERFSTAT STAMPP STRATAG
##   &lt;chr&gt; &lt;chr&gt;    &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;
## 1 G100  G102     0.123     0.123    0.115     0.123  0.123   0.123
## # â€¦ with 2 more variables: SNPRELATE_MEAN &lt;dbl&gt;, SNPRELATE_WEIGHTED &lt;dbl&gt;</code></pre>
<p><strong>Gold Standard</strong>: HIERFSTAT 0.12</p>
<ul>
<li>min value: SNPRELATE_MEAN 0.08</li>
<li>max value: SNPRELATE_WEIGHTED 0.14</li>
<li>Weir and Cockerham mean Fst estimate: 0.036903 !!!</li>
<li>Weir and Cockerham weighted Fst estimate: 0.16222 !!!</li>
<li>thatâ€™s a bit scaryâ€¦</li>
</ul>
</div>
<div id="generate-a-correlation-matrix" class="section level3">
<h3 class="hasAnchor">
<a href="#generate-a-correlation-matrix" class="anchor"></a>Generate a correlation matrix</h3>
<p>Thereâ€™s obviously differences when we look at the pairwise comparisons separately. The difference between the approaches taken in the packages/software probably depends on numerous things:</p>
<ul>
<li>how common polymorphic markers are handle</li>
<li>how sample size differences are managed</li>
<li>missing data</li>
<li>rouding during steps of the codes</li>
</ul>
<p>Letâ€™s look at it globally with a correlation matrix:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" title="1">fst.corr &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/cor">cor</a></span>(<span class="dt">x =</span> dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(fst.summary, <span class="op">-</span>POP1, <span class="op">-</span>POP2)), <span class="dv">6</span>)</a>
<a class="sourceLine" id="cb38-2" title="2">fst.corr</a></code></pre></div>
<pre><code>##                    ASSIGNER DIVERSITY GENODIVE HIERFSTAT   STAMPP  STRATAG
## ASSIGNER           1.000000  0.999998 0.997723  1.000000 0.999996 0.999996
## DIVERSITY          0.999998  1.000000 0.997635  0.999998 0.999995 0.999995
## GENODIVE           0.997723  0.997635 1.000000  0.997723 0.997702 0.997702
## HIERFSTAT          1.000000  0.999998 0.997723  1.000000 0.999996 0.999996
## STAMPP             0.999996  0.999995 0.997702  0.999996 1.000000 1.000000
## STRATAG            0.999996  0.999995 0.997702  0.999996 1.000000 1.000000
## SNPRELATE_MEAN     0.973732  0.974023 0.965324  0.973732 0.973953 0.973953
## SNPRELATE_WEIGHTED 0.996752  0.996691 0.998345  0.996752 0.996729 0.996729
##                    SNPRELATE_MEAN SNPRELATE_WEIGHTED
## ASSIGNER                 0.973732           0.996752
## DIVERSITY                0.974023           0.996691
## GENODIVE                 0.965324           0.998345
## HIERFSTAT                0.973732           0.996752
## STAMPP                   0.973953           0.996729
## STRATAG                  0.973953           0.996729
## SNPRELATE_MEAN           1.000000           0.962916
## SNPRELATE_WEIGHTED       0.962916           1.000000</code></pre>
<p>I prefer to look at it in a tibbleâ€¦</p>
</div>
<div id="generate-a-tibble-for-correlation-results" class="section level3">
<h3 class="hasAnchor">
<a href="#generate-a-tibble-for-correlation-results" class="anchor"></a>Generate a tibble for correlation results</h3>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" title="1">fst.corr.tibble &lt;-<span class="st"> </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb40-2" title="2">    <span class="dt">x =</span> fst.corr, </a>
<a class="sourceLine" id="cb40-3" title="3">    <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb40-4" title="4">    <span class="dt">relative =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb40-5" title="5">    <span class="dt">pop.levels =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"HIERFSTAT"</span>, <span class="st">"STRATAG"</span>, <span class="st">"ASSIGNER"</span>, <span class="st">"DIVERSITY"</span>, <span class="st">"STAMPP"</span>, <span class="st">"GENODIVE"</span>, <span class="st">"SNPRELATE_MEAN"</span>, <span class="st">"SNPRELATE_WEIGHTED"</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb40-6" title="6"><span class="st">      </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"ID1"</span>, <span class="st">"ID2"</span>, <span class="st">"CORRELATION"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb40-7" title="7"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span>(CORRELATION))</a>
<a class="sourceLine" id="cb40-8" title="8">fst.corr.tibble</a></code></pre></div>
<pre><code>## # A tibble: 28 x 3
##    ID1       ID2       CORRELATION
##    &lt;ord&gt;     &lt;ord&gt;           &lt;dbl&gt;
##  1 ASSIGNER  HIERFSTAT       1    
##  2 STAMPP    STRATAG         1    
##  3 ASSIGNER  DIVERSITY       1.000
##  4 DIVERSITY HIERFSTAT       1.000
##  5 ASSIGNER  STAMPP          1.000
##  6 HIERFSTAT STAMPP          1.000
##  7 ASSIGNER  STRATAG         1.000
##  8 HIERFSTAT STRATAG         1.000
##  9 DIVERSITY STAMPP          1.000
## 10 DIVERSITY STRATAG         1.000
## # â€¦ with 18 more rows</code></pre>
<p>Very interesting results!</p>
</div>
<div id="figure-of-results" class="section level3">
<h3 class="hasAnchor">
<a href="#figure-of-results" class="anchor"></a>Figure of results</h3>
<p>Letâ€™s plot the results:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" title="1">plot.fst &lt;-<span class="st"> </span>fst.summary <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb42-2" title="2"><span class="st">  </span>tidyr<span class="op">::</span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/unite.html">unite</a></span>(<span class="dt">data =</span> ., <span class="dt">col =</span> PAIRS, POP1, POP2, <span class="dt">sep =</span> <span class="st">"-"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb42-3" title="3"><span class="st">  </span>tidyr<span class="op">::</span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span>(<span class="dt">data =</span> ., <span class="dt">key =</span> SOFTWARE, <span class="dt">value =</span> FST, <span class="op">-</span>PAIRS) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb42-4" title="4"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(</a>
<a class="sourceLine" id="cb42-5" title="5">    <span class="dt">SOFTWARE =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(</a>
<a class="sourceLine" id="cb42-6" title="6">      <span class="dt">x =</span> SOFTWARE, </a>
<a class="sourceLine" id="cb42-7" title="7">      <span class="dt">levels =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"HIERFSTAT"</span>, <span class="st">"ASSIGNER"</span>, <span class="st">"STAMPP"</span>, <span class="st">"STRATAG"</span>, <span class="st">"DIVERSITY"</span>, <span class="st">"GENODIVE"</span>, <span class="st">"SNPRELATE_WEIGHTED"</span>, <span class="st">"SNPRELATE_MEAN"</span>), <span class="dt">ordered =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb42-8" title="8"></a>
<a class="sourceLine" id="cb42-9" title="9">pairs.levels &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(plot.fst, PAIRS) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb42-10" title="10"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span>(<span class="dt">FST =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/mean">mean</a></span>(FST, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb42-11" title="11"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(FST) <span class="op">%$%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb42-12" title="12"><span class="st">  </span>PAIRS</a>
<a class="sourceLine" id="cb42-13" title="13">under.the.sea.palette &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"#1C2344"</span>, <span class="st">"#163D7D"</span>, <span class="st">"#118386"</span>, <span class="st">"#89D6D6"</span>, <span class="st">"#6E783D"</span>, <span class="st">"#FC564F"</span>, <span class="st">"#901546"</span>, <span class="st">"#FFD521"</span>)<span class="co">#, "#010101")</span></a>
<a class="sourceLine" id="cb42-14" title="14">plot.fst <span class="op">%&lt;&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb42-15" title="15"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">PAIRS =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(<span class="dt">x =</span> PAIRS, <span class="dt">levels =</span> pairs.levels, <span class="dt">ordered =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb42-16" title="16"><span class="st">  </span>ggplot2<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(<span class="dt">data =</span> ., ggplot2<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x =</span> FST, <span class="dt">y =</span> PAIRS, <span class="dt">colour =</span> SOFTWARE)) <span class="op">+</span></a>
<a class="sourceLine" id="cb42-17" title="17"><span class="st">  </span>ggplot2<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/geom_jitter">geom_jitter</a></span>(<span class="dt">alpha =</span> <span class="fl">0.8</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb42-18" title="18"><span class="st">  </span>ggplot2<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/scale_manual">scale_colour_manual</a></span>(<span class="dt">values =</span> under.the.sea.palette)</a></code></pre></div>
<p><img src="fst_comparisons_files/figure-html/plot%20the%20results-1.png" width="700"></p>
</div>
</div>
<div id="simulated-datasets-10" class="section level2">
<h2 class="hasAnchor">
<a href="#simulated-datasets-10" class="anchor"></a>Simulated datasets</h2>
<div id="sim_01-high-fst" class="section level3">
<h3 class="hasAnchor">
<a href="#sim_01-high-fst" class="anchor"></a>sim_01: high Fst</h3>
<p><strong>Generate a summary tibble:</strong></p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" title="1">fst.summary.sim01&lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/warning">suppressWarnings</a></span>(</a>
<a class="sourceLine" id="cb43-2" title="2">  assigner.sim01.fst <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb43-3" title="3"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(diversity.sim01.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb43-4" title="4"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(genodive.sim01.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb43-5" title="5"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(hierfstat.sim01.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb43-6" title="6"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(stampp.sim01.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb43-7" title="7"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(stratag.sim01.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb43-8" title="8"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(snprelate.sim01.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb43-9" title="9"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html">mutate_if</a></span>(<span class="dt">.tbl =</span> ., <span class="dt">.predicate =</span> is.numeric, <span class="dt">.funs =</span> round, <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb43-10" title="10">)</a></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" title="1">fst.summary.sim01 &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(<span class="dt">file =</span> <span class="st">"fst.summary.sim01.tsv"</span>, <span class="dt">col_types =</span> <span class="st">"ccnnnnnnnn"</span>)</a>
<a class="sourceLine" id="cb44-2" title="2">fst.summary.sim01</a></code></pre></div>
<pre><code>## # A tibble: 10 x 10
##    POP1  POP2  ASSIGNER_SIM01 DIVERSITY_SIM01 GENODIVE_SIM01
##    &lt;chr&gt; &lt;chr&gt;          &lt;dbl&gt;           &lt;dbl&gt;          &lt;dbl&gt;
##  1 POP_1 POP_2          0.514           0.514          0.514
##  2 POP_1 POP_3          0.481           0.481          0.481
##  3 POP_1 POP_4          0.425           0.425          0.425
##  4 POP_1 POP_5          0.410           0.410          0.410
##  5 POP_2 POP_3          0.412           0.412          0.412
##  6 POP_2 POP_4          0.370           0.370          0.370
##  7 POP_2 POP_5          0.360           0.360          0.360
##  8 POP_3 POP_4          0.356           0.356          0.356
##  9 POP_3 POP_5          0.341           0.341          0.341
## 10 POP_4 POP_5          0.301           0.301          0.301
## # â€¦ with 5 more variables: HIERFSTAT_SIM01 &lt;dbl&gt;, STAMPP_SIM01 &lt;dbl&gt;,
## #   STRATAG_SIM01 &lt;dbl&gt;, SNPRELATE_MEAN_SIM01 &lt;dbl&gt;,
## #   SNPRELATE_WEIGHTED_SIM01 &lt;dbl&gt;</code></pre>
<p><strong>Generate the correlation tibble:</strong></p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" title="1">fst.corr.tibble.sim01 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/cor">cor</a></span>(<span class="dt">x =</span> dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(fst.summary.sim01, <span class="op">-</span>POP1, <span class="op">-</span>POP2)), <span class="dv">6</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb46-2" title="2"><span class="st"> </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb46-3" title="3">  <span class="dt">x =</span> ., </a>
<a class="sourceLine" id="cb46-4" title="4">  <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb46-5" title="5">  <span class="dt">relative =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb46-6" title="6">  <span class="dt">pop.levels =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"HIERFSTAT_SIM01"</span>, <span class="st">"STRATAG_SIM01"</span>, <span class="st">"ASSIGNER_SIM01"</span>, <span class="st">"DIVERSITY_SIM01"</span>, <span class="st">"STAMPP_SIM01"</span>, <span class="st">"GENODIVE_SIM01"</span>, <span class="st">"SNPRELATE_MEAN_SIM01"</span>, <span class="st">"SNPRELATE_WEIGHTED_SIM01"</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb46-7" title="7"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"ID1"</span>, <span class="st">"ID2"</span>, <span class="st">"CORRELATION_SIM01"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb46-8" title="8"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span>(CORRELATION_SIM01))</a>
<a class="sourceLine" id="cb46-9" title="9">fst.corr.tibble.sim01</a></code></pre></div>
<pre><code>## # A tibble: 28 x 3
##    ID1             ID2             CORRELATION_SIM01
##    &lt;ord&gt;           &lt;ord&gt;                       &lt;dbl&gt;
##  1 ASSIGNER_SIM01  DIVERSITY_SIM01                 1
##  2 ASSIGNER_SIM01  GENODIVE_SIM01                  1
##  3 DIVERSITY_SIM01 GENODIVE_SIM01                  1
##  4 ASSIGNER_SIM01  HIERFSTAT_SIM01                 1
##  5 DIVERSITY_SIM01 HIERFSTAT_SIM01                 1
##  6 GENODIVE_SIM01  HIERFSTAT_SIM01                 1
##  7 ASSIGNER_SIM01  STAMPP_SIM01                    1
##  8 DIVERSITY_SIM01 STAMPP_SIM01                    1
##  9 GENODIVE_SIM01  STAMPP_SIM01                    1
## 10 HIERFSTAT_SIM01 STAMPP_SIM01                    1
## # â€¦ with 18 more rows</code></pre>
</div>
<div id="sim_02-low-fst" class="section level3">
<h3 class="hasAnchor">
<a href="#sim_02-low-fst" class="anchor"></a>sim_02: low Fst</h3>
<p><strong>Generate a summary tibble:</strong></p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" title="1">fst.summary.sim02&lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/warning">suppressWarnings</a></span>(</a>
<a class="sourceLine" id="cb48-2" title="2">  assigner.sim02.fst <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb48-3" title="3"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(diversity.sim02.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb48-4" title="4"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(genodive.sim02.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb48-5" title="5"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(hierfstat.sim02.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb48-6" title="6"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(stampp.sim02.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb48-7" title="7"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(stratag.sim02.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb48-8" title="8"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(snprelate.sim02.fst, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"POP1"</span>, <span class="st">"POP2"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb48-9" title="9"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html">mutate_if</a></span>(<span class="dt">.tbl =</span> ., <span class="dt">.predicate =</span> is.numeric, <span class="dt">.funs =</span> round, <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb48-10" title="10">)</a></code></pre></div>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" title="1">fst.summary.sim02 &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(<span class="dt">file =</span> <span class="st">"fst.summary.sim02.tsv"</span>, <span class="dt">col_types =</span> <span class="st">"ccnnnnnnnn"</span>)</a>
<a class="sourceLine" id="cb49-2" title="2">fst.summary.sim02</a></code></pre></div>
<pre><code>## # A tibble: 10 x 10
##    POP1  POP2  ASSIGNER_SIM02 DIVERSITY_SIM02 GENODIVE_SIM02
##    &lt;chr&gt; &lt;chr&gt;          &lt;dbl&gt;           &lt;dbl&gt;          &lt;dbl&gt;
##  1 POP_1 POP_2         0.001           0.001          0.001 
##  2 POP_1 POP_3         0.0032          0.0032         0.0032
##  3 POP_1 POP_4         0.0021          0.0021         0.0021
##  4 POP_1 POP_5         0.0022          0.0022         0.0022
##  5 POP_2 POP_3         0.0001          0.0001         0.0001
##  6 POP_2 POP_4         0.0002          0.0002         0.0002
##  7 POP_2 POP_5         0.0005          0.0005         0.0005
##  8 POP_3 POP_4         0              -0.0006        -0.0006
##  9 POP_3 POP_5         0.0028          0.0028         0.0028
## 10 POP_4 POP_5         0.0017          0.0017         0.0017
## # â€¦ with 5 more variables: HIERFSTAT_SIM02 &lt;dbl&gt;, STAMPP_SIM02 &lt;dbl&gt;,
## #   STRATAG_SIM02 &lt;dbl&gt;, SNPRELATE_MEAN_SIM02 &lt;dbl&gt;,
## #   SNPRELATE_WEIGHTED_SIM02 &lt;dbl&gt;</code></pre>
<p><strong>Generate the correlation tibble:</strong></p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" title="1">fst.corr.tibble.sim02 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/cor">cor</a></span>(<span class="dt">x =</span> dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(fst.summary.sim02, <span class="op">-</span>POP1, <span class="op">-</span>POP2)), <span class="dv">6</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb51-2" title="2"><span class="st"> </span>radiator<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/radiator/topics/distance2tibble">distance2tibble</a></span>(</a>
<a class="sourceLine" id="cb51-3" title="3">  <span class="dt">x =</span> ., </a>
<a class="sourceLine" id="cb51-4" title="4">  <span class="dt">na.diag =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb51-5" title="5">  <span class="dt">relative =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb51-6" title="6">  <span class="dt">pop.levels =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"HIERFSTAT_SIM02"</span>, <span class="st">"STRATAG_SIM02"</span>, <span class="st">"ASSIGNER_SIM02"</span>, <span class="st">"DIVERSITY_SIM02"</span>, <span class="st">"STAMPP_SIM02"</span>, <span class="st">"GENODIVE_SIM02"</span>, <span class="st">"SNPRELATE_MEAN_SIM02"</span>, <span class="st">"SNPRELATE_WEIGHTED_SIM02"</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-7" title="7"><span class="st">  </span>magrittr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/magrittr/topics/aliases">set_colnames</a></span>(<span class="dt">x =</span> ., <span class="dt">value =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"ID1"</span>, <span class="st">"ID2"</span>, <span class="st">"CORRELATION_SIM02"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb51-8" title="8"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span>(CORRELATION_SIM02))</a>
<a class="sourceLine" id="cb51-9" title="9">fst.corr.tibble.sim02</a></code></pre></div>
<pre><code>## # A tibble: 28 x 3
##    ID1             ID2             CORRELATION_SIM02
##    &lt;ord&gt;           &lt;ord&gt;                       &lt;dbl&gt;
##  1 DIVERSITY_SIM02 GENODIVE_SIM02                  1
##  2 DIVERSITY_SIM02 HIERFSTAT_SIM02                 1
##  3 GENODIVE_SIM02  HIERFSTAT_SIM02                 1
##  4 DIVERSITY_SIM02 STAMPP_SIM02                    1
##  5 GENODIVE_SIM02  STAMPP_SIM02                    1
##  6 HIERFSTAT_SIM02 STAMPP_SIM02                    1
##  7 DIVERSITY_SIM02 STRATAG_SIM02                   1
##  8 GENODIVE_SIM02  STRATAG_SIM02                   1
##  9 HIERFSTAT_SIM02 STRATAG_SIM02                   1
## 10 STAMPP_SIM02    STRATAG_SIM02                   1
## # â€¦ with 18 more rows</code></pre>
</div>
</div>
</div>
<div id="table-of-features" class="section level1">
<h1 class="hasAnchor">
<a href="#table-of-features" class="anchor"></a>Table of features</h1>
<p>The table timing was produced using the empirical dataset.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" title="1">summary.table &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(<span class="dt">file =</span> <span class="st">"fst_comparisons_packages_features.tsv"</span>, <span class="dt">col_types =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/cols.html">cols</a></span>(<span class="dt">.default =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>()))</a>
<a class="sourceLine" id="cb53-2" title="2">knitr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/knitr/topics/kable">kable</a></span>(summary.table) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb53-3" title="3"><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/kableExtra/topics/kable_styling">kable_styling</a></span>(<span class="dt">bootstrap_options =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"striped"</span>, <span class="st">"hover"</span>, <span class="st">"condensed"</span>, <span class="st">"responsive"</span>), <span class="dt">full_width =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>## Registered S3 method overwritten by 'rvest':
##   method            from
##   read_xml.response xml2</code></pre>
<pre><code>## Warning in kableExtra::kable_styling(., bootstrap_options = c("striped", :
## Please specify format in kable. kableExtra can customize either HTML or
## LaTeX outputs. See https://haozhu233.github.io/kableExtra/ for details.</code></pre>
<table class="table">
<thead><tr class="header">
<th align="left">FEATURES</th>
<th align="left">HIERFSTAT</th>
<th align="left">STRATAG</th>
<th align="left">ASSIGNER</th>
<th align="left">DIVERSITY</th>
<th align="left">STAMPP</th>
<th align="left">GENODIVE</th>
<th align="left">SNPRELATE</th>
<th align="left">VCFtools</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Function</td>
<td align="left">pairwise.WCfst</td>
<td align="left">popStructTest</td>
<td align="left">fst_WC84</td>
<td align="left">diffCalc</td>
<td align="left">stamppFst</td>
<td align="left">F_st</td>
<td align="left">snpgdsFst</td>
<td align="left">â€“weir-fst-pop</td>
</tr>
<tr class="even">
<td align="left">W&amp;C 1984 Fst</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
</tr>
<tr class="odd">
<td align="left">Language</td>
<td align="left">R</td>
<td align="left">R, C</td>
<td align="left">R</td>
<td align="left">R, C++</td>
<td align="left">R</td>
<td align="left">Objective-C</td>
<td align="left">R, C++</td>
<td align="left">Perl, C++</td>
</tr>
<tr class="even">
<td align="left">Data</td>
<td align="left">object</td>
<td align="left">object</td>
<td align="left">both</td>
<td align="left">file</td>
<td align="left">object</td>
<td align="left">file</td>
<td align="left">object</td>
<td align="left">file</td>
</tr>
<tr class="odd">
<td align="left">Format</td>
<td align="left">hierfstat</td>
<td align="left">gtypes</td>
<td align="left">several</td>
<td align="left">genepop</td>
<td align="left">genlight</td>
<td align="left">several</td>
<td align="left">SNPRelate GDS</td>
<td align="left">VCF</td>
</tr>
<tr class="even">
<td align="left">Polymorphic markers</td>
<td align="left">yes</td>
<td align="left">?</td>
<td align="left">yes</td>
<td align="left">?</td>
<td align="left">yes</td>
<td align="left">?</td>
<td align="left">yes</td>
<td align="left">yes</td>
</tr>
<tr class="odd">
<td align="left">Tailor common markers</td>
<td align="left">yes</td>
<td align="left">?</td>
<td align="left">yes</td>
<td align="left">?</td>
<td align="left">yes</td>
<td align="left">?</td>
<td align="left">?</td>
<td align="left">yes</td>
</tr>
<tr class="even">
<td align="left">Output number of markers</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">no</td>
</tr>
<tr class="odd">
<td align="left">Global fst</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
</tr>
<tr class="even">
<td align="left">Pairwise fst</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">no*</td>
<td align="left">no</td>
</tr>
<tr class="odd">
<td align="left">Bootstrap individuals</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
</tr>
<tr class="even">
<td align="left">Boostrap loci</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
</tr>
<tr class="odd">
<td align="left">CI</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
</tr>
<tr class="even">
<td align="left">P-Value</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
</tr>
<tr class="odd">
<td align="left">Locus pairwise</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
</tr>
<tr class="even">
<td align="left">Figures</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
</tr>
<tr class="odd">
<td align="left">Tibble</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
</tr>
<tr class="even">
<td align="left">Full upper and lower matrix</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">no</td>
</tr>
<tr class="odd">
<td align="left">Upper matrix</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
</tr>
<tr class="even">
<td align="left">Lower matrix</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
</tr>
<tr class="odd">
<td align="left">Combo matrix (Fst + stats)</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
</tr>
<tr class="even">
<td align="left">Fst markers</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
</tr>
<tr class="odd">
<td align="left">Fis</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
</tr>
<tr class="even">
<td align="left">Write</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">no</td>
<td align="left">yes</td>
</tr>
<tr class="odd">
<td align="left">Parallelize</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes*</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
</tr>
<tr class="even">
<td align="left">OS (linux, macOS, windows)</td>
<td align="left">all</td>
<td align="left">all</td>
<td align="left">all</td>
<td align="left">all</td>
<td align="left">all</td>
<td align="left">macOS</td>
<td align="left">all</td>
<td align="left">all</td>
</tr>
<tr class="odd">
<td align="left">Impacted by missingness</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">no</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
<td align="left">yes</td>
</tr>
<tr class="even">
<td align="left">Timing (sec)</td>
<td align="left">1234</td>
<td align="left">657</td>
<td align="left">49</td>
<td align="left">43</td>
<td align="left">39</td>
<td align="left">3</td>
<td align="left">0.245</td>
<td align="left">2*</td>
</tr>
</tbody>
</table>
<p><em>Note:</em></p>
<ul>
<li>VCFtools: we used only 2 populations and timing is inside VCFtools only, it doesnâ€™t account for manipulations of output.</li>
<li>SNPRelate: No out-of-the-box pairwise Fst analysis, but very quick solution is proposed in this vignette.</li>
</ul>
</div>
<div id="conclusion" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h1>
<p>You could go and look at the 36 pairwise comparisons, but the end of the story is very simple:</p>
<ul>
<li>For Fst, Iâ€™m definitely staying away from VCFtools and SNPRelate <code>MeanFst</code>.</li>
<li>
<code>SNPRelate</code> MeanFst as a systematic downward bias and the weighted estimate is usually a little higher.</li>
<li>Empirical and simulated data: <code>assigner</code> and <code>hierfstat</code> have equal values down to a very large number of decimals!</li>
<li>The packages show different colors with empirical and simulated data.
<ul>
<li>contrary to empirical data, with simulated data all the packages have the same results.</li>
<li>missing data is probably the stats driving those differences.</li>
<li>
<code>assigner</code> and <code>hierfstat</code> show consistent behaviors in the presence of missing data.</li>
<li>high or low Fst in simulated data: doesnâ€™t impact the packages.</li>
<li>
<code>assigner</code> is different than the others: it transform <code>-</code> Fst to 0.</li>
<li>it should be worth exploring the impact of missing data patterns and level of missingness on those Fst calculationsâ€¦</li>
</ul>
</li>
</ul>
<p>In the absence of missing data, because overall Fst values are similar, in the end, itâ€™s worth looking at:</p>
<ul>
<li>
<strong>speed</strong>: if this is an issue for you (e.g.Â if running multiple simulations, assignment analysis, etc).</li>
<li>
<strong>features</strong>: lots of differences between packages/software.</li>
</ul>
<p><em>Please let me know if I did something wrong or forgot to mention an interesting feature of your favorite package.</em></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#prepare-your-r-session">Prepare your R session</a></li>
      <li><a href="#goal-of-this-exercise">Goal of this exercise</a></li>
      <li>
<a href="#datasets">Datasets</a><ul class="nav nav-pills nav-stacked">
<li><a href="#empirical-dataset">Empirical dataset</a></li>
      <li><a href="#simulated-datasets">Simulated datasets</a></li>
      </ul>
</li>
      <li>
<a href="#generate-the-output-files">Generate the output files</a><ul class="nav nav-pills nav-stacked">
<li><a href="#empirical-dataset-1">Empirical dataset</a></li>
      <li><a href="#simulated-datasets-1">Simulated datasets</a></li>
      </ul>
</li>
      <li>
<a href="#hierfstat">hierfstat</a><ul class="nav nav-pills nav-stacked">
<li><a href="#empirical-dataset-2">Empirical dataset</a></li>
      <li><a href="#simulated-datasets-2">Simulated datasets</a></li>
      </ul>
</li>
      <li>
<a href="#stratag">strataG</a><ul class="nav nav-pills nav-stacked">
<li><a href="#empirical-dataset-3">Empirical dataset</a></li>
      <li><a href="#simulated-datasets-3">Simulated datasets</a></li>
      </ul>
</li>
      <li>
<a href="#diversity">diveRsity</a><ul class="nav nav-pills nav-stacked">
<li><a href="#empirical-dataset-4">Empirical dataset</a></li>
      <li><a href="#simulated-datasets-4">Simulated datasets</a></li>
      </ul>
</li>
      <li>
<a href="#snprelate">SNPRelate</a><ul class="nav nav-pills nav-stacked">
<li><a href="#empirical-dataset-5">Empirical dataset</a></li>
      <li><a href="#simulated-datasets-5">Simulated datasets</a></li>
      </ul>
</li>
      <li>
<a href="#genodive">GenoDive</a><ul class="nav nav-pills nav-stacked">
<li><a href="#empirical-dataset-6">Empirical dataset</a></li>
      <li><a href="#simulated-datasets-6">Simulated datasets</a></li>
      </ul>
</li>
      <li>
<a href="#stampp">StAMPP</a><ul class="nav nav-pills nav-stacked">
<li><a href="#empirical-dataset-7">Empirical dataset</a></li>
      <li><a href="#simulated-datasets-7">Simulated datasets</a></li>
      </ul>
</li>
      <li>
<a href="#vcftools">VCFtools</a><ul class="nav nav-pills nav-stacked">
<li><a href="#empirical-dataset-8">Empirical dataset</a></li>
      <li><a href="#simulated-datasets-8">Simulated datasets</a></li>
      </ul>
</li>
      <li>
<a href="#assigner">assigner</a><ul class="nav nav-pills nav-stacked">
<li><a href="#empirical-dataset-9">Empirical dataset</a></li>
      <li><a href="#simulated-datasets-9">Simulated datasets</a></li>
      </ul>
</li>
      <li>
<a href="#comparisons">Comparisons</a><ul class="nav nav-pills nav-stacked">
<li><a href="#empirical-dataset-10">Empirical dataset</a></li>
      <li><a href="#simulated-datasets-10">Simulated datasets</a></li>
      </ul>
</li>
      <li><a href="#table-of-features">Table of features</a></li>
      <li><a href="#conclusion">Conclusion</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
