<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Assigner features • assigner</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png" />
<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script>



<meta property="og:title" content="Assigner features" />

<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-title-body">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">assigner</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="FEATURES.html">features</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/assigner">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Assigner features</h1>
    </div>

<div id="assigner-features" class="section level1">

<div id="features" class="section level2">
<h2 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h2>
<table style="width:100%;">
<colgroup>
<col width="26%">
<col width="73%">
</colgroup>
<thead><tr class="header">
<th align="left">Caracteristics</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><strong>Analysis</strong></td>
<td align="left">
<a href="https://github.com/eriqande/gsi_sim">gsi_sim</a> (Anderson et al., 2008; Anderson, 2010) or <a href="https://github.com/thibautjombart/adegenet">adegenet</a> (Jombart, 2008; Jombart and Ahmed, 2011)<br><br>Conducted on: samples of <strong>known origin (population)</strong> or <strong>unknown/mixture</strong>
</td>
</tr>
<tr class="even">
<td align="left"><strong>Input files</strong></td>
<td align="left">assigner relies on <code><a href="https://www.rdocumentation.org/packages/radiator/topics/tidy_genomic_data">radiator::tidy_genomic_data</a></code> for imports</td>
</tr>
<tr class="odd">
<td align="left"><strong>Markers selections</strong></td>
<td align="left">Random or ranked based on W&amp;C’s Fst (Weir and Cockerham, 1984) or Nei’s Gst (Nei, 1987)</td>
</tr>
<tr class="even">
<td align="left"><strong>Cross-Validations</strong></td>
<td align="left">Avoid high-grading bias, use <strong>classic LOO</strong> (Leave-One-Out) or <strong>THL</strong> (Training, Holdout, Leave-one-out) (see Anderson, 2010)</td>
</tr>
<tr class="odd">
<td align="left"><strong>Resampling/Bootstrap/Subsampling</strong></td>
<td align="left">markers and individuals</td>
</tr>
<tr class="even">
<td align="left"><strong>Filters</strong></td>
<td align="left">
<code>strata</code>: use the strata to <strong>whitelist samples</strong> and/or <strong>select populations</strong> <code>??radiator::read_strata</code><br><code>whitelist.markers</code>: kleep favorite markers<br><code>marker.number</code>: test different groupings of marker numbers automatically (selected randomly or based on the ranking method described in the table above)<br><br>For more info on how to filter RADseq data, see <a href="https://github.com/thierrygosselin/radiator">radiator</a>.</td>
</tr>
<tr class="odd">
<td align="left"><strong><a href="http://ggplot2.org">ggplot2</a>-based plotting</strong></td>
<td align="left">View assignment results and create publication-ready figures</td>
</tr>
<tr class="even">
<td align="left"><strong>Parallel</strong></td>
<td align="left">Codes designed and optimized for fast computations running imputations, iterations, etc. in parallel. Works with all OS: Linux, Mac and PC.</td>
</tr>
</tbody>
</table>
</div>
<div id="independent-modules" class="section level2">
<h2 class="hasAnchor">
<a href="#independent-modules" class="anchor"></a>Independent modules</h2>
<ul>
<li>
<strong>Fst</strong>: assigner provides <strong>one of the fastest</strong> implementation of Weir and Cockerham (1984) Fst/Theta and Nei’s fst (1987). Both <strong>overall</strong> and <strong>pairwise Fst</strong> can be estimated with <strong>confidence intervals</strong> based on bootstrap of markers (resampling with replacement).</li>
<li>
<strong>Dlr:</strong> Compute the genotype likelihood ratio distance metric (Paetkau’s et al. 1997, 2004).</li>
</ul>
</div>
<div id="vignettes" class="section level2">
<h2 class="hasAnchor">
<a href="#vignettes" class="anchor"></a>Vignettes</h2>
<ul>
<li>Learn how to do a pairwise and overall Fst with confidence intervals + building phylogenetic tree <a href="https://www.dropbox.com/s/tiq4yenzmgzc2f5/fst_confidence_intervals.html?dl=0">R Notebook vignette</a>
</li>
<li>Why use assigner to compute Fst of W&amp;C (1984) in R ? <a href="https://www.dropbox.com/s/s4j6y498j1smohs/fst_comparisons.nb.html?dl=0">R Notebook vignette</a>
</li>
<li>Run the assignment with several whitelists of markers in <code><a href="reference/assignment_ngs.html">assigner::assignment_ngs</a></code><a href="https://www.dropbox.com/s/jzv91hbxi0g18kg/assignment.whitelists.nb.html?dl=0">(html vignette)</a>
</li>
<li><a href="https://www.dropbox.com/s/1kz59xpolb5y52m/rad_genomics_computer_setup.nb.html?dl=0">computer setup and troubleshooting</a></li>
</ul>
</div>
<div id="parallel-computation-and-cloud-computing" class="section level2">
<h2 class="hasAnchor">
<a href="#parallel-computation-and-cloud-computing" class="anchor"></a>Parallel computation and Cloud Computing</h2>
<p>Most of the function in <strong>assigner</strong> were designed to be as fast as possible. Using computer with 16GB RAM is recommended. With more CPU and Memory comes faster computation time. If you decide to keep intermediate files during assignment analysis, you will need a large external drive (disk space is cheap). Solid State Drive and thunderbolt cables will provide fast input/output.</p>
<p>If disk space and computer power is an issue, cloud computing with <a href="https://cloud.google.com/compute/">Google Cloud Compute Engine</a> and <a href="https://aws.amazon.com/ec2/">Amazon Elastic Cloud Compute</a> is cheap and can be used easily.</p>
<p>A tutorial and pipeline along an Amazon Machine Image (AMI) are available in our <a href="http://gbs-cloud-tutorial.readthedocs.org/en/latest/">tutorial-workflow</a>.</p>
<p>The AMI is preloaded with <strong>gsi_sim</strong> and the required <strong>R packages</strong>. Following a few steps: <a href="http://gbs-cloud-tutorial.readthedocs.org/en/latest/10_use_rstudio.html">link</a>, you can have <a href="https://www.rstudio.com/">RStudio server</a> running and used through your web browser!</p>
<p>The Amazon image can be imported into Google Cloud Compute Engine to start a new compute engine virtual machine: <a href="https://cloud.google.com/compute/docs/creating-custom-image#import_an_ami_image">link</a>.</p>
</div>
<div id="new-features" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features" class="anchor"></a>New features</h2>
<p>Change log, version, new features and bug history now lives in the <a href="https://github.com/thierrygosselin/assigner/blob/master/NEWS.md">NEWS.md file</a></p>
</div>
<div id="life-cycle" class="section level2">
<h2 class="hasAnchor">
<a href="#life-cycle" class="anchor"></a>Life cycle</h2>
<p><strong>Missing data imputations: now in <a href="https://github.com/thierrygosselin/grur">grur</a></strong></p>
<p>The imputation of missing data requires special attention that fall outside the scope of <strong>assigner</strong>. Consequently, these options are no longer available. For assignment, it’s better to do no imputation then quickly do imputations with defaults.</p>
<p>Inside my package called <a href="https://github.com/thierrygosselin/grur">grur</a>, users can <strong>visualize patterns of missingness</strong> associated with different variables (lanes, chips, sequencers, populations, sample sites, reads/samples, homozygosity, etc). Several <strong>Map-independent imputations</strong> of missing genotypes are available: <strong>Random Forests</strong> (on-the-fly-imputations or predictive modeling), <strong>Extreme Gradient Tree Boosting</strong>, Strawman imputations (~ max/mean/mode: the most frequently observed, non-missing genotypes is used). Imputations can be conducted <strong>overall samples</strong> or <strong>by populations/strata/grouping</strong>. <code><a href="https://www.rdocumentation.org/packages/radiator/topics/genomic_converter">radiator::genomic_converter</a></code> is integrated with the imputation function of <strong>grur</strong>.</p>
<ul>
<li>filter your data correctly in <a href="https://github.com/thierrygosselin/radiator">radiator</a>
</li>
<li>read on imputations an do tests in <a href="https://github.com/thierrygosselin/grur">grur</a>
</li>
<li>then use assigner…</li>
</ul>
</div>
<div id="roadmap-of-future-developments" class="section level2">
<h2 class="hasAnchor">
<a href="#roadmap-of-future-developments" class="anchor"></a>Roadmap of future developments</h2>
<ul>
<li>The ability to provide the ranking of markers based on other statistics, something else than the Fst (Weir and Cockerham, 1984) currently used in the function. e.g. Informativeness for Assignment Measure (In, Rosenberg et al. 2003), the Absolute Allele Frequency Differences (delta, δ, Shriver et al., 1997).</li>
<li>Provide ranking from other software: e.g. Toolbox for Ranking and Evaluation of SNPs <a href="http://mlkd.csd.auth.gr/bio/tres/">TRES</a>, <a href="http://cmpg.unibe.ch/software/BayeScan/">BayeScan</a> and <a href="https://github.com/whitlock/OutFLANK">OutFLANK</a>.</li>
<li>Would be very cool to use genotype likelihood information to get more accurate assignment.</li>
<li>Use Shiny and ggvis when subplots or facets are available</li>
<li>CRAN</li>
<li>…suggestions ?</li>
</ul>
</div>
<div id="contributions" class="section level2">
<h2 class="hasAnchor">
<a href="#contributions" class="anchor"></a>Contributions</h2>
<p>This package has been developed in the open, and it wouldn’t be nearly as good without your contributions. There are a number of ways you can help me make this package even better:</p>
<ul>
<li>If you don’t understand something, please let me know.</li>
<li>Your feedback on what is confusing or hard to understand is valuable.</li>
<li>If you spot a typo, feel free to edit the underlying page and send a pull request.</li>
</ul>
<p>New to pull request on github ? The process is very easy:</p>
<ul>
<li>Click the edit this page on the sidebar.</li>
<li>Make the changes using github’s in-page editor and save.</li>
<li>Submit a pull request and include a brief description of your changes.</li>
<li>“Fixing typos” is perfectly adequate.</li>
</ul>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>To get the citation for <strong>assigner</strong>, inside R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/citation">citation</a></span>(<span class="st">"assigner"</span>)</code></pre></div>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>Anderson EC, Waples RS, Kalinowski ST (2008) An improved method for predicting the accuracy of genetic stock identification. Canadian Journal of Fisheries and Aquatic Sciences, 65, 1475–1486.</p>
<p>Anderson EC (2010) Assessing the power of informative subsets of loci for population assignment: standard methods are upwardly biased. Molecular Ecology Resources, 10, 701–710.</p>
<p>Catchen JM, Amores A, Hohenlohe PA et al. (2011) Stacks: Building and Genotyping Loci De Novo From Short-Read Sequences. G3, 1, 171–182.</p>
<p>Catchen JM, Hohenlohe PA, Bassham S, Amores A, Cresko WA (2013) Stacks: an analysis tool set for population genomics. Molecular Ecology, 22, 3124–3140.</p>
<p>Danecek P, Auton A, Abecasis G et al. (2011) The variant call format and VCFtools. Bioinformatics, 27, 2156–2158.</p>
<p>Foll M, Gaggiotti O (2008) A Genome-Scan Method to Identify Selected Loci Appropriate for Both Dominant and Codominant Markers: A Bayesian Perspective. Genetics, 180, 977–993.</p>
<p>Goudet J.hierfstat, a package for r to compute and test hierarchical F-statistics. Molecular Ecology Notes.2005; 5: 184-186. <a href="doi:10.1111/j.1471-8286.2004.00828.x" class="uri">doi:10.1111/j.1471-8286.2004.00828.x</a></p>
<p>Ishwaran H. and Kogalur U.B. (2015). Random Forests for Survival, Regression and Classification (RF-SRC), R package version 1.6.1.</p>
<p>Jombart T (2008) adegenet: a R package for the multivariate analysis of genetic markers. Bioinformatics, 24, 1403–1405.</p>
<p>Jombart T, Ahmed I (2011) adegenet 1.3-1: new tools for the analysis of genome-wide SNP data. Bioinformatics, 27, 3070–3071.</p>
<p>Kavakiotis I, Triantafyllidis A, Ntelidou D et al. (2015) TRES: Identification of Discriminatory and Informative SNPs from Population Genomic Data. Journal of Heredity, 106, 672–676.</p>
<p>Meirmans PG, Van Tienderen PH (2004) genotype and genodive: two programs for the analysis of genetic diversity of asexual organisms. Molecular Ecology Notes, 4, 792-794.</p>
<p>Nei M. (1987) Molecular Evolutionary Genetics. Columbia University Press.</p>
<p>Paetkau D, Slade R, Burden M, Estoup A (2004) Genetic assignment methods for the direct, real-time estimation of migration rate: a simulation-based exploration of accuracy and power. Molecular Ecology, 13, 55-65.</p>
<p>Paetkau D, Waits LP, Clarkson PL, Craighead L, Strobeck C (1997) An empirical evaluation of genetic distance statistics using microsatellite data from bear (Ursidae) populations. Genetics, 147, 1943-1957.</p>
<p>Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MAR, Bender D, et al. PLINK: a tool set for whole-genome association and population-based linkage analyses. American Journal of Human Genetics. 2007; 81: 559–575. <a href="doi:10.1086/519795" class="uri">doi:10.1086/519795</a></p>
<p>Raymond M. &amp; Rousset F, (1995). GENEPOP (version 1.2): population genetics software for exact tests and ecumenicism. J. Heredity, 86:248-249</p>
<p>Rosenberg NA, Li LM, Ward R, Pritchard JK (2003) Informativeness of genetic markers for inference of ancestry. American Journal of Human Genetics, 73, 1402–1422.</p>
<p>Rousset F. genepop’007: a complete re-implementation of the genepop software for Windows and Linux. Molecular Ecology Resources. 2008, 8: 103-106. <a href="doi:10.1111/j.1471-8286.2007.01931.x" class="uri">doi:10.1111/j.1471-8286.2007.01931.x</a></p>
<p>Shriver MD, Smith MW, Jin L et al. (1997) Ethnic-affiliation estimation by use of population-specific DNA markers. American Journal of Human Genetics, 60, 957.</p>
<p>Weir BS, Cockerham CC (1984) Estimating F-Statistics for the Analysis of Population Structure. Evolution, 38, 1358–1370.</p>
<p>Whitlock MC, Lotterhos KE (2015) Reliable Detection of Loci Responsible for Local Adaptation: Inference of a Null Model through Trimming the Distribution of FST*. The American Naturalist, S000–S000.</p>
<p>Zheng X, Levine D, Shen J, Gogarten SM, Laurie C, Weir BS. A high-performance computing toolset for relatedness and principal component analysis of SNP data. Bioinformatics. 2012;28: 3326-3328. <a href="doi:10.1093/bioinformatics/bts606" class="uri">doi:10.1093/bioinformatics/bts606</a></p>
</div>
</div>

  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

