<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Assigner features • assigner</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png"><script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Assigner features"><meta property="og:image" content="https://thierrygosselin.github.io/assigner/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">assigner</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="index.html">Installation</a></li>
<li class="active nav-item"><a class="nav-link" href="FEATURES.html">Features</a></li>
<li class="nav-item"><a class="nav-link" href="articles/get_started.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="articles/get_started.html">Get started</a></li>
    <li><a class="dropdown-item" href="articles/fst_confidence_intervals.html">Fst analysis</a></li>
    <li><a class="dropdown-item" href="articles/web_only/fst_comparisons.html">Comparisons of Fst in 8 packages/software</a></li>
    <li><a class="dropdown-item" href="articles/assignment.whitelists.html">Using whitelists in assignment</a></li>
    <li><a class="dropdown-item" href="articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a></li>
    <li><a class="dropdown-item" href="articles/life_cycle.html">Life cycle</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/thierrygosselin/assigner" aria-label="GitHub"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-title-body">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="logo.png" class="logo" alt=""><h1>Assigner features</h1>
      <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/assigner/blob/HEAD/FEATURES.md" class="external-link"><code>FEATURES.md</code></a></small>
    </div>

<div id="assigner-features" class="section level1">

<div class="section level2">
<h2 id="features">Features<a class="anchor" aria-label="anchor" href="#features"></a></h2>
<table class="table"><colgroup><col width="25%"><col width="74%"></colgroup><thead><tr class="header"><th align="left">Caracteristics</th>
<th align="left">Description</th>
</tr></thead><tbody><tr class="odd"><td align="left"><strong>Analysis</strong></td>
<td align="left">
<a href="https://github.com/eriqande/gsi_sim" class="external-link">gsi_sim</a> (Anderson et al., 2008; Anderson, 2010) or <a href="https://github.com/thibautjombart/adegenet" class="external-link">adegenet</a> (Jombart, 2008; Jombart and Ahmed, 2011)<br><br>Conducted on: samples of <strong>known origin (population)</strong> or <strong>unknown/mixture</strong>
</td>
</tr><tr class="even"><td align="left"><strong>Input files</strong></td>
<td align="left">assigner relies on <code><a href="https://thierrygosselin.github.io/radiator/reference/tidy_genomic_data.html" class="external-link">radiator::tidy_genomic_data</a></code> for imports</td>
</tr><tr class="odd"><td align="left"><strong>Markers selections</strong></td>
<td align="left">Random or ranked based on W&amp;C’s Fst (Weir and Cockerham, 1984) or Nei’s Gst (Nei, 1987)</td>
</tr><tr class="even"><td align="left"><strong>Cross-Validations</strong></td>
<td align="left">Avoid high-grading bias, use <strong>classic LOO</strong> (Leave-One-Out) or <strong>THL</strong> (Training, Holdout, Leave-one-out) (see Anderson, 2010)</td>
</tr><tr class="odd"><td align="left"><strong>Resampling/Bootstrap/Subsampling</strong></td>
<td align="left">markers and individuals</td>
</tr><tr class="even"><td align="left"><strong>Filters</strong></td>
<td align="left">
<code>strata</code>: use the strata to <strong>whitelist samples</strong> and/or <strong>select populations</strong> <code>??radiator::read_strata</code><br><code>whitelist.markers</code>: kleep favorite markers<br><code>marker.number</code>: test different groupings of marker numbers automatically (selected randomly or based on the ranking method described in the table above)<br><br>For more info on how to filter RADseq data, see <a href="https://github.com/thierrygosselin/radiator" class="external-link">radiator</a>.</td>
</tr><tr class="odd"><td align="left"><strong><a href="http://ggplot2.org" class="external-link">ggplot2</a>-based plotting</strong></td>
<td align="left">View assignment results and create publication-ready figures</td>
</tr><tr class="even"><td align="left"><strong>Parallel</strong></td>
<td align="left">Codes designed and optimized for fast computations running imputations, iterations, etc. in parallel. Works with all OS: Linux, Mac and PC.</td>
</tr></tbody></table></div>
<div class="section level2">
<h2 id="independent-modules">Independent modules<a class="anchor" aria-label="anchor" href="#independent-modules"></a></h2>
<ul><li>
<strong>Fst</strong>: assigner provides <strong>one of the fastest</strong> implementation of Weir and Cockerham (1984) Fst/Theta and Nei’s fst (1987). Both <strong>overall</strong> and <strong>pairwise Fst</strong> can be estimated with <strong>confidence intervals</strong> based on bootstrap of markers (resampling with replacement).</li>
<li>
<strong>Dlr:</strong> Compute the genotype likelihood ratio distance metric (Paetkau’s et al. 1997, 2004).</li>
</ul></div>
<div class="section level2">
<h2 id="parallel-computation-and-cloud-computing">Parallel computation and Cloud Computing<a class="anchor" aria-label="anchor" href="#parallel-computation-and-cloud-computing"></a></h2>
<p>Most of the function in <strong>assigner</strong> were designed to be as fast as possible. Using computer with 16GB RAM is recommended. With more CPU and Memory comes faster computation time. If you decide to keep intermediate files during assignment analysis, you will need a large external drive (disk space is cheap). Solid State Drive and thunderbolt cables will provide fast input/output.</p>
<p>If disk space and computer power is an issue, cloud computing with <a href="https://www.r-bloggers.com/a-new-image-on-digitalocean-to-start-using-rstudio-server-without-waiting-more-than-2-minutes-2/" class="external-link">DigitalOcean</a>, <a href="https://cloud.google.com/compute/" class="external-link">Google Cloud Compute Engine</a> or <a href="https://aws.amazon.com/ec2/" class="external-link">Amazon Elastic Cloud Compute</a> is cheap and can be used easily.</p>
<p>A tutorial and pipeline along an Amazon Machine Image (AMI) are available in our <a href="http://gbs-cloud-tutorial.readthedocs.org/en/latest/" class="external-link">tutorial-workflow</a>.</p>
<p>The AMI is preloaded with <strong>gsi_sim</strong> and the required <strong>R packages</strong>. Following a <a href="http://gbs-cloud-tutorial.readthedocs.org/en/latest/10_use_rstudio.html" class="external-link">few steps</a>, you can have <a href="https://www.rstudio.com/" class="external-link">RStudio server</a> running and used through your web browser!</p>
<p>The Amazon image can be <a href="https://cloud.google.com/compute/docs/creating-custom-image#import_an_ami_image" class="external-link">imported into Google Cloud Compute Engine to start a new compute engine virtual machine</a>.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
<p>Anderson EC, Waples RS, Kalinowski ST (2008) An improved method for predicting the accuracy of genetic stock identification. Canadian Journal of Fisheries and Aquatic Sciences, 65, 1475–1486.</p>
<p>Anderson EC (2010) Assessing the power of informative subsets of loci for population assignment: standard methods are upwardly biased. Molecular Ecology Resources, 10, 701–710.</p>
<p>Catchen JM, Amores A, Hohenlohe PA et al. (2011) Stacks: Building and Genotyping Loci De Novo From Short-Read Sequences. G3, 1, 171–182.</p>
<p>Catchen JM, Hohenlohe PA, Bassham S, Amores A, Cresko WA (2013) Stacks: an analysis tool set for population genomics. Molecular Ecology, 22, 3124–3140.</p>
<p>Danecek P, Auton A, Abecasis G et al. (2011) The variant call format and VCFtools. Bioinformatics, 27, 2156–2158.</p>
<p>Foll M, Gaggiotti O (2008) A Genome-Scan Method to Identify Selected Loci Appropriate for Both Dominant and Codominant Markers: A Bayesian Perspective. Genetics, 180, 977–993.</p>
<p>Goudet J.hierfstat, a package for r to compute and test hierarchical F-statistics. Molecular Ecology Notes.2005; 5: 184-186. <a href="doi:10.1111/j.1471-8286.2004.00828.x" class="uri">doi:10.1111/j.1471-8286.2004.00828.x</a></p>
<p>Ishwaran H. and Kogalur U.B. (2015). Random Forests for Survival, Regression and Classification (RF-SRC), R package version 1.6.1.</p>
<p>Jombart T (2008) adegenet: a R package for the multivariate analysis of genetic markers. Bioinformatics, 24, 1403–1405.</p>
<p>Jombart T, Ahmed I (2011) adegenet 1.3-1: new tools for the analysis of genome-wide SNP data. Bioinformatics, 27, 3070–3071.</p>
<p>Kavakiotis I, Triantafyllidis A, Ntelidou D et al. (2015) TRES: Identification of Discriminatory and Informative SNPs from Population Genomic Data. Journal of Heredity, 106, 672–676.</p>
<p>Meirmans PG, Van Tienderen PH (2004) genotype and genodive: two programs for the analysis of genetic diversity of asexual organisms. Molecular Ecology Notes, 4, 792-794.</p>
<p>Nei M. (1987) Molecular Evolutionary Genetics. Columbia University Press.</p>
<p>Paetkau D, Slade R, Burden M, Estoup A (2004) Genetic assignment methods for the direct, real-time estimation of migration rate: a simulation-based exploration of accuracy and power. Molecular Ecology, 13, 55-65.</p>
<p>Paetkau D, Waits LP, Clarkson PL, Craighead L, Strobeck C (1997) An empirical evaluation of genetic distance statistics using microsatellite data from bear (Ursidae) populations. Genetics, 147, 1943-1957.</p>
<p>Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MAR, Bender D, et al.  PLINK: a tool set for whole-genome association and population-based linkage analyses. American Journal of Human Genetics. 2007; 81: 559–575. <a href="doi:10.1086/519795" class="uri">doi:10.1086/519795</a></p>
<p>Raymond M. &amp; Rousset F, (1995). GENEPOP (version 1.2): population genetics software for exact tests and ecumenicism. J. Heredity, 86:248-249</p>
<p>Rosenberg NA, Li LM, Ward R, Pritchard JK (2003) Informativeness of genetic markers for inference of ancestry. American Journal of Human Genetics, 73, 1402–1422.</p>
<p>Rousset F. genepop’007: a complete re-implementation of the genepop software for Windows and Linux. Molecular Ecology Resources. 2008, 8: 103-106. <a href="doi:10.1111/j.1471-8286.2007.01931.x" class="uri">doi:10.1111/j.1471-8286.2007.01931.x</a></p>
<p>Shriver MD, Smith MW, Jin L et al. (1997) Ethnic-affiliation estimation by use of population-specific DNA markers. American Journal of Human Genetics, 60, 957.</p>
<p>Weir BS, Cockerham CC (1984) Estimating F-Statistics for the Analysis of Population Structure. Evolution, 38, 1358–1370.</p>
<p>Whitlock MC, Lotterhos KE (2015) Reliable Detection of Loci Responsible for Local Adaptation: Inference of a Null Model through Trimming the Distribution of FST*. The American Naturalist, S000–S000.</p>
<p>Zheng X, Levine D, Shen J, Gogarten SM, Laurie C, Weir BS. A high-performance computing toolset for relatedness and principal component analysis of SNP data. Bioinformatics. 2012;28: 3326-3328. <a href="doi:10.1093/bioinformatics/bts606" class="uri">doi:10.1093/bioinformatics/bts606</a></p>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

