<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pairwise and overall Fst with confidence intervals + building phylogenetic tree • assigner</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Pairwise and overall Fst with confidence intervals + building phylogenetic tree">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">assigner</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/fst_confidence_intervals.html">Fst analysis</a>
    </li>
    <li>
      <a href="../articles/assignment.whitelists.html">Using whitelists in assignment</a>
    </li>
    <li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../news/index.html">Change log</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/thierrygosselin/assigner">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Pairwise and overall Fst with confidence intervals + building phylogenetic tree</h1>
                        <h4 class="author">Thierry Gosselin</h4>
            
            <h4 class="date">2019-04-26</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/assigner/blob/master/vignettes/fst_confidence_intervals.Rmd"><code>vignettes/fst_confidence_intervals.Rmd</code></a></small>
      <div class="hidden name"><code>fst_confidence_intervals.Rmd</code></div>

    </div>

    
    
<div id="objectives" class="section level1">
<h1 class="hasAnchor">
<a href="#objectives" class="anchor"></a>Objectives</h1>
<ul>
<li>learn how to run the function <code><a href="../reference/fst_WC84.html">assigner::fst_WC84</a></code>
</li>
<li>compute pairwise and overall Weir and Cockerham (1984) Fst estimates</li>
<li>build confidence intervals</li>
<li>visualize the estimates with data frame and matrices</li>
<li>learn how to integrate the function with other package, e.g. to build a phylogenetic tree</li>
<li>make Louis Bernatchez happy !</li>
</ul>
</div>
<div id="special-concerns-for-genome-wide-estimate-and-filtering-bias" class="section level1">
<h1 class="hasAnchor">
<a href="#special-concerns-for-genome-wide-estimate-and-filtering-bias" class="anchor"></a>Special concerns for genome-wide estimate and filtering bias</h1>
<p>During computation, the function first starts by keeping only the polymorphic markers in common between the populations.</p>
<p>When your project involves &gt; 2 populations that evolved more by neutral processes (e.g. genetic drift) than by natural selection, filtering to keep only the markers in common between your populations will drastically reduce the number of markers, and consequently your genome coverage for the estimate. Strategically, you should reduce the threshold of number of populations required to keep a marker (e.g. use a minimum of 2 populations genotyped for the markers to keep it). This is easily tested with <code>stackr::filter_population</code> or in stacks population module.</p>
<p>The overall fst and fis estimates will have reduce genome-wide coverage, because the computation is conducted with shared markers accross all your populations. However, for the pairwise comparisons, the markers will be tallied automatically for the pair. This usually results in more legitimate markers (i.e. less filtering bias) being used during computation, providing better genome-wide estimates for that pair.</p>
</div>
<div id="workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow" class="anchor"></a>Workflow</h1>
<div id="preparation" class="section level2">
<h2 class="hasAnchor">
<a href="#preparation" class="anchor"></a>Preparation</h2>
<ol style="list-style-type: decimal">
<li>
<p>Clean your desk and remove the clutter in the global environment</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rm">rm</a></span>(<span class="dt">list=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ls">ls</a></span>())</code></pre></div>
</li>
<li>
<p>Required libraries</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"thierrygosselin/assigner"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"assigner"</span>)</code></pre></div>
</li>
<li>
<p>Get the data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data.fst &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(<span class="dt">file =</span> <span class="st">"https://raw.githubusercontent.com/thierrygosselin/package_data/master/assigner_data_fst.tsv.gz"</span>)</code></pre></div>
</li>
</ol>
</div>
<div id="fst-calculations" class="section level2">
<h2 class="hasAnchor">
<a href="#fst-calculations" class="anchor"></a>Fst calculations</h2>
<p><code><a href="../reference/fst_WC84.html">assigner::fst_WC84</a></code> first run: overall and pairwise, but without boostrap (i.e. no confidence intervals). Get the function documentation with: <code>??assigner::fst_WC84</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fst &lt;-<span class="st"> </span>assigner<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/assigner/topics/fst_WC84">fst_WC84</a></span>(
    <span class="dt">data =</span> data.fst, 
    <span class="dt">pop.levels =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"pop1"</span>, <span class="st">"pop2"</span>, <span class="st">"pop3"</span>, <span class="st">"pop4"</span>, <span class="st">"pop5"</span>, <span class="st">"pop6"</span>, <span class="st">"pop7"</span>, <span class="st">"pop8"</span>, <span class="st">"pop9"</span>, <span class="st">"pop10"</span>, <span class="st">"pop11"</span>),
    <span class="dt">pairwise =</span> <span class="ot">TRUE</span>,
    <span class="dt">filename =</span> <span class="st">"testing_fst"</span>, <span class="co">#will trigger the function to write the results in directory as well</span>
    <span class="dt">verbose =</span> <span class="ot">TRUE</span>
)</code></pre></div>
</div>
<div id="results" class="section level2">
<h2 class="hasAnchor">
<a href="#results" class="anchor"></a>Results</h2>
<p>After 10 sec max you should get the Fst (overall) value of 0.05</p>
<ol style="list-style-type: decimal">
<li>
<p>To view the content of <code>fst</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(fst)</code></pre></div>
</li>
<li>Get access to specific content with <code>$</code>:
<ul>
<li><p><code>$sigma.loc</code>: the variance components per locus: <code>lsiga</code>: among populations, <code>lsigb</code>: among individuals within populations, <code>lsigw</code>: within individuals.</p></li>
<li>
<code>$fst.markers</code>: the fst by markers</li>
<li>
<code>$fst.ranked</code>: the fst ranked</li>
<li>
<code>$fst.overall</code>: the mean fst overall markers and number of markers</li>
<li>
<code>$fis.markers</code>: the fis by markers</li>
<li>
<code>$fis.overall</code>: the mean fis overall markers and number of markers</li>
<li>
<code>$fst.plot</code>: the histogram of the overall Fst per markers</li>
<li>
<code>$pairwise.fst</code>: the pairwise fst in long/tidy data frame and number of markers per pairwise comparison</li>
<li>
<code>$pairwise.fst.upper.matrix</code>: the pairwise fst in a upper triangle matrix</li>
<li>
<code>$pairwise.fst.full.matrix</code>: the pairwise fst matrix (duplicated upper and lower triangle)</li>
<li><p><code>$pairwise.fst.ci.matrix</code>: matrix with pairwise fst in the upper triangle and the confidence intervals in the lower triangle.</p></li>
</ul>
</li>
<li>
<p>To view the pairwise Fst values in a tibble with 4 columns:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df &lt;-<span class="st"> </span>fst<span class="op">$</span>pairwise.fst</code></pre></div>
</li>
<li>
<p>The first 10 parwise comparisons in <code>df</code></p>
<p><em>N_MARKERS</em>: is the number of markers used between the pairs to calculate the Fst.</p>
<table class="table">
<thead><tr class="header">
<th align="center">POP1</th>
<th align="center">POP2</th>
<th align="center">FST</th>
<th align="center">N_MARKERS</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">pop1</td>
<td align="center">pop2</td>
<td align="center">0.0442</td>
<td align="center">3457</td>
</tr>
<tr class="even">
<td align="center">pop1</td>
<td align="center">pop3</td>
<td align="center">0.0507</td>
<td align="center">3457</td>
</tr>
<tr class="odd">
<td align="center">pop1</td>
<td align="center">pop4</td>
<td align="center">0.0341</td>
<td align="center">3457</td>
</tr>
<tr class="even">
<td align="center">pop1</td>
<td align="center">pop5</td>
<td align="center">0.0300</td>
<td align="center">3457</td>
</tr>
<tr class="odd">
<td align="center">pop1</td>
<td align="center">pop6</td>
<td align="center">0.0434</td>
<td align="center">3457</td>
</tr>
<tr class="even">
<td align="center">pop1</td>
<td align="center">pop7</td>
<td align="center">0.0367</td>
<td align="center">3457</td>
</tr>
<tr class="odd">
<td align="center">pop1</td>
<td align="center">pop8</td>
<td align="center">0.0256</td>
<td align="center">3457</td>
</tr>
<tr class="even">
<td align="center">pop1</td>
<td align="center">pop9</td>
<td align="center">0.0475</td>
<td align="center">3457</td>
</tr>
<tr class="odd">
<td align="center">pop1</td>
<td align="center">pop10</td>
<td align="center">0.0522</td>
<td align="center">3457</td>
</tr>
<tr class="even">
<td align="center">pop1</td>
<td align="center">pop11</td>
<td align="center">0.0473</td>
<td align="center">3457</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>The matrix with fst values:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># to see as a tibble:</span>
fst.matrix &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span>(fst<span class="op">$</span>pairwise.fst.full.matrix, <span class="dt">rownames =</span> <span class="st">"POP"</span>)
<span class="co"># to keep matrix:</span>
fst.matrix &lt;-<span class="st"> </span>fst<span class="op">$</span>pairwise.fst.full.matrix</code></pre></div>
</li>
<li>
<p>If you try to get the <code>pairwise.fst.ci.matrix</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fst<span class="op">$</span>pairwise.fst.ci.matrix <span class="co"># you will get this:</span>
[<span class="dv">1</span>] <span class="st">"confidence intervals not selected"</span></code></pre></div>
</li>
<li>
<p>So let’s make another run with <code>fst_WC84</code>, but this time will ask for the confidence intervals and the heatmap:</p>
<div class="sourceCode"><pre class="sourceCode rs"><code class="sourceCode rust">fst.ci &lt;- fst_WC84(data = data.fst, 
pop.levels = c(<span class="st">"pop1"</span>, <span class="st">"pop2"</span>, <span class="st">"pop3"</span>, <span class="st">"pop4"</span>, <span class="st">"pop5"</span>, <span class="st">"pop6"</span>, <span class="st">"pop7"</span>, <span class="st">"pop8"</span>, <span class="st">"pop9"</span>, <span class="st">"pop10"</span>, <span class="st">"pop11"</span>),
pairwise = TRUE,
ci = TRUE, 
iteration.ci = <span class="dv">100</span>,
quantiles.ci = c(<span class="dv">0.025</span>, <span class="dv">0.975</span>),
parallel.core = <span class="dv">12</span>,
heatmap.fst = TRUE,
filename = <span class="st">"testing_fst"</span>,
verbose = TRUE
)</code></pre></div>
<p><code>fst.ci$pairwise.fst</code> is the tibble with the Fst values and now with the confidence intervals.</p>
</li>
<li>
<p>Try again to get the matrix with fst and ci:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fst.ci<span class="op">$</span>pairwise.fst.ci.matrix</code></pre></div>
</li>
<li>
<p>To have this in a handy table for publication, we will create a new object <code>pairwise.fst.ci.df</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pairwise.fst.ci.df &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span>(fst.ci<span class="op">$</span>pairwise.fst.ci.matrix, <span class="dt">rownames =</span> <span class="st">"POP"</span>)
<span class="co"># if you have numeric pop_id identifier you might have to do this to get proper column names:</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(pairwise.fst.ci.df) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(fst.ci<span class="op">$</span>pairwise.fst.ci.matrix)
<span class="co"># to save:</span>
readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_tsv</a></span>(<span class="dt">x =</span> pairwise.fst.ci.df, <span class="dt">path =</span> <span class="st">"pairwise.fst.ci.df.tsv"</span>)</code></pre></div>
</li>
<li><p>Or even better, use the heatmap of Fst and CI values…: <img src="../../../../../Mobile%20Documents/com~apple~CloudDocs/r_packages/assigner/vignettes/fst_WC84_heatmap_fst.png" id="id" class="class" width="600" height="600"></p></li>
</ol>
</div>
<div id="phylogenetic-tree" class="section level2">
<h2 class="hasAnchor">
<a href="#phylogenetic-tree" class="anchor"></a>Phylogenetic tree</h2>
<p>For the next steps, we need the full distance matrix created in step 10.</p>
<ol style="list-style-type: decimal">
<li>
<p>For a Neighbor joining tree:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># build the tree:</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(ape)
tree &lt;-<span class="st"> </span>ape<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ape/topics/nj">nj</a></span>(<span class="dt">X =</span> fst.matrix) <span class="co"># fst.matrix as a matrix</span>
<span class="co"># for annotating bootstraps values on the tree:</span>
bootstrap.value &lt;-<span class="st"> </span>ape<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ape/topics/boot.phylo">boot.phylo</a></span>(
<span class="dt">phy =</span> tree, 
<span class="dt">x =</span> fst.matrix, 
<span class="dt">FUN =</span> <span class="cf">function</span>(x) ape<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ape/topics/nj">nj</a></span>(x), 
<span class="dt">block =</span> <span class="dv">1</span>, 
<span class="dt">B =</span> <span class="dv">10000</span>, 
<span class="dt">trees =</span> <span class="ot">FALSE</span>, 
<span class="dt">rooted =</span> <span class="ot">FALSE</span>
)
 <span class="co"># to get percentage values</span>
bootstrap.value &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>((bootstrap.value<span class="op">/</span><span class="dv">10000</span>)<span class="op">*</span><span class="dv">100</span>, <span class="dv">0</span>)
bootstrap.value
<span class="co"># to include in the tree</span>
tree<span class="op">$</span>node.label &lt;-<span class="st"> </span>bootstrap.value </code></pre></div>
</li>
<li>
<p>For a UPGMA tree:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(stats)
tree &lt;-<span class="st"> </span>ape<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ape/topics/as.phylo">as.phylo</a></span>(stats<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/hclust">hclust</a></span>(stats<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/dist">dist</a></span>(fst.matrix), <span class="dt">method =</span> <span class="st">"average"</span>))
bootstrap.value &lt;-<span class="st"> </span>ape<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ape/topics/boot.phylo">boot.phylo</a></span>(<span class="dt">phy =</span> tree, <span class="dt">x =</span> fst.matrix, <span class="dt">FUN =</span> <span class="cf">function</span>(xx) ape<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ape/topics/as.phylo">as.phylo</a></span>(stats<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/hclust">hclust</a></span>(stats<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/dist">dist</a></span>(xx), <span class="dt">method =</span> <span class="st">"average"</span>)) , <span class="dt">block =</span> <span class="dv">1</span>, <span class="dt">B =</span> <span class="dv">10000</span>, <span class="dt">trees =</span> <span class="ot">FALSE</span>, <span class="dt">rooted =</span> <span class="ot">TRUE</span>) 
bootstrap.value &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>((bootstrap.value<span class="op">/</span><span class="dv">10000</span>)<span class="op">*</span><span class="dv">100</span>, <span class="dv">0</span>)
bootstrap.value
tree<span class="op">$</span>node.label &lt;-<span class="st"> </span>bootstrap.value</code></pre></div>
</li>
<li>
<p>To build the tree we will need to install </p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get the latest development version of ggtree:</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"ggtree"</span>)) <span class="kw">install_github</span>(<span class="st">"GuangchuangYu/ggtree"</span>)
<span class="co"># If it's not working, use the Bioconductor version:</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ns-load">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)) <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"BiocManager"</span>)
BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"ggtree"</span>)</code></pre></div>
</li>
<li><p>Several vignettes are available to get to know <a href="http://guangchuangyu.github.io/ggtree/">ggtree</a></p></li>
<li>
<p>Build a very basic tree figure:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(ggtree)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(ggplot2)
tree.figure &lt;-<span class="st"> </span>ggplot2<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(tree, ggplot2<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(x, y), <span class="dt">ladderize =</span> <span class="ot">TRUE</span>) <span class="op">+</span><span class="st"> </span>
ggtree<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggtree/topics/geom_tree">geom_tree</a></span>() <span class="op">+</span><span class="st"> </span>
<span class="co"># geom_tiplab(size = 3, hjust = -0.05, vjust = 0.5)+ # for just the tip label</span>
ggplot2<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/geom_text">geom_text</a></span>(ggplot2<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">label =</span> label), <span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">hjust =</span> <span class="op">-</span><span class="fl">0.05</span>, <span class="dt">vjust =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span><span class="co"># for both tips and nodes</span>
ggtree<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggtree/topics/theme_tree">theme_tree</a></span>() <span class="op">+</span><span class="st"> </span>
ggplot2<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/lims">xlim</a></span>(<span class="dv">0</span>, <span class="fl">0.16</span>) <span class="co"># to allocate more space for tip labels (trial/error)</span>
tree.figure
ggplot2<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/ggsave">ggsave</a></span>(<span class="dt">filename =</span> <span class="st">"assigner.fst.tree.example.pdf"</span>, <span class="dt">width =</span> <span class="dv">15</span>, <span class="dt">height =</span> <span class="dv">15</span>, <span class="dt">dpi =</span> <span class="dv">600</span>, <span class="dt">units =</span> <span class="st">"cm"</span>, <span class="dt">useDingbats =</span> <span class="ot">FALSE</span>)</code></pre></div>
</li>
</ol>
<div class="figure">
<img src="../../../../../Mobile%20Documents/com~apple~CloudDocs/r_packages/assigner/vignettes/assigner_fst_tree_example.png" id="id" class="class" width="600" height="600">
</div>
</div>
</div>
<div id="conclusion" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h1>
<p>Please send me suggestions and bug reports through github</p>
<p><strong>References</strong></p>
<p>Weir BS, Cockerham CC (1984) Estimating F-Statistics for the Analysis of Population Structure. Evolution, 38, 1358–1370.</p>
<p>G Yu, D Smith, H Zhu, Y Guan, TTY Lam, ggtree: an R package for visualization and annotation of phylogenetic tree with different types of meta-data. revised.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#objectives">Objectives</a></li>
      <li><a href="#special-concerns-for-genome-wide-estimate-and-filtering-bias">Special concerns for genome-wide estimate and filtering bias</a></li>
      <li>
<a href="#workflow">Workflow</a><ul class="nav nav-pills nav-stacked">
<li><a href="#preparation">Preparation</a></li>
      <li><a href="#fst-calculations">Fst calculations</a></li>
      <li><a href="#results">Results</a></li>
      <li><a href="#phylogenetic-tree">Phylogenetic tree</a></li>
      </ul>
</li>
      <li><a href="#conclusion">Conclusion</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
