<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Assignment analysis tailored for RADseq data — assignment_ngs • assigner</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Assignment analysis tailored for RADseq data — assignment_ngs"><meta name="description" content="The arguments in the assignment_ngs function were tailored for the
reality of RADseq data for assignment analysis while
maintaining a reproducible workflow.
Assignment assumptions are listed in the section below.

Input file: various file format are supported (see data argument below).
Cross-Validations: Markers can be randomly selected for a classic LOO (Leave-One-Out)
  assignment or chosen based on ranked Fst for a thl
  (Training, Holdout, Leave-one-out) assignment analysis.
Assignment analysis: conducted in
  gsi_sim, a tool
  for doing and simulating genetic stock identification and
  developed by Eric C. Anderson, or
  adegenet,
 an R package developed by Thibaul Jombart.
Parallel: The assignment can be conduncted on multiple CPUs.
  The R GUI is unstable with this functions, I recommend using
  RStudio.

"><meta property="og:description" content="The arguments in the assignment_ngs function were tailored for the
reality of RADseq data for assignment analysis while
maintaining a reproducible workflow.
Assignment assumptions are listed in the section below.

Input file: various file format are supported (see data argument below).
Cross-Validations: Markers can be randomly selected for a classic LOO (Leave-One-Out)
  assignment or chosen based on ranked Fst for a thl
  (Training, Holdout, Leave-one-out) assignment analysis.
Assignment analysis: conducted in
  gsi_sim, a tool
  for doing and simulating genetic stock identification and
  developed by Eric C. Anderson, or
  adegenet,
 an R package developed by Thibaul Jombart.
Parallel: The assignment can be conduncted on multiple CPUs.
  The R GUI is unstable with this functions, I recommend using
  RStudio.

"><meta property="og:image" content="https://thierrygosselin.github.io/assigner/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">assigner</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../FEATURES.html">Features</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/get_started.html">Get started</a></li>
    <li><a class="dropdown-item" href="../articles/fst_confidence_intervals.html">Fst analysis</a></li>
    <li><a class="dropdown-item" href="../articles/web_only/fst_comparisons.html">Comparisons of Fst in 8 packages/software</a></li>
    <li><a class="dropdown-item" href="../articles/assignment.whitelists.html">Using whitelists in assignment</a></li>
    <li><a class="dropdown-item" href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a></li>
    <li><a class="dropdown-item" href="../articles/life_cycle.html">Life cycle</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/thierrygosselin/assigner" aria-label="GitHub"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Assignment analysis tailored for RADseq data</h1>
      <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/assigner/blob/HEAD/R/assignment_ngs.R" class="external-link"><code>R/assignment_ngs.R</code></a></small>
      <div class="d-none name"><code>assignment_ngs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The arguments in the <code>assignment_ngs</code> function were tailored for the
reality of RADseq data for assignment analysis while
maintaining a reproducible workflow.
Assignment assumptions are listed in the section below.</p>
<ul><li><p><strong>Input file:</strong> various file format are supported (see <code>data</code> argument below).</p></li>
<li><p><strong>Cross-Validations:</strong> Markers can be randomly selected for a classic LOO (Leave-One-Out)
  assignment or chosen based on ranked Fst for a thl
  (Training, Holdout, Leave-one-out) assignment analysis.</p></li>
<li><p><strong>Assignment analysis:</strong> conducted in
  <a href="https://github.com/eriqande/gsi_sim" class="external-link">gsi_sim</a>, a tool
  for doing and simulating genetic stock identification and
  developed by Eric C. Anderson, or
  <a href="https://github.com/thibautjombart/adegenet" class="external-link">adegenet</a>,
 an R package developed by Thibaul Jombart.</p></li>
<li><p><strong>Parallel:</strong> The assignment can be conduncted on multiple CPUs.
  The R GUI is unstable with this functions, I recommend using
  <a href="https://www.rstudio.com/products/rstudio/download/" class="external-link">RStudio</a>.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">assignment_ngs</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  strata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pop.levels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  assignment.analysis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gsim_sim"</span>, <span class="st">"adegenet"</span><span class="op">)</span>,</span>
<span>  markers.sampling <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ranked"</span>, <span class="st">"random"</span><span class="op">)</span>,</span>
<span>  marker.number <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  thl <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  iteration.method <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  subsample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  iteration.subsample <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>Several input format are accepted. assigner uses <span class="pkg">radiator</span>
<code><a href="https://thierrygosselin.github.io/radiator/reference/tidy_genomic_data.html" class="external-link">tidy_genomic_data</a></code> module to import the data.
See function documentation for more details.</p></dd>


<dt id="arg-strata">strata<a class="anchor" aria-label="anchor" href="#arg-strata"></a></dt>
<dd><p>(optional)
The strata file is a tab delimited file with a minimum of 2 columns headers:
<code>INDIVIDUALS</code> and <code>STRATA</code>. Documented in <code><a href="https://thierrygosselin.github.io/radiator/reference/read_strata.html" class="external-link">read_strata</a></code>.
DArT data: a third column <code>TARGET_ID</code> is required.
Documented on <code><a href="https://thierrygosselin.github.io/radiator/reference/read_dart.html" class="external-link">read_dart</a></code>. Also use the strata read function to
blacklist individuals.
Default: <code>strata = NULL</code>.</p></dd>


<dt id="arg-pop-levels">pop.levels<a class="anchor" aria-label="anchor" href="#arg-pop-levels"></a></dt>
<dd><p>(optional, string) This refers to the levels in a factor. In this
case, the id of the pop.
Use this argument to have the pop ordered your way instead of the default
alphabetical or numerical order. e.g. <code>pop.levels = c("QUE", "ONT", "ALB")</code>
instead of the default <code>pop.levels = c("ALB", "ONT", "QUE")</code>.
White spaces in population names are replaced by underscore.
Default: <code>pop.levels = NULL</code>.</p></dd>


<dt id="arg-assignment-analysis">assignment.analysis<a class="anchor" aria-label="anchor" href="#arg-assignment-analysis"></a></dt>
<dd><p>(character) Assignment analysis conducted with
<code>assignment.analysis = "gsi_sim"</code> or
<code>assignment.analysis = "adegenet"</code>.
See <strong>Details</strong> section below for installing
<a href="https://github.com/eriqande/gsi_sim" class="external-link">gsi_sim</a>.</p></dd>


<dt id="arg-markers-sampling">markers.sampling<a class="anchor" aria-label="anchor" href="#arg-markers-sampling"></a></dt>
<dd><p>(character) 2 options for markers selection:</p><ol><li><p><code>markers.sampling == "random"</code> the subset of markers are selected
randomly, this is the classic <em>Leave-One-Out</em> (LOO) assignment.</p></li>
<li><p><code>markers.sampling == "ranked"</code> the subset of markers are first
ranked based on an overall <em>decreasing</em> Fst values.
The Fst is computed with <code><a href="fst_WC84.html">fst_WC84</a></code> function, that uses a fast
implementation of Weir and Cockerham 1984 Fst/Theta equations. This selection
method is used during <em>Training-Holdout-Leave One Out</em> (thl)
assignment. How many markers are selected is controlled with argument <code>thl</code>.</p></li>
</ol></dd>


<dt id="arg-marker-number">marker.number<a class="anchor" aria-label="anchor" href="#arg-marker-number"></a></dt>
<dd><p>(Integer or string of number or "all") The assignment
analysis can use all your markers (default) or a subset of your markers.
e.g. To test 500, 1000, 2000 and all the markers:
<code>marker.number = c(500, 1000, 2000, "all")</code>.
To use only 500 makers <code>marker.number = 500</code>. How those markers are sampled
is determined with the argument <code>markers.sampling</code>, next.
Default = <code>marker.number = "all"</code>.</p></dd>


<dt id="arg-thl">thl<a class="anchor" aria-label="anchor" href="#arg-thl"></a></dt>
<dd><p>(character, integer, proportion) For <code>markers.sampling = "ranked"</code> only.
Several options are available:</p><ol><li><p><code>thl = 1</code>: Only 1 individual sample is used as holdout. This individual is not
participating in the markers ranking. For each marker number,
the analysis will be repeated with all the indiviuals in the data set
(e.g. 500 individuals, 500 times 500, 1000, 2000 markers). This is the default.</p></li>
<li><p><code>proportion</code>: e.g. <code>thl = 0.15</code>, 15 percent of the individuals,
in each strata, are chosen randomly as holdout individuals.</p></li>
<li><p><code>thl = "all"</code>: all individuals are used for ranking (not good) and
the argument <code>iteration.method = 1</code> is set by default. This is for testing
only.</p></li>
</ol><p>Different lists of holdout individuals are generated when the argument
<code>iteration.method</code> (bootstrap) is used.</p></dd>


<dt id="arg-iteration-method">iteration.method<a class="anchor" aria-label="anchor" href="#arg-iteration-method"></a></dt>
<dd><p>(integer)
With <strong>random markers selection</strong> method, the iterations argument =
the number of iterations to repeat marker resampling.
Default: <code>iteration.method = 10</code>.</p>
<p>With <code>marker.number = c(500, 1000)</code> and default iterations setting,
500 markers will be randomly chosen 10 times and 1000 markers will be randomly
chosen 10 times.</p>
<p><strong>Notes:</strong> If all the markers are used, with <code>marker.number = "all"</code>
or in a series of marker number groupings <code>marker.number = c(200, 500, "all")</code>,
the number of iteration is automatically set to 1. The remaining groupings
are unaffected.</p>
<p>With <strong>ranked makers selection</strong> method, using <code>thl = 1</code>, the analysis
will be repeated for each individuals in the data set for every
<code>marker.number</code> selected. With a proportion argument <code>thl = 0.15</code>,
15 percent of individuals in each populations are chosen randomly as holdout
individuals and this process is reapeated the number of times chosen by the
<code>iteration.method</code> value.</p></dd>


<dt id="arg-subsample">subsample<a class="anchor" aria-label="anchor" href="#arg-subsample"></a></dt>
<dd><p>(Integer or Character, optional)
This argument subsample individuals.
With <code>subsample = 36</code>, 36 individuals in each populations are chosen
randomly to represent the dataset. This integer as to be smaller than the
population with min sample size, if higher, the minimum sample size found
in the data will be used as default. In doubt, use <code>subsample = "min"</code>,
the function will use the smallest population sample size found in the data.
The number of times this process is repeated is controlled by the argument
<code>iteration.subsample</code>.
Default: <code>subsample = NULL</code> (no subsampling).</p></dd>


<dt id="arg-iteration-subsample">iteration.subsample<a class="anchor" aria-label="anchor" href="#arg-iteration-subsample"></a></dt>
<dd><p>(Integer) The number of iterations to repeat
subsampling of individuals.
With <code>subsample = 20</code> and <code>iteration.subsample = 10</code>,
20 individuals/populations will be randomly chosen 10 times.
Default: <code>iteration.subsample = 1</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>(optional) To pass further argument for fine-tuning the
function (see advanced section below).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Depending on arguments selected, several folders and files are written:</p><ol><li><p><code>01_radiator_tidy_genomic</code> this is the result of importing the data
with radiator import module <a href="https://thierrygosselin.github.io/radiator/reference/tidy_genomic_data.html" class="external-link">tidy_genomic_data</a>.</p></li>
<li><p><code>assigner_assignment_ngs_args_date@time.tsv</code>: For reproducibility,
the function call, arguments and values used along the default arguments.</p></li>
<li><p><code>assignment.plot.pdf</code>: The assignment figure.</p></li>
<li><p><code>assignment.results.summary.stats.tsv</code>: tibble
of summarized assignment statistics.</p></li>
<li><p><code>assignment.ranked.results.summary.stats.all.subsamples.tsv</code>:
When subsampling is used, this file contains the raw results of all subsample before
summarizing.</p></li>
</ol><p><strong>THL: Training, Holdout, Leave-one-out</strong>:</p>
<p>Intermediate files are written, you can inspect specific iterations
and/or subsample:</p><ol><li><p><code>assignment.ranked.results.iterations.raw.tsv</code>: tibble
with raw intermediate results of all iterations.</p></li>
<li><p><code>assignment.ranked.results.iterations.summary.tsv</code>: tibble
with intermediate summary over iterations.</p></li>
<li><p><code>holdout.individuals.tsv</code>: tibble with the holdout individuals and
associated iteration and random seed number.</p></li>
</ol><p><strong>LOO: Leave-one-out</strong>:</p>
<p>Intermediate files are written, you can inspect specific iterations
and/or subsample:</p><ol><li><p><code>assignment.random.results.iterations.raw.tsv</code>: tibble
with raw intermediate results of all iterations.</p></li>
<li><p><code>markers.random.tsv</code>: tibble with the random markers selected for
each iteration with associated random seed number.</p></li>
</ol><p><strong>Folders</strong></p>
<p>The names have the different iterations <em>i</em>
starting with <code>assignment_</code><em>i</em> contains:</p><ul><li><p><code>assignment_</code><em>i</em><code>.tsv</code>: the assignment result, for the
iteration.</p></li>
<li><p><code>fst.ranked_</code><em>i</em><code>.tsv</code>: for THL method, the ranked Fst
per markers, for the iteration.</p></li>
<li><p><code>gsi_sim_seeds</code>: the <code>gsi_sim</code> random seeds when this program
is used, for the iteration.</p></li>
</ul><p>The output in your global environment is a list. To view the assignment results
<code>$assignment</code> to view the ggplot2 figure <code>$assignment.plot</code>.
See example below.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Using <a href="https://github.com/eriqande/gsi_sim" class="external-link">gsi_sim</a>:</p>
<p><code>assignment_ngs</code> assumes that the command line version of
<a href="https://github.com/eriqande/gsi_sim" class="external-link">gsi_sim</a>
is properly installed into <code>file.path(system.file(package = "assigner"), "bin", "gsi_sim")</code>.
Things are set up so that it will try running gsi_sim, and if it does not find it, the
program will throw an error and ask the user to run <code><a href="install_gsi_sim.html">install_gsi_sim</a></code>
which will do its best to put a usable copy of gsi_sim where it is needed.
To do so, you must be connected to the internet. If that doesn't work, you will
need to compile the program yourself, or get it yourself, and the manually copy
it to <code>file.path(system.file(package = "assigner"), "bin", "gsi_sim")</code>.
To compile <a href="https://github.com/eriqande/gsi_sim" class="external-link">gsi_sim</a>, follow the
instruction here: <a href="https://github.com/eriqande/gsi_sim" class="external-link">https://github.com/eriqande/gsi_sim</a>.</p>
    </div>
    <div class="section level2">
    <h2 id="advance-mode">Advance mode<a class="anchor" aria-label="anchor" href="#advance-mode"></a></h2>



<p>Ideally, forget about this section.
For advance users, through <em>dots-dots-dots ...</em> you can pass several
arguments for fine-tuning the function:</p><ol><li><p><code>adegenet.dapc.opt</code> (optional, character) Argument available only when
using:
<code>assignment.analysis = "adegenet"</code> with
<code>markers.sampling == "random"</code>.</p>
<p>The assignment using dapc can use the optimized alpha score
<code>adegenet.dapc.opt == "optim.a.score"</code> or
cross-validation <code>adegenet.dapc.opt == "xval"</code>
for stability of group membership probabilities.
For fine tuning the trade-off between power of discrimination and over-fitting.
See <span class="pkg">adegenet</span> documentation for more details.
<code>adegenet.dapc.opt == "xval"</code> doesn't work with missing data, so it's
only available with full dataset or <strong>imputed dataset</strong>.
With non imputed data or the default: <code>adegenet.dapc.opt == "optim.a.score"</code>.</p></li>
<li><p><code>adegenet.n.rep</code>: (optional, integer)
When <code>adegenet.dapc.opt == "xval"</code>, the number of replicates to be
carried out at each level of PC retention.
Default: <code>adegenet.n.rep = 30</code>.
See <span class="pkg">adegenet</span> documentation for more details.</p></li>
<li><p><code>adegenet.training</code>: (optional, numeric)
When <code>adegenet.dapc.opt == "xval"</code>, the proportion of data (individuals)
to be used for the training set.
Default: <code>adegenet.training = 0.9</code>, if all groups have &gt;= 10 members.
Otherwise, training.set scales automatically to the largest proportion
that still ensures all groups will be present in both training
and validation sets.
See <span class="pkg">adegenet</span> documentation for more details.</p></li>
<li><p><code>folder</code>: (optional) The name of the folder created in the working
directory to save the files/results. Default: <code>folder = NULL</code> will create
the folder for you with data and time.</p></li>
<li><p><code>filename</code>: (optional) The name of the file written to the directory.
Use the extension ".txt" at the end.
Several info will be appended to the name of the file.
Default <code>assignment_data.txt</code>.</p></li>
<li><p><code>keep.gsi.files</code>: (logical, optional) With the default,
the intermediate input and output gsi_sim files will be deleted from the
directory when finished processing. I you decide to keep the files, with
<code>keep.gsi.files = TRUE</code>, remember to allocate a large chunk of the disk
space for the analysis.
Default: <code>keep.gsi.files = FALSE</code></p></li>
<li><p><code>random.seed</code>: (integer, optional) For reproducibility, set an integer
that will be used inside function that requires randomness. With default,
a random number is generated and printed in the appropriate output.
Default: <code>random.seed = NULL</code>.</p></li>
<li><p><code>full.y.range</code>: (logical, optional) By default the y axis print
min and max values are determied  automatically from the data. It might
be more usefull to see the full range of the scale from 0 to 100, in
this case use <code>full.y.range = TRUE</code>. This can also be modified after
running the analysis. See example below.
Default: <code>full.y.range = FALSE</code>.</p></li>
</ol></div>
    <div class="section level2">
    <h2 id="assumptions">Assumptions<a class="anchor" aria-label="anchor" href="#assumptions"></a></h2>


<ol><li><p><strong>Individuals QC</strong>: Bad individual QC <strong>will bias</strong>
the assignment results.</p><ul><li><p><strong>remove duplicates samples: </strong> when found within the same strata,
duplicates generate a false population signal, when they are found
between strata (yes, I've seen it),
it's generating noise and the core population signal is diluted.</p></li>
<li><p><strong>remove individual with outlier heterozygosity: </strong>
unchecked, outlier individuals based on heterozygosity will generate false population
signal when the sample as lower heterozygosity and match against several
strata (week assignment) when the sample as higher heterozygosity.</p></li>
<li><p><strong>remove individuals with too many missing: </strong> these individuals
will exacerbate or dilute the signal, depending on correlation with heterozygosity
and presence of pattern of missingness.</p></li>
</ul></li>
<li><p><strong>Markers QC</strong>: Bad markers QC <strong>will bias</strong>
the assignment results.</p><ul><li><p><strong>low MAC</strong>: improper Minor Allele Count filtering generate noise.
The LOO and THL methods, both removes samples during model construction,
if MAC is too low, the population core signature is greatly impacted at each iteration.</p></li>
<li><p><strong>Linkage disequilibrium</strong>: remove highly linked markers.</p></li>
<li><p><strong>HWE</strong>: remove markers in very strong Hardy-Weinberg disequilibrium
likely artefactual and/or genotyping errors.</p></li>
</ul></li>
<li><p><strong>Strata</strong>: bad K selection will result in poor assingment results.</p></li>
<li><p><strong>filtered data:</strong> Don't expect to filter your data here.
<span class="pkg">radiator</span> was designed for this, and <code>filter_rad</code> will handle
the issues mentioned above. By default, the function will only remove
monomorphic markers and keep markers in common between strata.</p></li>
</ol></div>
    <div class="section level2">
    <h2 id="life-cycle">Life cycle<a class="anchor" aria-label="anchor" href="#life-cycle"></a></h2>



<p>Map-independent imputation of missing genotype is avaible in my other R
package called <a href="https://github.com/thierrygosselin/grur" class="external-link">grur</a>.</p>
<p>Use <a href="https://github.com/thierrygosselin/grur" class="external-link">grur</a> to :</p><ol><li><p><strong>Visualize your missing data: </strong> before imputing your genotypes,
visualize your missing data.
Several visual tools are available inside <a href="https://github.com/thierrygosselin/grur" class="external-link">grur</a> to
help you decide the best strategy after.</p></li>
<li><p><strong>Optimize: </strong>
use <a href="https://github.com/thierrygosselin/grur" class="external-link">grur</a> imputation module
and other functions to optimize the imputations of your dataset.
You need to test arguments. Failing to conduct tests and adjust imputations arguments
will <strong>generate artifacts</strong> and/or <strong>exacerbate bias</strong>.
Using defaults is not optional here...</p></li>
<li><p><strong>genomic_converter: </strong>
use the output argument inside <a href="https://github.com/thierrygosselin/grur" class="external-link">grur</a>
imputation module to generate the required formats for assigner (e.g. a tidy dataset)</p></li>
</ol><p><strong>Deprecated arguments: </strong></p>
<ul><li><p><code>sampling.method</code>: renamed <code>markers.sampling</code>.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Anderson, Eric C., Robin S. Waples, and Steven T. Kalinowski. (2008)
An improved method for predicting the accuracy of genetic stock identification.
Canadian Journal of Fisheries and Aquatic Sciences 65, 7:1475-1486.</p>
<p>Anderson, E. C. (2010) Assessing the power of informative subsets of
loci for population assignment: standard methods are upwardly biased.
Molecular ecology resources 10, 4:701-710.</p>
<p>Weir BS, Cockerham CC (1984) Estimating F-Statistics for the
Analysis of Population Structure. Evolution, 38, 1358–1370.</p>
<p>Jombart T, Devillard S, Balloux F.
Discriminant analysis of principal components: a new method for the analysis
of genetically structured populations.
BMC Genet. 2010:11: 94. doi:10.1186/1471-2156-11-94</p>
<p>Jombart T, Ahmed I. adegenet 1.3-1: new tools for the analysis
of genome-wide SNP data.
Bioinformatics. 2011:27: 3070–3071. doi:10.1093/bioinformatics/btr521</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="https://github.com/eriqande/gsi_sim" class="external-link">gsi_sim</a> and
<a href="https://github.com/eriqande/rubias" class="external-link">rubias</a></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Thierry Gosselin <a href="mailto:thierrygosselin@icloud.com">thierrygosselin@icloud.com</a> and Eric C. Anderson</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">assignment.treefrog</span> <span class="op">&lt;-</span> <span class="fu">assignment_ngs</span><span class="op">(</span></span></span>
<span class="r-in"><span> data <span class="op">=</span> <span class="st">"batch_1.vcf"</span>,</span></span>
<span class="r-in"><span> strata <span class="op">=</span> <span class="st">"strata.treefrog.tsv"</span>,</span></span>
<span class="r-in"><span> assignment.analysis <span class="op">=</span> <span class="st">"gsi_sim"</span>,</span></span>
<span class="r-in"><span> marker.number <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">5000</span>, <span class="st">"all"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span> markers.sampling <span class="op">=</span> <span class="st">"ranked"</span>,</span></span>
<span class="r-in"><span> thl <span class="op">=</span> <span class="fl">0.3</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-in"><span> <span class="co"># To create a dataframe with the assignment results:</span></span></span>
<span class="r-in"><span>   <span class="va">assignment</span> <span class="op">&lt;-</span> <span class="va">assignment.treefrog</span><span class="op">$</span><span class="va">assignment</span></span></span>
<span class="r-in"><span> <span class="co"># To plot the assignment using ggplot2 and facet</span></span></span>
<span class="r-in"><span>   <span class="va">fig</span> <span class="op">&lt;-</span> <span class="va">assignment.treefrog</span><span class="op">$</span><span class="va">assignment.plot</span></span></span>
<span class="r-in"><span> <span class="co"># To view the full range of y values = Assignment success(%):</span></span></span>
<span class="r-in"><span>   <span class="va">fig</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="co"># Or use the ... argument: full.y.range = TRUE</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

