<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>A fast implementation of Nei's 1987 Fst (overall and paiwise estimates) — fst_NEI87 • assigner</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />
<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="A fast implementation of Nei's 1987 Fst (overall and paiwise estimates) — fst_NEI87" />

<meta property="og:description" content="The function calculates for diploid genomes the classic Nei's Gst (1987), 
Nei's G'st (prime), that comes with a correction for the bias that stems 
from sampling a limited number of populations. Also calculated is Jost's D,
an index of population differentiation that is independent of the amount of
within-population diversity (Jost, 2008). 
Both overall and pairwise Fst can be estimated with 
confidence intervals based on bootstrap of markers (resampling with replacement). 
The function should give identical results at the 4th decimal when tested 
against genet.dist in hierfstat and with 
the Fst computed in Calculate Distances or 
GenoDive.
The fastest computation is still 
GenoDive, 
but here, the R solution computes confidence intervals and it's very fast. 
The computations takes advantage of tidyverse packages and parallel.
The impact of unbalanced design on estimates can be tested by using the 
subsample argument.
Special concerns for genome-wide estimate and filtering bias
During computation, the function first starts by keeping only the polymorphic 
markers in common between the populations. Keep this in mind when filtering 
your markers to use this function characteristic strategically to get 
better genome-wide estimate. This is even more important when your project 
involves more than 2 populations that evolved more by neutral processes 
(e.g. genetic drift) than by natural selection (see the vignette for more details)." />

<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">assigner</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../FEATURES.html">Features</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/fst_confidence_intervals.html">Fst analysis</a>
    </li>
    <li>
      <a href="../articles/assignment.whitelists.html">Using whitelists in assignment</a>
    </li>
    <li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/assigner">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>A fast implementation of Nei's 1987 Fst (overall and paiwise estimates)</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/assigner/blob/master/R/fst_NEI87.R'><code>R/fst_NEI87.R</code></a></small>
    <div class="hidden name"><code>fst_NEI87.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>The function calculates for diploid genomes the classic Nei's Gst (1987), 
Nei's G'st (prime), that comes with a correction for the bias that stems 
from sampling a limited number of populations. Also calculated is Jost's D,
an index of population differentiation that is independent of the amount of
within-population diversity (Jost, 2008). 
Both overall and pairwise Fst can be estimated with 
confidence intervals based on bootstrap of markers (resampling with replacement). 
The function should give identical results <em>at the 4th decimal</em> when tested 
against <code>genet.dist</code> in <code>hierfstat</code> and with 
the Fst computed in <code>Calculate Distances</code> or 
<a href='http://www.bentleydrummer.nl/software/software/GenoDive.html'>GenoDive</a>.
The fastest computation is still 
<a href='http://www.bentleydrummer.nl/software/software/GenoDive.html'>GenoDive</a>, 
but here, the R solution computes confidence intervals and it's very fast. 
The computations takes advantage of <span class="pkg">tidyverse</span> packages and <span class="pkg">parallel</span>.
The impact of unbalanced design on estimates can be tested by using the 
subsample argument.</p>
<p><em>Special concerns for genome-wide estimate and filtering bias</em></p>
<p>During computation, the function first starts by keeping only the polymorphic 
markers in common between the populations. Keep this in mind when filtering 
your markers to use this function characteristic strategically to get 
better genome-wide estimate. This is even more important when your project 
involves more than 2 populations that evolved more by neutral processes 
(e.g. genetic drift) than by natural selection (see the vignette for more details).</p>
    
    </div>

    <pre class="usage"><span class='fu'>fst_NEI87</span>(<span class='no'>data</span>, <span class='kw'>pop.levels</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>pop.labels</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>strata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>holdout.samples</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>pairwise</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>ci</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>iteration.ci</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>quantiles.ci</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>0.025</span>, <span class='fl'>0.975</span>),
  <span class='kw'>subsample</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>iteration.subsample</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>digits</span> <span class='kw'>=</span> <span class='fl'>9</span>,
  <span class='kw'>parallel.core</span> <span class='kw'>=</span> <span class='kw pkg'>parallel</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/parallel/topics/detectCores'>detectCores</a></span>() - <span class='fl'>1</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>A file in the working directory or object in the global environment 
in wide or long (tidy) formats. To import, the function uses internally
<a href='https://github.com/thierrygosselin/radiator'>radiator</a> 
<code><a href='https://www.rdocumentation.org/packages/radiator/topics/tidy_wide'>tidy_wide</a></code>. See details for more info.</p>
<p><em>How to get a tidy data frame ?</em>
<a href='https://github.com/thierrygosselin/radiator'>radiator</a> 
<code><a href='https://www.rdocumentation.org/packages/radiator/topics/tidy_genomic_data'>tidy_genomic_data</a></code> can transform 11 genomic data formats 
in a tidy data frame (VCF, PLINK, genind, genlight, gtypes, genepop,
stacks haplotype file, hierfstat, ...). 
You can also use this function to filter your dataset using
whitelist of markers, blacklist of individuals and genotypes.</p></td>
    </tr>
    <tr>
      <th>pop.levels</th>
      <td><p>(optional, string) This refers to the levels in a factor. In this 
case, the id of the pop.
Use this argument to have the pop ordered your way instead of the default 
alphabetical or numerical order. e.g. <code>pop.levels = c("QUE", "ONT", "ALB")</code> 
instead of the default <code>pop.levels = c("ALB", "ONT", "QUE")</code>. 
Default: <code>pop.levels = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>pop.labels</th>
      <td><p>(optional, string) Use this argument to rename/relabel
your pop or combine your pop. e.g. To combine <code>"QUE"</code> and <code>"ONT"</code> 
into a new pop called <code>"NEW"</code>:
(1) First, define the levels for your pop with <code>pop.levels</code> argument: 
<code>pop.levels = c("QUE", "ONT", "ALB")</code>. 
(2) then, use <code>pop.labels</code> argument: 
<code>pop.levels = c("NEW", "NEW", "ALB")</code>.#' 
To rename <code>"QUE"</code> to <code>"TAS"</code>:
<code>pop.labels = c("TAS", "ONT", "ALB")</code>.
Default: <code>pop.labels = NULL</code>. If you find this too complicated, there is also the
<code>strata</code> argument that can do the same thing, see below.</p></td>
    </tr>
    <tr>
      <th>strata</th>
      <td><p>(optional) A tab delimited file with 2 columns with header:
<code>INDIVIDUALS</code> and <code>STRATA</code>.
If a <code>strata</code> file is specified, the strata file will have
precedence over any grouping found input file (<code>data</code>). 
The <code>STRATA</code> column can be any hierarchical grouping.
Default: <code>strata = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>holdout.samples</th>
      <td><p>(optional) Samples that don't participate in the Fst 
computation (supplementary). Data frame with one column <code>INDIVIDUALS</code>.
Default: <code>holdout.samples = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>pairwise</th>
      <td><p>(logical, optional) With <code>pairwise = TRUE</code>, the 
pairwise NEI87 Fst is calculated between populations. 
Default: <code>pairwise = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>ci</th>
      <td><p>(logical, optional) Compute bootstrapped confidence intervals. 
Default: <code>ci = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>iteration.ci</th>
      <td><p>(integer, optional) The number of iterations for 
the boostraps (resampling with replacement of markers). 
Default: <code>iteration.ci = 100</code>.</p></td>
    </tr>
    <tr>
      <th>quantiles.ci</th>
      <td><p>(double, optional) 
The quantiles for the bootstrapped confidence intervals. 
Default: <code>quantiles.ci = c(0.025,0.975)</code>.</p></td>
    </tr>
    <tr>
      <th>subsample</th>
      <td><p>(Integer or character) 
With <code>subsample = 36</code>, 36 individuals in each populations are chosen
randomly to represent the dataset. With <code>subsample = "min"</code>, the 
minimum number of individual/population found in the data is used automatically.
Default is no subsampling, <code>subsample = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>iteration.subsample</th>
      <td><p>(Integer) The number of iterations to repeat 
subsampling.
With <code>subsample = 20</code> and <code>iteration.subsample = 10</code>,
20 individuals/populations will be randomly chosen 10 times.
Default: <code>iteration.subsample = 1</code>.</p></td>
    </tr>
    <tr>
      <th>digits</th>
      <td><p>(optional, integer) The number of decimal places to be used in 
results.
Default: <code>digits = 9</code>.</p></td>
    </tr>
    <tr>
      <th>parallel.core</th>
      <td><p>(optional) The number of core for parallel computation 
of pairwise Fst. 
If not selected <code>detectCores() - 1</code> is used as default.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>(logical, optional) <code>verbose = TRUE</code> to be chatty 
during execution. 
Default: <code>verbose = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>other parameters passed to the function.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The function returns a list with several objects.
When sumsample is selected the objects end with <code>.subsample</code>.</p><ul>
<li><p><code>$subsampling.individuals</code>: the combinations of individuals and subsamples,</p></li>
<li><p><code>$fst.markers</code>: Nei's Gst, Nei's G'st and Jost's D by markers,</p></li>
<li><p><code>$fst.ranked</code>: Nei's Gst, Nei's G'st and Jost's D by markers ranked by Nei's Gst,</p></li>
<li><p><code>$fst.overall</code>: the overall Nei's Gst, Nei's G'st and Jost's D by markers with confidence intervals.</p></li>
<li><p><code>$fis.markers</code>: the fis by markers,</p></li>
<li><p><code>$fis.overall</code>: the mean fis overall markers with confidence intervals and the number of markers,</p></li>
<li><p><code>$fst.plot</code>: the histogram of the overall G'st per markers,</p></li>
<li><p><code>$pairwise.fst</code>: pairwise Nei's Gst, Nei's G'st and Jost's D in long/tidy data frame and the number of markers ,</p></li>
<li><p><code>$pairwise.fst.upper.matrix</code>: the pairwise fst prime in a upper triangle matrix,</p></li>
<li><p><code>$pairwise.fst.full.matrix</code>: the pairwise fst prime matrix (duplicated upper and lower triangle),</p></li>
<li><p><code>$pairwise.fst.ci.matrix</code>: matrix with pairwise fst prime in the upper triangle
 and the confidence intervals in the lower triangle.</p></li>
<li><p>when subsample is selected <code>$pairwise.fst.subsample.mean</code> is a summary
 of all pairwise comparisons subsample. The mean is calculated accross summary
 statistics.</p></li>
</ul>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><strong>Input data:</strong></p> 
<p>To discriminate the long from the wide format, 
the function <span class="pkg">radiator</span> <code><a href='https://www.rdocumentation.org/packages/radiator/topics/tidy_wide'>tidy_wide</a></code> searches 
for <code>MARKERS or LOCUS</code> in column names (TRUE = long format).
The data frame is tab delimitted.
<strong>Wide format:</strong>
The wide format cannot store metadata info.
The wide format starts with these 2 id columns: 
<code>INDIVIDUALS</code>, <code>POP_ID</code> (that refers to any grouping of individuals), 
the remaining columns are the markers in separate columns storing genotypes.</p>
<p><strong>Long/Tidy format:</strong>
The long format is considered to be a tidy data frame and can store metadata info. 
(e.g. from a VCF see <span class="pkg">radiator</span> <code>tidy_genomic_data</code>). A minimum of 4 columns
are required in the long format: <code>INDIVIDUALS</code>, <code>POP_ID</code>, 
<code>MARKERS or LOCUS</code> and <code>GENOTYPE or GT</code>. The rest are considered metata info.</p>
<p><strong>2 genotypes formats are available:</strong>
6 characters no separator: e.g. <code>001002 of 111333</code> (for heterozygote individual).
6 characters WITH separator: e.g. <code>001/002 of 111/333</code> (for heterozygote individual).
The separator can be any of these: <code>"/", ":", "_", "-", "."</code>.</p>
<p><em>How to get a tidy data frame ?</em>
<span class="pkg">radiator</span> <code>tidy_genomic_data</code> can transform 6 genomic data formats 
in a tidy data frame.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p><strong>Negative Fst</strong> are technical artifact of the computation 
(see Roesti el al. 2012) and are automatically replaced with zero inside 
this function.</p>
<p><strong>Why no p-values ?</strong></p>
<p>There is no null hypothesis testing with <em>P</em>-values.
Confidence intervals provided with the <em>F</em>-statistics 
enables more reliable conclusions about the biological trends in the data.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Nei M. (1987) Molecular Evolutionary Genetics.
Columbia University Press</p>
<p>Meirmans PG, Van Tienderen PH (2004) genotype and genodive: 
two programs for the analysis of genetic diversity of asexual organisms. 
Molecular Ecology Notes, 4, 792-794.</p>
<p>Roesti M, Salzburger W, Berner D. (2012)
Uninformative polymorphisms bias genome scans for signatures of selection. 
BMC Evol Biol., 12:94. doi:10.1111/j.1365-294X.2012.05509.x</p>
<p>Jost L. (2008)
G(ST) and its relatives do not measure differentiation.
Molecular Ecology. 17: 4015-4026.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>From <a href='http://www.bentleydrummer.nl/software/software/GenoDive.html'>GenoDive</a> manual:
<em>'In general, rather than to test differentiation between all pairs of 
populations,
it is adviseable to perform an overall test of population differentiation, 
possibly using a hierarchical population structure, (see AMOVA)'</em></p>
<p>To compute an AMOVA, use <a href='http://www.bentleydrummer.nl/software/software/GenoDive.html'>GenoDive</a></p>
<p><code>hierfstat</code> is available on 
CRAN <a href='http://cran.r-project.org/web/packages/hierfstat/'>http://cran.r-project.org/web/packages/hierfstat/</a> and 
github <a href='https://github.com/jgx65/hierfstat/'>https://github.com/jgx65/hierfstat/</a></p>
<p>Link for <a href='http://www.bentleydrummer.nl/software/software/GenoDive.html'>GenoDive</a></p>
<p>For Fisher's exact test and p-values per markers 
see <code>mmod</code> <code>diff_test</code>.</p>
<p><code><a href='https://www.rdocumentation.org/packages/radiator/topics/tidy_genomic_data'>tidy_genomic_data</a></code> to transform numerous genomic data 
format in tidy data frames.</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"># NOT RUN {
wombat.fst.pairwise <- fst_NEI87(
data = "wombat.filtered.tidy.tsv", 
sep = "/", 
pop.levels = c("ATL", "MLE", "BIS", "PMO", "SOL", "TAS", "ECU"),
holdout.samples = NULL,
pairwise = TRUE,
ci = TRUE, 
iteration.ci = 10000, 
quantiles.ci = c(0.025,0.975),
parallel.core = 8,
verbose = TRUE
)
To get the overall Fst estimate:
wombat.fst.pairwise$fst.overall
To get the Fst plot:
wombat.fst.pairwise$fst.plots
To get the pairwise Fst values with confidence intervals in a data frame:
wombat.fst.pairwise$pairwise.fst
To get the pairwise fst and ci matrix in a data frame:
# rename, data frame, put rownames in column
pairwise.fst.ci.df <- data.frame(pairwise.fst.ci.matrix) %>% add_rownames("POP")
# }
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Thierry Gosselin <a href='mailto:thierrygosselin@icloud.com'>thierrygosselin@icloud.com</a></p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

