<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>A fast implementation of Weir and Cockerham (1984) Fst/Theta (overall and paiwise estimates) — fst_WC84 • assigner</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="A fast implementation of Weir and Cockerham (1984) Fst/Theta (overall and paiwise estimates) — fst_WC84"><meta name="description" content="The function calculates Weir and Cockerham (1984)
Fst for diploid genomes. Both overall and pairwise Fst can be estimated with
confidence intervals based on bootstrap of markers (resampling with replacement).
The function gives identical results at the 9th decimal when tested
against genet.dist in hierfstat. Using the
argument snprelate = TRUE will compute the Fst with
SNPRelate. This implementation
gives slightly upward bias values but provided the fastest computations I know,
but it doesn't compute confidence intervals, for now.
For an R implementation, fst_WC84 is very fast.
The computations takes advantage of dplyr, tidyr, purrr,
parallel and SNPRelate.
The impact of unbalanced design on estimates can be tested by using the
subsample argument (see advance mode section).
Special concerns for genome-wide estimate and filtering bias
During computation, the function first starts by keeping only the polymorphic
markers in common between the populations. Keep this in mind when filtering
your markers to use this function characteristic strategically to get
better genome-wide estimate. This is even more important when your project
involves more than 2 populations that evolved more by neutral processes
(e.g. genetic drift) than by natural selection (see the vignette for more details)."><meta property="og:description" content="The function calculates Weir and Cockerham (1984)
Fst for diploid genomes. Both overall and pairwise Fst can be estimated with
confidence intervals based on bootstrap of markers (resampling with replacement).
The function gives identical results at the 9th decimal when tested
against genet.dist in hierfstat. Using the
argument snprelate = TRUE will compute the Fst with
SNPRelate. This implementation
gives slightly upward bias values but provided the fastest computations I know,
but it doesn't compute confidence intervals, for now.
For an R implementation, fst_WC84 is very fast.
The computations takes advantage of dplyr, tidyr, purrr,
parallel and SNPRelate.
The impact of unbalanced design on estimates can be tested by using the
subsample argument (see advance mode section).
Special concerns for genome-wide estimate and filtering bias
During computation, the function first starts by keeping only the polymorphic
markers in common between the populations. Keep this in mind when filtering
your markers to use this function characteristic strategically to get
better genome-wide estimate. This is even more important when your project
involves more than 2 populations that evolved more by neutral processes
(e.g. genetic drift) than by natural selection (see the vignette for more details)."><meta property="og:image" content="https://thierrygosselin.github.io/assigner/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">assigner</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../FEATURES.html">Features</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/get_started.html">Get started</a></li>
    <li><a class="dropdown-item" href="../articles/fst_confidence_intervals.html">Fst analysis</a></li>
    <li><a class="dropdown-item" href="../articles/web_only/fst_comparisons.html">Comparisons of Fst in 8 packages/software</a></li>
    <li><a class="dropdown-item" href="../articles/assignment.whitelists.html">Using whitelists in assignment</a></li>
    <li><a class="dropdown-item" href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a></li>
    <li><a class="dropdown-item" href="../articles/life_cycle.html">Life cycle</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/thierrygosselin/assigner" aria-label="GitHub"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>A fast implementation of Weir and Cockerham (1984) Fst/Theta (overall and paiwise estimates)</h1>
      <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/assigner/blob/HEAD/R/fst_WC84.R" class="external-link"><code>R/fst_WC84.R</code></a></small>
      <div class="d-none name"><code>fst_WC84.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The function calculates Weir and Cockerham (1984)
Fst for diploid genomes. Both overall and pairwise Fst can be estimated with
confidence intervals based on bootstrap of markers (resampling with replacement).
The function gives identical results <em>at the 9th decimal</em> when tested
against <code>genet.dist</code> in <code>hierfstat</code>. Using the
argument <code>snprelate = TRUE</code> will compute the Fst with
<a href="https://github.com/zhengxwen/SNPRelate" class="external-link">SNPRelate</a>. This implementation
gives slightly upward bias values but provided the fastest computations I know,
but it doesn't compute confidence intervals, for now.
For an R implementation, <code>fst_WC84</code> is very fast.
The computations takes advantage of <span class="pkg">dplyr</span>, <span class="pkg">tidyr</span>, <span class="pkg">purrr</span>,
<span class="pkg">parallel</span> and <span class="pkg">SNPRelate</span>.
The impact of unbalanced design on estimates can be tested by using the
subsample argument (see advance mode section).</p>
<p><em>Special concerns for genome-wide estimate and filtering bias</em></p>
<p>During computation, the function first starts by keeping only the polymorphic
markers in common between the populations. Keep this in mind when filtering
your markers to use this function characteristic strategically to get
better genome-wide estimate. This is even more important when your project
involves more than 2 populations that evolved more by neutral processes
(e.g. genetic drift) than by natural selection (see the vignette for more details).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fst_WC84</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  snprelate <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  strata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pop.levels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pairwise <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ci <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  iteration.ci <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  quantiles.ci <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.025</span>, <span class="fl">0.975</span><span class="op">)</span>,</span>
<span>  heatmap.fst <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  digits <span class="op">=</span> <span class="fl">9</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">2</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>A tidy data frame object in the global environment or
a tidy data frame in wide or long format in the working directory.
<em>How to get a tidy data frame ?</em>
Look into <span class="pkg">radiator</span> <code>tidy_genomic_data</code>.
You can also use this function to filter your dataset using
whitelist of markers, blacklist of individuals and genotypes.</p></dd>


<dt id="arg-snprelate">snprelate<a class="anchor" aria-label="anchor" href="#arg-snprelate"></a></dt>
<dd><p>(optional, logical) Use <a href="https://github.com/zhengxwen/SNPRelate" class="external-link">SNPRelate</a>
to compute the Fst.
It's the fastest computation I've seen so far!</p>
<p>However, testing with different RADseq datasets as shown several upward bias
with <code><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsFst.html" class="external-link">SNPRelate::snpgdsFst</a></code> (last version tested was v.1.16.0).
I compared the results with assigner, hierfstat and strataG
(results available upon request).
The SNPRelate author as not given me good reason to belive the issue is fully
resolved, consequently, the option is no longer available, until further notice.
Default: <code>snprelate = FALSE</code></p></dd>


<dt id="arg-strata">strata<a class="anchor" aria-label="anchor" href="#arg-strata"></a></dt>
<dd><p>(optional, data frame) A tab delimited file with 2 columns with header:
<code>INDIVIDUALS</code> and <code>STRATA</code>.
If a <code>strata</code> file is specified, the strata file will have
precedence over any grouping found data file (<code>data</code>).
The <code>STRATA</code> column can be any hierarchical grouping.
Default: <code>strata = NULL</code>.</p></dd>


<dt id="arg-pop-levels">pop.levels<a class="anchor" aria-label="anchor" href="#arg-pop-levels"></a></dt>
<dd><p>(optional, string) This refers to the levels in a factor. In this
case, the id of the pop.
Use this argument to have the pop ordered your way instead of the default
alphabetical or numerical order. e.g. <code>pop.levels = c("QUE", "ONT", "ALB")</code>
instead of the default <code>pop.levels = c("ALB", "ONT", "QUE")</code>.
Default: <code>pop.levels = NULL</code>.</p></dd>


<dt id="arg-pairwise">pairwise<a class="anchor" aria-label="anchor" href="#arg-pairwise"></a></dt>
<dd><p>(optional, logical) With <code>pairwise = TRUE</code>, the
pairwise WC84 Fst is calculated between populations.
Default: <code>pairwise = FALSE</code>.</p></dd>


<dt id="arg-ci">ci<a class="anchor" aria-label="anchor" href="#arg-ci"></a></dt>
<dd><p>(optional, logical) Compute bootstrapped confidence intervals.
Default: <code>ci = FALSE</code>.</p></dd>


<dt id="arg-iteration-ci">iteration.ci<a class="anchor" aria-label="anchor" href="#arg-iteration-ci"></a></dt>
<dd><p>(optional, integer) The number of iterations for
the boostraps (resampling with replacement of markers).
Default: <code>iteration.ci = 100</code>.</p></dd>


<dt id="arg-quantiles-ci">quantiles.ci<a class="anchor" aria-label="anchor" href="#arg-quantiles-ci"></a></dt>
<dd><p>(optional, double)
The quantiles for the bootstrapped confidence intervals.
Default: <code>quantiles.ci = c(0.025,0.975)</code>.</p></dd>


<dt id="arg-heatmap-fst">heatmap.fst<a class="anchor" aria-label="anchor" href="#arg-heatmap-fst"></a></dt>
<dd><p>(logical) Generate a heatmap with the Fst values in
lower matrix and CI in the upper matrix.
The heatmap can also be generated separately after the Fst
analysis using the separate function: <code><a href="heatmap_fst.html">heatmap_fst</a></code>.
Default: <code>heatmap.fst = FALSE</code>.</p></dd>


<dt id="arg-digits">digits<a class="anchor" aria-label="anchor" href="#arg-digits"></a></dt>
<dd><p>(optional, integer) The number of decimal places to be used in
results.
Default: <code>digits = 9</code>.</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>(optional, character) Give filename prefix, this will trigger
saving results in a directory.
Default: <code>filename = NULL</code>.</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional, integer) The number of core for parallel computation
of pairwise Fst. See also the advance mode section below.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>(optional, logical) <code>verbose = TRUE</code> to be chatty
during execution.
Default: <code>verbose = FALSE</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>other parameters passed to the function.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The function returns a list with several objects.
When sumsample is selected the objects end with <code>.subsample</code>.</p><ul><li><p><code>$subsampling.individuals</code>: the combinations of individuals and subsamples,</p></li>
<li><p><code>$sigma.loc</code>: the variance components per locus, with
      (<code>lsiga</code>: among populations,
      <code>lsigb</code>: among individuals within populations,
      <code>lsigw</code>: within individuals)</p></li>
<li><p><code>$fst.markers</code>: the fst by markers,</p></li>
<li><p><code>$fst.ranked</code>: the fst ranked,</p></li>
<li><p><code>$fst.overall</code>: the mean fst overall markers and the number of markers</p></li>
<li><p><code>$fis.markers</code>: the fis by markers,</p></li>
<li><p><code>$fis.overall</code>: the mean fis overall markers and the number of markers,</p></li>
<li><p><code>$fst.plot</code>: the histogram of the overall Fst per markers,</p></li>
<li><p><code>$pairwise.fst</code>: the pairwise fst in long/tidy data frame and the number of markers ,</p></li>
<li><p><code>$pairwise.fst.upper.matrix</code>: the pairwise fst in a upper triangle matrix,</p></li>
<li><p><code>$pairwise.fst.full.matrix</code>: the pairwise fst matrix (duplicated upper and lower triangle),</p></li>
<li><p><code>$pairwise.fst.ci.matrix</code>: matrix with pairwise fst in the upper triangle
 and the confidence intervals in the lower triangle.</p></li>
<li><p>when subsample is selected <code>$pairwise.fst.subsample.mean</code> is a summary
 of all pairwise comparisons subsample. The mean is calculated accross summary
 statistics.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p><strong>Negative Fst</strong> are technical artifact of the computation
(see Roesti el al. 2012) and are automatically replaced with zero inside
this function.</p>
<p><strong>Why no p-values ?</strong></p>
<p>There is no null hypothesis testing with <em>P</em>-values and its rarely if ever
the appropriate model in population genomics, despite its popularity
with molecular ecologists interested in population differentiation.</p>
<p>"The important scientific question is the real magnitude of the differentiation,
not the smallness of the <em>P</em> value" —Lou Jost</p>
<p>Confidence intervals provided with the <em>F</em>-statistics
enables more reliable conclusions about the biological trends in the data.
A confidence interval describes the statistical uncertainty of the <em>F</em>-statistics
estimate. If the confidence interval include zero, then the null hypothesis
cannot be rejected. If the confidence interval does not include zero, the null
hypothesis can be rejected and you can also have an appreciation of the real
magnitude of the statistical differentiation whether its large or small.</p>
    </div>
    <div class="section level2">
    <h2 id="advance-mode">Advance mode<a class="anchor" aria-label="anchor" href="#advance-mode"></a></h2>



<p><em>dots-dots-dots ...</em> allows to pass several arguments for fine-tuning the function:</p><ol><li><p><code>filter.monomorphic</code> (logical, optional) By default monomorphic
markers present in the dataset are removed (and it should stay that way...).
Default: <code>filter.monomorphic = TRUE</code>.</p></li>
<li><p><code>holdout.samples</code> (optional, data frame) Samples that don't participate in the Fst
computation (supplementary). Data frame with one column <code>INDIVIDUALS</code>.
This argument is used inside assignment analysis.
Default: <code>holdout.samples = NULL</code>.</p></li>
<li><p><code>subsample</code> (Integer or character)
With <code>subsample = 36</code>, 36 individuals in each populations are chosen
randomly to represent the dataset. With <code>subsample = "min"</code>, the
minimum number of individual/population found in the data is used automatically.
Default is no subsampling, <code>subsample = NULL</code>.</p></li>
<li><p><code>iteration.subsample</code> (Integer) The number of iterations to repeat
subsampling.
With <code>subsample = 20</code> and <code>iteration.subsample = 10</code>,
20 individuals/populations will be randomly chosen 10 times.
Default: <code>iteration.subsample = 1</code>.</p></li>
<li><p><code>calibrate.alleles</code> (logical)
Un-calibrated alleles can bias estimate and by default the function expect that
the REF/ALT alleles are calibrated. Using <code>calibrate.alleles = TRUE</code>,
can take a bit more time.
Default: <code>calibrate.alleles = FALSE</code>.</p></li>
<li><p><code>forking</code> (logical)
For macOS, forking is a much faster option, despite having the reputation of
being unreliable in GUI environment like RStudio.
The implementation in assigner is stable and up to 3-4 times faster.
It's easier on memory.
By default the function uses <code>multisession</code> from the future package
to implement parallel processing. If you have macOS, test by toggling to
<code>TRUE</code> this argument.
Default: <code>forking = FALSE</code>.</p></li>
</ol></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Excoffier L, Smouse PE, Quattro JM.
Analysis of molecular variance inferred from metric distances among
DNA haplotypes: application to human mitochondrial DNA restriction data.
Genetics. 1992;131: 479-491.</p>
<p>Jost L, D vs. GST: Response to Heller and Siegismund (2009) and
Ryman and Leimar (2009).
Molecular Ecology. 2009; 18:10 2088-2091.
https://doi.org/10.1111/j.1365-294x.2009.04186.x</p>
<p>Meirmans PG, Van Tienderen PH (2004) genotype and genodive:
two programs for the analysis of genetic diversity of asexual organisms.
Molecular Ecology Notes, 4, 792-794.</p>
<p>Michalakis Y, Excoffier L.
A generic estimation of population
subdivision using distances between alleles with special reference for
microsatellite loci.
Genetics. 1996;142: 1061-1064.</p>
<p>Weir BS, Cockerham CC (1984) Estimating F-Statistics for the
Analysis of Population Structure.
Evolution, 38, 1358-1370.</p>
<p>Roesti M, Salzburger W, Berner D. (2012)
Uninformative polymorphisms bias genome scans for signatures of selection.
BMC Evol Biol., 12:94. doi:10.1111/j.1365-294X.2012.05509.x</p>
<p>Zheng X, Levine D, Shen J, Gogarten SM, Laurie C, Weir BS.
A high-performance computing toolset for relatedness and principal component
analysis of SNP data. Bioinformatics. 2012;28: 3326-3328.
doi:10.1093/bioinformatics/bts606</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>From <a href="http://www.bentleydrummer.nl/software/software/GenoDive.html" class="external-link">GenoDive</a> manual:
<em>'In general, rather than to test differentiation between all pairs of
populations,
it is advisable to perform an overall test of population differentiation,
possibly using a hierarchical population structure, (see AMOVA)'</em>.
To compute an AMOVA, use <a href="http://www.bentleydrummer.nl/software/software/GenoDive.html" class="external-link">GenoDive</a>
or <code>Phi_st_Meirmans</code> in <code>mmod</code>.</p>
<p><a href="https://github.com/jgx65/hierfstat/" class="external-link">hierfstat</a></p>
<p>For Fisher's exact test and p-values per markers
see <code>mmod</code> <code>diff_test</code>.</p>
<p><strong>Vignette for this function:</strong> <a href="https://www.dropbox.com/s/tiq4yenzmgzc2f5/fst_confidence_intervals.html?dl=0" class="external-link">how to do the pairwise and overall Fst with confidence intervals and build the phylogenetic tree</a></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Thierry Gosselin <a href="mailto:thierrygosselin@icloud.com">thierrygosselin@icloud.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">wombat.fst.pairwise</span> <span class="op">&lt;-</span> <span class="fu">fst_WC84</span><span class="op">(</span></span></span>
<span class="r-in"><span>    data <span class="op">=</span> <span class="st">"wombat.filtered.tidy.tsv"</span>,</span></span>
<span class="r-in"><span>    pop.levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ATL"</span>, <span class="st">"MLE"</span>, <span class="st">"BIS"</span>, <span class="st">"PMO"</span>, <span class="st">"SOL"</span>, <span class="st">"TAS"</span>, <span class="st">"ECU"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    pairwise <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    ci <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    iteration.ci <span class="op">=</span> <span class="fl">10000</span>,</span></span>
<span class="r-in"><span>    quantiles.ci <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.025</span>,<span class="fl">0.975</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    parallel.core <span class="op">=</span> <span class="fl">8</span>,</span></span>
<span class="r-in"><span>    verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    filename <span class="op">=</span> <span class="st">"wombat"</span>,</span></span>
<span class="r-in"><span>    heatmap.fst <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To get the overall Fst estimate:</span></span></span>
<span class="r-in"><span><span class="va">wombat.fst.pairwise</span><span class="op">$</span><span class="va">fst.overall</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To get the Fst plot:</span></span></span>
<span class="r-in"><span><span class="va">wombat.fst.pairwise</span><span class="op">$</span><span class="va">fst.plot</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#To get the pairwise Fst values with confidence intervals in a data frame:</span></span></span>
<span class="r-in"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="va">wombat.fst.pairwise</span><span class="op">$</span><span class="va">pairwise.fst</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

